<!DOCTYPE HTML>
<html lang="zh-CN">
<head>
    <!--Setting-->
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, user-scalable=no, initial-scale=1.0, maximum-scale=1.0, minimum-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1">
    <meta http-equiv="Cache-Control" content="no-siteapp">
    <meta http-equiv="Cache-Control" content="no-transform">
    <meta name="renderer" content="webkit|ie-comp|ie-stand">
    <meta name="apple-mobile-web-app-capable" content="木叶村">
    <meta name="apple-mobile-web-app-status-bar-style" content="black">
    <meta name="format-detection" content="telephone=no,email=no,adress=no">
    <meta name="browsermode" content="application">
    <meta name="screen-orientation" content="portrait">
    <link rel="dns-prefetch" href="https://jeweizhong.github.io">
    <!--SEO-->

    <meta name="keywords" content="软件" />


    <meta name="description" content="samtools的说明文档：http://samtools.sourceforge.net/samtools.shtmlsamtools是一个用于操作sam和bam文件的工具合集。包含有许多命令..." />



<meta name="robots" content="all" />
<meta name="google" content="all" />
<meta name="googlebot" content="all" />
<meta name="verify" content="all" />

    <!--Title-->


<title>(转载)samtools常用命令详解 | 木叶村</title>


    <link rel="alternate" href="/atom.xml" title="木叶村" type="application/atom+xml">


    <link rel="icon" href="/bitbug_favicon.ico">

    



<link rel="stylesheet" href="/css/bootstrap.min.css?rev=3.3.7">
<link rel="stylesheet" href="/css/font-awesome.min.css?rev=4.5.0">
<link rel="stylesheet" href="/css/style.css?rev=@@hash">




    
	<div class="hide">
		<script type="text/javascript">
			var cnzz_protocol = (("https:" == document.location.protocol) ? " https://" : " http://");document.write(unescape("%3Cspan class='cnzz_stat_icon_1263868967 hide' %3E%3Cscript%20src%3D%22https%3A%2F%2Fs95.cnzz.com%2Fz_stat.php%3Fweb_id%3D1272564536%22%3E%3C%2Fscript%3E%3C/span%3E%3Cscript src='" + cnzz_protocol + "s19.cnzz.com/z_stat.php%3Fid%3D1263868967%26show%3Dpic1' type='text/javascript'%3E%3C/script%3E"));
		</script>
	</div>






    

</head>


<!--[if lte IE 8]>
<style>
    html{ font-size: 1em }
</style>
<![endif]-->
<!--[if lte IE 9]>
<div style="ie">你使用的浏览器版本过低，为了你更好的阅读体验，请更新浏览器的版本或者使用其他现代浏览器，比如Chrome、Firefox、Safari等。</div>
<![endif]-->

<body>
    <header class="main-header"  style="background-image:url(http://snippet.shenliyang.com/img/banner.jpg)"  >
    <div class="main-header-box">
        <a class="header-avatar" href="/" title='Naruto'>
            <img src="/img/avatar.jpg" alt="logo头像" class="img-responsive center-block">
        </a>
        <div class="branding">
        	<!--<h2 class="text-hide">Snippet主题,从未如此简单有趣</h2>-->
            
                <h2> 无人为孤岛，一书一世界 </h2>
            
    	</div>
    </div>
</header>
    <nav class="main-navigation">
    <div class="container">
        <div class="row">
            <div class="col-sm-12">
                <div class="navbar-header"><span class="nav-toggle-button collapsed pull-right" data-toggle="collapse" data-target="#main-menu" id="mnav">
                    <span class="sr-only"></span>
                        <i class="fa fa-bars"></i>
                    </span>
                    <a class="navbar-brand" href="https://jeweizhong.github.io">木叶村</a>
                </div>
                <div class="collapse navbar-collapse" id="main-menu">
                    <ul class="menu">
                        
                            <li role="presentation" class="text-center">
                                <a href="/"><i class="fa "></i>首页</a>
                            </li>
                        
                            <li role="presentation" class="text-center">
                                <a href="/categories/计算机"><i class="fa "></i>计算机</a>
                            </li>
                        
                            <li role="presentation" class="text-center">
                                <a href="/categories/生信"><i class="fa "></i>生信</a>
                            </li>
                        
                            <li role="presentation" class="text-center">
                                <a href="/categories/数学与统计"><i class="fa "></i>数学与统计</a>
                            </li>
                        
                            <li role="presentation" class="text-center">
                                <a href="/archives/"><i class="fa "></i>时间轴</a>
                            </li>
                        
                    </ul>
                </div>
            </div>
        </div>
    </div>
</nav>
    <section class="content-wrap">
        <div class="container">
            <div class="row">
                <main class="col-md-8 main-content m-post">
                    <p id="process"></p>
<article class="post">
    <div class="post-head">
        <h1 id="(转载)samtools常用命令详解">
            
	            (转载)samtools常用命令详解
            
        </h1>
        <div class="post-meta">
    
        <span class="categories-meta fa-wrap">
            <i class="fa fa-folder-open-o"></i>
            <a class="category-link" href="/categories/生信/">生信</a>
        </span>
    

    
        <span class="fa-wrap">
            <i class="fa fa-tags"></i>
            <span class="tags-meta">
                
                    <a class="tag-link" href="/tags/软件/">软件</a>
                
            </span>
        </span>
    

    
        
        <span class="fa-wrap">
            <i class="fa fa-clock-o"></i>
            <span class="date-meta">2018/04/29</span>
        </span>
        
    
</div>
            
            
            <p class="fa fa-exclamation-triangle warning">
                本文于<strong>382</strong>天之前发表，文中内容可能已经过时。
            </p>
        
    </div>
    
    <div class="post-body post-content">
        <p>samtools的说明文档：<a href="http://samtools.sourceforge.net/samtools.shtml" target="_blank" rel="noopener">http://samtools.sourceforge.net/samtools.shtml</a><br>samtools是一个用于操作sam和bam文件的工具合集。包含有许多命令。以下是常用命令的介绍</p>
<h2 id="view"><a href="#view" class="headerlink" title="view"></a>view</h2><p>view命令的主要功能是：将sam文件转换成bam文件；然后对bam文件进行各种操作，比如数据的排序(不属于本命令的功能)和提取(这些操作是对bam文件进行的，因而当输入为sam文件的时候，不能进行该操作)；最后将排序或提取得到的数据输出为bam或sam（默认的）格式。</p>
<p>bam文件优点：bam文件为二进制文件，占用的磁盘空间比sam文本文件小；利用bam二进制文件的运算速度快。</p>
<p>view命令中，对sam文件头部的输入(-t或-T）和输出(-h)是单独的一些参数来控制的。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br></pre></td><td class="code"><pre><span class="line">Usage: samtools view [options] &lt;in.bam&gt;|&lt;in.sam&gt; [region1 [...]]</span><br><span class="line"> 默认情况下不加 region，则是输出所有的 region.</span><br><span class="line"></span><br><span class="line">Options: -b       output BAM</span><br><span class="line">                  默认下输出是 SAM 格式文件，该参数设置输出 BAM 格式</span><br><span class="line">         -h       print header for the SAM output</span><br><span class="line">                  默认下输出的 sam 格式文件不带 header，该参数设定输出sam文件时带 header 信息</span><br><span class="line">         -H       print header only (no alignments)</span><br><span class="line">         -S       input is SAM</span><br><span class="line">                  默认下输入是 BAM 文件，若是输入是 SAM 文件，则最好加该参数，否则有时候会报错。</span><br><span class="line">         -u       uncompressed BAM output (force -b)</span><br><span class="line">                  该参数的使用需要有-b参数，能节约时间，但是需要更多磁盘空间。</span><br><span class="line">         -c       Instead of printing the alignments, only count them and print the </span><br><span class="line">                  total number. All filter options, such as ‘-f’, ‘-F’ and ‘-q’ , </span><br><span class="line">                  are taken into account.</span><br><span class="line">         -1       fast compression (force -b)</span><br><span class="line">         -x       output FLAG in HEX (samtools-C specific)</span><br><span class="line">         -X       output FLAG in string (samtools-C specific)</span><br><span class="line">         -c       print only the count of matching records</span><br><span class="line">         -L FILE  output alignments overlapping the input BED FILE [null]</span><br><span class="line">         -t FILE  list of reference names and lengths (force -S) [null]</span><br><span class="line">                  使用一个list文件来作为header的输入</span><br><span class="line">         -T FILE  reference sequence file (force -S) [null]</span><br><span class="line">                  使用序列fasta文件作为header的输入</span><br><span class="line">         -o FILE  output file name [stdout]</span><br><span class="line">         -R FILE  list of read groups to be outputted [null]</span><br><span class="line">         -f INT   required flag, 0 for unset [0]</span><br><span class="line">         -F INT   filtering flag, 0 for unset [0] </span><br><span class="line">                  Skip alignments with bits present in INT [0]</span><br><span class="line">                  数字4代表该序列没有比对到参考序列上</span><br><span class="line">                  数字8代表该序列的mate序列没有比对到参考序列上</span><br><span class="line">         -q INT   minimum mapping quality [0]</span><br><span class="line">         -l STR   only output reads in library STR [null]</span><br><span class="line">         -r STR   only output reads in read group STR [null]</span><br><span class="line">         -s FLOAT fraction of templates to subsample; integer part as seed [-1]</span><br><span class="line">         -?       longer help</span><br></pre></td></tr></table></figure>
<p>例子：<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">将sam文件转换成bam文件</span><br><span class="line">$ samtools view -bS abc.sam &gt; abc.bam</span><br><span class="line">$ samtools view -b -S abc.sam -o abc.bam</span><br><span class="line"></span><br><span class="line">提取比对到参考序列上的比对结果</span><br><span class="line">$ samtools view -bF 4 abc.bam &gt; abc.F.bam</span><br><span class="line"></span><br><span class="line">提取paired reads中两条reads都比对到参考序列上的比对结果，只需要把两个4+8的值12作为过滤参数即可</span><br><span class="line">$ samtools view -bF 12 abc.bam &gt; abc.F12.bam</span><br><span class="line"></span><br><span class="line">提取没有比对到参考序列上的比对结果</span><br><span class="line">$ samtools view -bf 4 abc.bam &gt; abc.f.bam</span><br><span class="line"></span><br><span class="line">提取bam文件中比对到caffold1上的比对结果，并保存到sam文件格式</span><br><span class="line">$ samtools view abc.bam scaffold1 &gt; scaffold1.sam</span><br><span class="line"></span><br><span class="line">提取scaffold1上能比对到30k到100k区域的比对结果</span><br><span class="line">$ samtools view abc.bam scaffold1:30000-100000 $gt; scaffold1_30k-100k.sam</span><br><span class="line"></span><br><span class="line">根据fasta文件，将 header 加入到 sam 或 bam 文件中</span><br><span class="line">$ samtools view -T genome.fasta -h scaffold1.sam &gt; scaffold1.h.sam</span><br></pre></td></tr></table></figure></p>
<h2 id="sort"><a href="#sort" class="headerlink" title="sort"></a>sort</h2><p>sort对bam文件进行排序。<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">Usage: samtools sort [-n] [-m &lt;maxMem&gt;] &lt;in.bam&gt; &lt;out.prefix&gt;  </span><br><span class="line">-m 参数默认下是 500,000,000 即500M（不支持K，M，G等缩写）。对于处理大数据时，如果内存够用，则设置大点的值，以节约时间。</span><br><span class="line">-n 设定排序方式按short reads的ID排序。默认下是按序列在fasta文件中的顺序（即header）和序列从左往右的位点排序。</span><br></pre></td></tr></table></figure></p>
<p>例子：<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$ samtools sort abc.bam abc.sort</span><br><span class="line">$ samtools view abc.sort.bam | less -S</span><br></pre></td></tr></table></figure></p>
<h2 id="merge"><a href="#merge" class="headerlink" title="merge"></a>merge</h2><p>将2个或2个以上的已经sort了的bam文件融合成一个bam文件。融合后的文件不需要则是已经sort过了的。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">Usage:   samtools merge [-nr] [-h inh.sam] &lt;out.bam&gt; &lt;in1.bam&gt; &lt;in2.bam&gt;[...]</span><br><span class="line"></span><br><span class="line">Options: -n       sort by read names</span><br><span class="line">         -r       attach RG tag (inferred from file names)</span><br><span class="line">         -u       uncompressed BAM output</span><br><span class="line">         -f       overwrite the output BAM if exist</span><br><span class="line">         -1       compress level 1</span><br><span class="line">         -R STR   merge file in the specified region STR [all]</span><br><span class="line">         -h FILE  copy the header in FILE to &lt;out.bam&gt; [in1.bam]</span><br><span class="line"></span><br><span class="line">Note: Samtools&apos; merge does not reconstruct the @RG dictionary in the header. Users</span><br><span class="line">      must provide the correct header with -h, or uses Picard which properly maintains</span><br><span class="line">      the header dictionary in merging.</span><br></pre></td></tr></table></figure>
<h2 id="index"><a href="#index" class="headerlink" title="index"></a>index</h2><p>必须对bam文件进行默认情况下的排序后，才能进行index。否则会报错。</p>
<p>建立索引后将产生后缀为.bai的文件，用于快速的随机处理。很多情况下需要有bai文件的存在，特别是显示序列比对情况下。比如samtool的tview命令就需要；gbrowse2显示reads的比对图形的时候也需要。<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">Usage: samtools index &lt;in.bam&gt; [out.index]</span><br></pre></td></tr></table></figure></p>
<p>例子：<br>以下两种命令结果一样<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$ samtools index abc.sort.bam</span><br><span class="line">$ samtools index abc.sort.bam abc.sort.bam.bai</span><br></pre></td></tr></table></figure></p>
<h2 id="faidx"><a href="#faidx" class="headerlink" title="faidx"></a>faidx</h2><p>对fasta文件建立索引,生成的索引文件以.fai后缀结尾。该命令也能依据索引文件快速提取fasta文件中的某一条（子）序列</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">Usage: samtools faidx &lt;in.bam&gt; [ [...]]</span><br><span class="line"></span><br><span class="line">对基因组文件建立索引</span><br><span class="line">$ samtools faidx genome.fasta</span><br><span class="line">生成了索引文件genome.fasta.fai,是一个文本文件，分成了5列。第一列是子序列的名称；</span><br><span class="line">第二列是子序列的长度；个人认为“第三列是序列所在的位置”，因为该数字从上往下逐渐变大，</span><br><span class="line">最后的数字是genome.fasta文件的大小；第4和5列不知是啥意思。于是通过此文件，可以定位子序列在fasta文件在磁盘上的存放位置，直接快速调出子序列。</span><br><span class="line"></span><br><span class="line">由于有索引文件，可以使用以下命令很快从基因组中提取到fasta格式的子序列</span><br><span class="line">$ samtools faidx genome.fasta scffold_10 &gt; scaffold_10.fasta</span><br></pre></td></tr></table></figure>
<h2 id="tview"><a href="#tview" class="headerlink" title="tview"></a>tview</h2><p>tview能直观的显示出reads比对基因组的情况，和基因组浏览器有点类似。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">Usage: samtools tview &lt;aln.bam&gt; [ref.fasta]</span><br><span class="line"></span><br><span class="line">当给出参考基因组的时候，会在第一排显示参考基因组的序列，否则，第一排全用N表示。</span><br><span class="line">按下 g ，则提示输入要到达基因组的某一个位点。例子“scaffold_10:1000&quot;表示到达第10号scaffold的第1000个碱基位点处。</span><br><span class="line">使用H(左）J（上）K（下）L（右）移动显示界面。大写字母移动快，小写字母移动慢。</span><br><span class="line">使用空格建向左快速移动（和 L 类似），使用Backspace键向左快速移动（和 H 类似）。</span><br><span class="line">Ctrl+H 向左移动1kb碱基距离； Ctrl+L 向右移动1kb碱基距离</span><br><span class="line">可以用颜色标注比对质量，碱基质量，核苷酸等。30～40的碱基质量或比对质量使用白色表示；</span><br><span class="line">20～30黄色；10～20绿色；0～10蓝色。</span><br><span class="line">使用点号&apos;.&apos;切换显示碱基和点号；使用r切换显示read name等</span><br><span class="line">还有很多其它的使用说明，具体按 ？ 键来查看。</span><br></pre></td></tr></table></figure>
<h2 id="flagstat"><a href="#flagstat" class="headerlink" title="flagstat"></a>flagstat</h2><p>给出BAM文件的比对结果<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line">Usage: samtools flagstat &lt;in.bam&gt;</span><br><span class="line"></span><br><span class="line">$ samtools flagstat example.bam</span><br><span class="line">11945742 + 0 in total (QC-passed reads + QC-failed reads)</span><br><span class="line">总共的reads数</span><br><span class="line">0 + 0 duplicates</span><br><span class="line">7536364 + 0 mapped (63.09%:-nan%)</span><br><span class="line">总体上reads的匹配率</span><br><span class="line">11945742 + 0 paired in sequencing</span><br><span class="line">有多少reads是属于paired reads</span><br><span class="line">5972871 + 0 read1</span><br><span class="line">reads1中的reads数</span><br><span class="line">5972871 + 0 read2</span><br><span class="line">reads2中的reads数</span><br><span class="line">6412042 + 0 properly paired (53.68%:-nan%)</span><br><span class="line">完美匹配的reads数：比对到同一条参考序列，并且两条reads之间的距离符合设置的阈值</span><br><span class="line">6899708 + 0 with itself and mate mapped</span><br><span class="line">paired reads中两条都比对到参考序列上的reads数</span><br><span class="line">636656 + 0 singletons (5.33%:-nan%)</span><br><span class="line">单独一条匹配到参考序列上的reads数，和上一个相加，则是总的匹配上的reads数。</span><br><span class="line">469868 + 0 with mate mapped to a different chr</span><br><span class="line">paired reads中两条分别比对到两条不同的参考序列的reads数</span><br><span class="line">243047 + 0 with mate mapped to a different chr (mapQ&gt;=5)</span><br><span class="line">同上一个，只是其中比对质量&gt;=5的reads的数量</span><br></pre></td></tr></table></figure></p>
<h2 id="depth"><a href="#depth" class="headerlink" title="depth"></a>depth</h2><p>得到每个碱基位点的测序深度,并输出到标准输出。<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">Usage: bam2depth [-r reg] [-q baseQthres] [-Q mapQthres] [-b in.bed] &lt;in1.bam&gt; [...]</span><br></pre></td></tr></table></figure></p>
<h2 id="其它有用的命令"><a href="#其它有用的命令" class="headerlink" title="其它有用的命令"></a>其它有用的命令</h2><p>reheader 替换bam文件的头<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ samtools reheader &lt;in.header.sam&gt; &lt;in.bam&gt;</span><br></pre></td></tr></table></figure></p>
<p>cat 连接多个bam文件，适用于非sorted的bam文件<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ samtools cat [-h header.sam] [-o out.bam] &lt;in1.bam&gt; &lt;in2.bam&gt; [ ... ]</span><br></pre></td></tr></table></figure></p>
<p>idxstats 统计一个表格，4列，分别为”序列名，序列长度，比对上的reads数，unmapped reads number”。第4列应该是paired reads中有一端能匹配到该scaffold上，而另外一端不匹配到任何scaffolds上的reads数。<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ samtools idxstats &lt;aln.bam&gt;</span><br></pre></td></tr></table></figure></p>
<h2 id="将bam文件转换为fastq文件"><a href="#将bam文件转换为fastq文件" class="headerlink" title="将bam文件转换为fastq文件"></a>将bam文件转换为fastq文件</h2><p>有时候，我们需要提取出比对到一段参考序列的reads，进行小范围的分析，以利于debug等。这时需要将bam或sam文件转换为fastq格式。<br>该网站提供了一个bam转换为fastq的程序：<a href="http://www.hudsonalpha.org/gsl/information/software/bam2fastq" target="_blank" rel="noopener">http://www.hudsonalpha.org/gsl/information/software/bam2fastq</a></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">$ wget http://www.hudsonalpha.org/gsl/static/software/bam2fastq-1.1.0.tgz</span><br><span class="line">$ tar zxf bam2fastq-1.1.0.tgz</span><br><span class="line">$ cd bam2fastq-1.1.0</span><br><span class="line">$ make</span><br><span class="line">$ ./bam2fastq &lt;in.bam&gt;</span><br></pre></td></tr></table></figure>
<h2 id="mpileup"><a href="#mpileup" class="headerlink" title="mpileup"></a>mpileup</h2><p>samtools还有个非常重要的命令mpileup，以前为pileup。该命令用于生成bcf文件，再使用bcftools进行SNP和Indel的分析。bcftools是samtool中附带的软件，在samtools的安装文件夹中可以找到。</p>
<p>最常用的参数有2： <code>-f</code> 来输入有索引文件的fasta参考序列； <code>-g</code> 输出到bcf格式。用法和最简单的例子如下</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">Usage: samtools mpileup [-EBug] [-C capQcoef] [-r reg] [-f in.fa] [-l list] [-M capMapQ] [-Q minBaseQ] [-q minMapQ] in.bam [in2.bam [...]]</span><br><span class="line"></span><br><span class="line">$ samtools mpileup -f genome.fasta abc.bam &gt; abc.txt</span><br><span class="line">$ samtools mpileup -gSDf genome.fasta abc.bam &gt; abc.bcf</span><br><span class="line">$ samtools mpileup -guSDf genome.fasta abc.bam | \</span><br><span class="line">           bcftools view -cvNg - &gt; abc.vcf</span><br></pre></td></tr></table></figure>
<p>mpileup不使用<code>-u</code>或<code>-g</code>参数时，则不生成二进制的bcf文件，而生成一个文本文件(输出到标准输出)。该文本文件统计了参考序列中每个碱基位点的比对情况；该文件每一行代表了参考序列中某一个碱基位点的比对结果。比如：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">scaffold_1      2841    A       11      ,,,...,....     BHIGDGIJ?FF</span><br><span class="line">scaffold_1      2842    C       12      ,$,,...,....^I. CFGEGEGGCFF+</span><br><span class="line">scaffold_1      2843    G       11      ,,...,.....     FDDDDCD?DD+</span><br><span class="line">scaffold_1      2844    G       11      ,,...,.....     FA?AAAA&lt;AA+</span><br><span class="line">scaffold_1      2845    G       11      ,,...,.....     F656666166*</span><br><span class="line">scaffold_1      2846    A       11      ,,...,.....     (1.1111)11*</span><br><span class="line">scaffold_1      2847    A       11      ,,+9acggtgaag.+9ACGGTGAAT.+9ACGGTGAAG.+9ACGGTGAAG,+9acggtgaag.+9ACGGTGAAG.+9ACGGTGAAG.+9ACGGTGAAG.+9ACGGTGAAG.+9ACGGTGAAG       %.+....-..)</span><br><span class="line">scaffold_1      2848    N       11      agGGGgGGGGG     !!$!!!!!!!!</span><br><span class="line">scaffold_1      2849    A       11      c$,...,.....    !0000000000</span><br><span class="line">scaffold_1      2850    A       10      ,...,.....      353333333</span><br></pre></td></tr></table></figure>
<p>mpileup生成的结果包含6行：参考序列名；位置；参考碱基；比对上的reads数；比对情况；比对上的碱基的质量。其中第5列比较复杂,解释如下：<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">1  ‘.’代表与参考序列正链匹配。</span><br><span class="line">2  ‘,’代表与参考序列负链匹配。</span><br><span class="line">3  ‘ATCGN’代表在正链上的不匹配。</span><br><span class="line">4  ‘atcgn’代表在负链上的不匹配。</span><br><span class="line">5  ‘*’代表模糊碱基</span><br><span class="line">6  ‘^’代表匹配的碱基是一个read的开始；’^’后面紧跟的ascii码减去33代表比对质量；这两个符号修饰的是后面的碱基，其后紧跟的碱基(.,ATCGatcgNn)代表该read的第一个碱基。</span><br><span class="line">7  ‘$’代表一个read的结束，该符号修饰的是其前面的碱基。</span><br><span class="line">8  正则式’\+[0-9]+[ACGTNacgtn]+’代表在该位点后插入的碱基；比如上例中在scaffold_1的2847后插入了9个长度的碱基acggtgaag。表明此处极可能是indel。</span><br><span class="line">9  正则式’-[0-9]+[ACGTNacgtn]+’代表在该位点后缺失的碱基；</span><br></pre></td></tr></table></figure></p>
<p><strong>pileup具体的参数如下：</strong></p>
<p><strong>输入参数</strong><br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">-6       Assume the quality is in the Illumina 1.3+ encoding. </span><br><span class="line"> -A Do not skip anomalous read pairs in variant calling. </span><br><span class="line">-B       Disable probabilistic realignment for the computation of base alignment quality (BAQ). BAQ is the Phred-scaled probability of a read base being misaligned. Applying this option greatly helps to reduce false SNPs caused by misalignments. </span><br><span class="line">-b FILE  List of input BAM files, one file per line [null]</span><br><span class="line">-C INT   Coefficient for downgrading mapping quality for reads containing excessive mismatches. Given a read with a phred-scaled probability q of being generated from the mapped position, the new mapping quality is about sqrt((INT-q)/INT)*INT. A zero value disables this functionality; if enabled, the recommended value for BWA is 50. [0] </span><br><span class="line">-d INT   At a position, read maximally INT reads per input BAM. [250] </span><br><span class="line">-E       Extended BAQ computation. This option helps sensitivity especially for MNPs, but may hurt specificity a little bit. </span><br><span class="line">-f FILE  The faidx-indexed reference file in the FASTA format. The file can be optionally compressed by razip. [null] </span><br><span class="line">-l FILE  BED or position list file containing a list of regions or sites where pileup or BCF should be generated [null] </span><br><span class="line">-M INT       cap mapping quality at INT [60]</span><br><span class="line">-q INT 	Minimum mapping quality for an alignment to be used [0] </span><br><span class="line">-Q INT 	Minimum base quality for a base to be considered [13]</span><br><span class="line">-r STR 	Only generate pileup in region STR [all sites]</span><br></pre></td></tr></table></figure></p>
<p><strong>输出参数</strong><br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"> -D 	Output per-sample read depth (require -g/-u)</span><br><span class="line"> -g 	Compute genotype likelihoods and output them in the binary call format (BCF). </span><br><span class="line"> -S 	Output per-sample Phred-scaled strand bias P-value (require -g/-u) </span><br><span class="line"> -u 	Similar to -g except that the output is uncompressed BCF, which is preferred for piping. </span><br><span class="line"></span><br><span class="line">Options for Genotype Likelihood Computation (for -g or -u):</span><br><span class="line"> -e INT 	Phred-scaled gap extension sequencing error probability. Reducing INT leads to longer indels. [20] </span><br><span class="line"> -h INT 	Coefficient for modeling homopolymer errors. Given an l-long homopolymer run, the sequencing error of an indel of size s is modeled as INT*s/l. [100] </span><br><span class="line"> -I 	Do not perform INDEL calling </span><br><span class="line"> -L INT 	Skip INDEL calling if the average per-sample depth is above INT. [250] </span><br><span class="line"> -o INT 	Phred-scaled gap open sequencing error probability. Reducing INT leads to more indel calls. [40] </span><br><span class="line"> -P STR 	Comma dilimited list of platforms (determined by @RG-PL) from which indel candidates are obtained. It is recommended to collect indel candidates from sequencing technologies that have low indel error rate such as ILLUMINA. [all]</span><br></pre></td></tr></table></figure></p>
<h2 id="使用bcftools"><a href="#使用bcftools" class="headerlink" title="使用bcftools"></a>使用bcftools</h2><p>bcftools和samtools类似，用于处理vcf(variant call format)文件和bcf(binary call format)文件。前者为文本文件，后者为其二进制文件。</p>
<p>bcftools使用简单，最主要的命令是view命令，其次还有index和cat等命令。index和cat命令和samtools中类似。此处主讲使用view命令来进行SNP和Indel calling。该命令的使用方法和例子为：<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">$ bcftools view [-AbFGNQSucgv] [-D seqDict] [-l listLoci] [-s listSample] </span><br><span class="line">          [-i gapSNPratio] [-t mutRate] [-p varThres] [-P prior] </span><br><span class="line">          [-1 nGroup1] [-d minFrac] [-U nPerm] [-X permThres] </span><br><span class="line">          [-T trioType] in.bcf [region]</span><br><span class="line"></span><br><span class="line">$ bcftools view -cvNg abc.bcf &gt; snp_indel.vcf</span><br></pre></td></tr></table></figure></p>
<p>生成的结果文件为vcf格式，有10列，分别是：</p>
<blockquote>
<p>1 参考序列名；<br>2 varianti所在的left-most位置；<br>3 variant的ID（默认未设置，用’.’表示)；<br>4 参考序列的allele；<br>5 variant的allele(有多个alleles，则用’,’分隔);<br>6 variant/reference QUALity;<br>7 FILTers applied;<br>8 variant的信息，使用分号隔开；<br>9 FORMAT of the genotype fields, separated by colon (optional)；<br>10 SAMPLE genotypes and per-sample information (optional)。</p>
</blockquote>
<p>例如：<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">scaffold_1      2847    .       A       AACGGTGAAG      194     .       INDEL;DP=11;VDB=0.0401;AF1=1;AC1=2;DP4=0,0,8,3;MQ=35;FQ=-67.5   GT:PL:GQ        1/1:235,33,0:63</span><br><span class="line">scaffold_1      3908    .       G       A       111     .       DP=13;VDB=0.0085;AF1=1;AC1=2;DP4=0,0,5,7;MQ=42;FQ=-63   GT:PL:GQ        1/1:144,36,0:69</span><br><span class="line">scaffold_1      4500    .       A       G       31.5    .       DP=8;VDB=0.0034;AF1=1;AC1=2;DP4=0,0,1,3;MQ=42;FQ=-39    GT:PL:GQ        1/1:64,12,0:21</span><br><span class="line">scaffold_1      4581    .       TGGNGG  TGG     145     .       INDEL;DP=8;VDB=0.0308;AF1=1;AC1=2;DP4=0,0,0,8;MQ=42;FQ=-58.5    GT:PL:GQ        1/1:186,24,0:45</span><br><span class="line">scaffold_1      4644    .       G       A       195     .       DP=21;VDB=0.0198;AF1=1;AC1=2;DP4=0,0,10,10;MQ=42;FQ=-87 GT:PL:GQ        1/1:228,60,0:99</span><br><span class="line">scaffold_1      4827    .       NACAAAGA        NA      4.42    .       INDEL;DP=1;AF1=1;AC1=2;DP4=0,0,1,0;MQ=40;FQ=-37.5       GT:PL:GQ        0/1:40,3,0:3</span><br><span class="line">scaffold_1      4854    .       A       G       48      .       DP=6;VDB=0.0085;AF1=1;AC1=2;DP4=0,0,2,1;MQ=41;FQ=-36    GT:PL:GQ        1/1:80,9,0:16</span><br><span class="line">scaffold_1      5120    .       A       G       85      .       DP=8;VDB=0.0355;AF1=1;AC1=2;DP4=0,0,5,3;MQ=42;FQ=-51    GT:PL:GQ        1/1:118,24,0:45</span><br></pre></td></tr></table></figure></p>
<p>第8列中显示了对variants的信息描述，比较重要，其中的 Tag 的描述如下：<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">Tag	Format	Description</span><br><span class="line">AF1	double	Max-likelihood estimate of the site allele frequency (AF) of the first ALT allele</span><br><span class="line">DP	int	Raw read depth (without quality filtering)</span><br><span class="line">DP4	int[4]	  high-quality reference forward bases, ref reverse, alternate for and alt rev bases</span><br><span class="line">FQ	int	Consensus quality. Positive: sample genotypes different; negative: otherwise</span><br><span class="line">MQ	int	Root-Mean-Square mapping quality of covering reads</span><br><span class="line">PC2	int[2]	Phred probability of AF in group1 samples being larger (,smaller) than in group2</span><br><span class="line">PCHI2	double	Posterior weighted chi^2 P-value between group1 and group2 samples</span><br><span class="line">PV4	double[4]	P-value for strand bias, baseQ bias, mapQ bias and tail distance bias</span><br><span class="line">QCHI2	int	Phred-scaled PCHI2</span><br><span class="line">RP	int	  permutations yielding a smaller PCHI2</span><br><span class="line">CLR	int	Phred log ratio of genotype likelihoods with and without the trio/pair constraint</span><br><span class="line">UGT	string	Most probable genotype configuration without the trio constraint</span><br><span class="line">CGT	string	Most probable configuration with the trio constraint</span><br></pre></td></tr></table></figure></p>
<p>bcftools view 的具体参数如下：<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><span class="line">Input/Output Options:</span><br><span class="line">-A 	Retain all possible alternate alleles at variant sites. By default, the view command discards unlikely alleles.</span><br><span class="line">-b 	Output in the BCF format. The default is VCF.</span><br><span class="line">-D FILE Sequence dictionary (list of chromosome names) for VCF-&gt;BCF conversion [null]</span><br><span class="line">-F 	Indicate PL is generated by r921 or before (ordering is different).</span><br><span class="line">-G 	Suppress all individual genotype information.</span><br><span class="line">-l FILE List of sites at which information are outputted [all sites]</span><br><span class="line">-N 	Skip sites where the REF field is not A/C/G/T</span><br><span class="line">-Q 	Output the QCALL likelihood format</span><br><span class="line">-s FILE List of samples to use. The first column in the input gives the sample names and the second gives the ploidy, which can only be 1 or 2. When the 2nd column is absent, the sample ploidy is assumed to be 2. In the output, the ordering of samples will be identical to the one in FILE. [null]</span><br><span class="line">-S 	The input is VCF instead of BCF.</span><br><span class="line">-u 	Uncompressed BCF output (force -b). </span><br><span class="line"></span><br><span class="line">Consensus/Variant Calling Options:</span><br><span class="line">-c 	Call variants using Bayesian inference. This option automatically invokes option -e.</span><br><span class="line">-d FLOAT When -v is in use, skip loci where the fraction of samples covered by reads is below FLOAT. [0]</span><br><span class="line">        当有多个sample用于variants calling时，比如多个转录组数据或多个重测序</span><br><span class="line">        数据需要比对到参考基因组上，设置该值，表明至少有该&lt;float 0~1&gt;比例的</span><br><span class="line">        samples在该位点都有覆盖才计算入variant.所以对于只有一个sample的情况</span><br><span class="line">        下，该值设置在0～1之间没有意义，大于1则得不到任何结果。</span><br><span class="line">-e 	Perform max-likelihood inference only, including estimating the site allele frequency, testing Hardy-Weinberg equlibrium and testing associations with LRT.</span><br><span class="line">-g 	Call per-sample genotypes at variant sites (force -c)</span><br><span class="line">-i FLOAT Ratio of INDEL-to-SNP mutation rate [0.15]</span><br><span class="line">-p FLOAT A site is considered to be a variant if P(ref|D)</span><br><span class="line">-t FLOAT Scaled muttion rate for variant calling [0.001]</span><br><span class="line">-T STR 	Enable pair/trio calling. For trio calling, option -s is usually needed to be applied to configure the trio members and their ordering. In the file supplied to the option -s, the first sample must be the child, the second the father and the third the mother. The valid values of STR are ‘pair’, ‘trioauto’, ‘trioxd’ and ‘trioxs’, where ‘pair’ calls differences between two input samples, and ‘trioxd’ (‘trioxs’) specifies that the input is from the X chromosome non-PAR regions and the child is a female (male). [null]</span><br><span class="line">-v 	Output variant sites only (force -c) </span><br><span class="line"></span><br><span class="line">Contrast Calling and Association Test Options:</span><br><span class="line">-1 INT 	Number of group-1 samples. This option is used for dividing the samples into two groups for contrast SNP calling or association test. When this option is in use, the following VCF INFO will be outputted: PC2, PCHI2 and QCHI2. [0]</span><br><span class="line">-U INT 	Number of permutations for association test (effective only with -1) [0]</span><br><span class="line">-X FLOAT Only perform permutations for P(chi^2)</span><br></pre></td></tr></table></figure></p>
<p>使用bcftools得到variant calling结果后。需要对结果再次进行过滤。主要依据比对结果中第8列信息。其中的 DP4 一行尤为重要，提供了4个数据：1 比对结果和正链一致的reads数、2 比对结果和负链一致的reads数、3 比对结果在正链的variant上的reads数、4 比对结果在负链的variant上的reads数。可以设定 （value3 + value4）大于某一阈值，才算是variant。比如：<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ perl -ne &apos;print $_ if /DP4=(\d+),(\d+),(\d+),(\d+)/ &amp;&amp; ($3+$4)&gt;=10 &amp;&amp; ($3+$4)/($1+$2+$3+$4)&gt;=0.8&apos; snp_indel.vcf &gt; snp_indel.final.vcf</span><br></pre></td></tr></table></figure></p>
<h2 id="rmdup"><a href="#rmdup" class="headerlink" title="rmdup"></a>rmdup</h2><p>NGS上机测序前需要进行PCR一步，使一个模板扩增出一簇，从而在上机测序的时候表现出为1个点，即一个reads。若一个模板扩增出了多簇，结果得到了多个reads，这些reads的坐标(coordinates)是相近的。在进行了reads比对后需要将这些由PCR duplicates获得的reads去掉，并只保留最高比对质量的read。使用rmdup命令即可完成.<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">Usage:  samtools rmdup [-sS]  </span><br><span class="line"> -s 对single-end reads。默认情况下，只对paired-end reads</span><br><span class="line"> -S 将Paired-end reads作为single-end reads处理。</span><br><span class="line"></span><br><span class="line">$ samtools input.sorted.bam output.bam</span><br></pre></td></tr></table></figure></p>

    </div>
    
    <div class="post-footer">
        <div>
            
                转载声明：商业转载请联系作者获得授权,非商业转载请注明出处 © <a href="" target="_blank">Snippet</a>
            
        </div>
        <div>
            
        </div>
    </div>
</article>

<div class="article-nav prev-next-wrap clearfix">
    
        <a href="/2018/04/29/linux-xia-duo-ge-jie-dian-ssh-hu-xiang-mian-mi-ma-deng-lu/" class="pre-post btn btn-default" title='linux下多个节点ssh互相免密码登陆'>
            <i class="fa fa-angle-left fa-fw"></i><span class="hidden-lg">上一篇</span>
            <span class="hidden-xs">linux下多个节点ssh互相免密码登陆</span>
        </a>
    
    
        <a href="/2018/04/01/c-yu-yan-du-xie-wen-jian/" class="next-post btn btn-default" title='C语言读写文件'>
            <span class="hidden-lg">下一篇</span>
            <span class="hidden-xs">C语言读写文件</span><i class="fa fa-angle-right fa-fw"></i>
        </a>
    
</div>


    <div id="comments">
        
   <p>评论系统未开启，无法评论！</p>

    </div>





                </main>
                
                    <aside id="article-toc" role="navigation" class="col-md-4">
    <div class="widget">
        <h3 class="title">文章目录</h3>
        
            <ol class="toc"><li class="toc-item toc-level-2"><a class="toc-link" href="#view"><span class="toc-text">view</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#sort"><span class="toc-text">sort</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#merge"><span class="toc-text">merge</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#index"><span class="toc-text">index</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#faidx"><span class="toc-text">faidx</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#tview"><span class="toc-text">tview</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#flagstat"><span class="toc-text">flagstat</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#depth"><span class="toc-text">depth</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#其它有用的命令"><span class="toc-text">其它有用的命令</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#将bam文件转换为fastq文件"><span class="toc-text">将bam文件转换为fastq文件</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#mpileup"><span class="toc-text">mpileup</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#使用bcftools"><span class="toc-text">使用bcftools</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#rmdup"><span class="toc-text">rmdup</span></a></li></ol>
        
    </div>
</aside>

                
            </div>
        </div>
    </section>
    <footer class="main-footer">
    <div class="container">
        <div class="row">
        </div>
    </div>
</footer>

<a id="back-to-top" class="icon-btn hide">
	<i class="fa fa-chevron-up"></i>
</a>




    <div class="copyright">
    <div class="container">
        <div class="row">
            <div class="col-sm-12">
                <div class="busuanzi">
    
</div>

            </div>
            <div class="col-sm-12">
                <span>Copyright &copy; 2017
                </span> |
                <span>
                    Powered by <a href="//hexo.io" class="copyright-links" target="_blank" rel="nofollow">Hexo</a>
                </span> |
                <span>
                    Theme by <a href="//github.com/shenliyang/hexo-theme-snippet.git" class="copyright-links" target="_blank" rel="nofollow">Snippet</a>
                </span>
            </div>
        </div>
    </div>
</div>







<script src="/js/app.js?rev=@@hash"></script>

<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        tex2jax: {
            inlineMath: [ ["$","$"], ["\\(","\\)"] ],
            skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code'],
            processEscapes: true
        }
    });
    MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax();
        for (var i = 0; i < all.length; ++i)
            all[i].SourceElement().parentNode.className += ' has-jax';
    });
</script>
<script src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
</body>
</html>