<!DOCTYPE html>



  


<html class="theme-next mist use-motion" lang="zh-Hans">
<head>
  <meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>
<meta name="theme-color" content="#222">









<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />















  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />




  
  
  
  

  
    
    
  

  

  

  

  

  
    
    
    <link href="//fonts.googleapis.com/css?family=Lato:300,300italic,400,400italic,700,700italic&subset=latin,latin-ext" rel="stylesheet" type="text/css">
  






<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.2" rel="stylesheet" type="text/css" />


  <meta name="keywords" content="Hexo, NexT" />








  <link rel="shortcut icon" type="image/x-icon" href="/favicon.ico?v=5.1.2" />






<meta property="og:type" content="website">
<meta property="og:title" content="木叶村">
<meta property="og:url" content="http://yoursite.com/page/4/index.html">
<meta property="og:site_name" content="木叶村">
<meta property="og:locale" content="zh-Hans">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="木叶村">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Mist',
    sidebar: {"position":"left","display":"post","offset":12,"offset_float":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: true,
    duoshuo: {
      userId: '0',
      author: '博主'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://yoursite.com/page/4/"/>





  <title>木叶村</title>
  














</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-Hans">

  
  
    
  

  <div class="container sidebar-position-left 
   page-home 
 ">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">木叶村</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle"></p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            首页
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            分类
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            归档
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            标签
          </a>
        </li>
      

      
    </ul>
  

  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            
  <section id="posts" class="posts-expand">
    
      

  

  
  
  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2018/04/29/samtools常用命令详解/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Naruto">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/137356299.jpg">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="木叶村">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/04/29/samtools常用命令详解/" itemprop="url">(转载)samtools常用命令详解</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-04-29T00:00:00+08:00">
                2018-04-29
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生信/" itemprop="url" rel="index">
                    <span itemprop="name">生信</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <p>samtools的说明文档：<a href="http://samtools.sourceforge.net/samtools.shtml" target="_blank" rel="noopener">http://samtools.sourceforge.net/samtools.shtml</a><br>samtools是一个用于操作sam和bam文件的工具合集。包含有许多命令。以下是常用命令的介绍</p>
<h2 id="view"><a href="#view" class="headerlink" title="view"></a>view</h2><p>view命令的主要功能是：将sam文件转换成bam文件；然后对bam文件进行各种操作，比如数据的排序(不属于本命令的功能)和提取(这些操作是对bam文件进行的，因而当输入为sam文件的时候，不能进行该操作)；最后将排序或提取得到的数据输出为bam或sam（默认的）格式。</p>
<p>bam文件优点：bam文件为二进制文件，占用的磁盘空间比sam文本文件小；利用bam二进制文件的运算速度快。</p>
<p>view命令中，对sam文件头部的输入(-t或-T）和输出(-h)是单独的一些参数来控制的。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br></pre></td><td class="code"><pre><span class="line">Usage: samtools view [options] &lt;in.bam&gt;|&lt;in.sam&gt; [region1 [...]]</span><br><span class="line"> 默认情况下不加 region，则是输出所有的 region.</span><br><span class="line"></span><br><span class="line">Options: -b       output BAM</span><br><span class="line">                  默认下输出是 SAM 格式文件，该参数设置输出 BAM 格式</span><br><span class="line">         -h       print header for the SAM output</span><br><span class="line">                  默认下输出的 sam 格式文件不带 header，该参数设定输出sam文件时带 header 信息</span><br><span class="line">         -H       print header only (no alignments)</span><br><span class="line">         -S       input is SAM</span><br><span class="line">                  默认下输入是 BAM 文件，若是输入是 SAM 文件，则最好加该参数，否则有时候会报错。</span><br><span class="line">         -u       uncompressed BAM output (force -b)</span><br><span class="line">                  该参数的使用需要有-b参数，能节约时间，但是需要更多磁盘空间。</span><br><span class="line">         -c       Instead of printing the alignments, only count them and print the </span><br><span class="line">                  total number. All filter options, such as ‘-f’, ‘-F’ and ‘-q’ , </span><br><span class="line">                  are taken into account.</span><br><span class="line">         -1       fast compression (force -b)</span><br><span class="line">         -x       output FLAG in HEX (samtools-C specific)</span><br><span class="line">         -X       output FLAG in string (samtools-C specific)</span><br><span class="line">         -c       print only the count of matching records</span><br><span class="line">         -L FILE  output alignments overlapping the input BED FILE [null]</span><br><span class="line">         -t FILE  list of reference names and lengths (force -S) [null]</span><br><span class="line">                  使用一个list文件来作为header的输入</span><br><span class="line">         -T FILE  reference sequence file (force -S) [null]</span><br><span class="line">                  使用序列fasta文件作为header的输入</span><br><span class="line">         -o FILE  output file name [stdout]</span><br><span class="line">         -R FILE  list of read groups to be outputted [null]</span><br><span class="line">         -f INT   required flag, 0 for unset [0]</span><br><span class="line">         -F INT   filtering flag, 0 for unset [0] </span><br><span class="line">                  Skip alignments with bits present in INT [0]</span><br><span class="line">                  数字4代表该序列没有比对到参考序列上</span><br><span class="line">                  数字8代表该序列的mate序列没有比对到参考序列上</span><br><span class="line">         -q INT   minimum mapping quality [0]</span><br><span class="line">         -l STR   only output reads in library STR [null]</span><br><span class="line">         -r STR   only output reads in read group STR [null]</span><br><span class="line">         -s FLOAT fraction of templates to subsample; integer part as seed [-1]</span><br><span class="line">         -?       longer help</span><br></pre></td></tr></table></figure>
<p>例子：<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">将sam文件转换成bam文件</span><br><span class="line">$ samtools view -bS abc.sam &gt; abc.bam</span><br><span class="line">$ samtools view -b -S abc.sam -o abc.bam</span><br><span class="line"></span><br><span class="line">提取比对到参考序列上的比对结果</span><br><span class="line">$ samtools view -bF 4 abc.bam &gt; abc.F.bam</span><br><span class="line"></span><br><span class="line">提取paired reads中两条reads都比对到参考序列上的比对结果，只需要把两个4+8的值12作为过滤参数即可</span><br><span class="line">$ samtools view -bF 12 abc.bam &gt; abc.F12.bam</span><br><span class="line"></span><br><span class="line">提取没有比对到参考序列上的比对结果</span><br><span class="line">$ samtools view -bf 4 abc.bam &gt; abc.f.bam</span><br><span class="line"></span><br><span class="line">提取bam文件中比对到caffold1上的比对结果，并保存到sam文件格式</span><br><span class="line">$ samtools view abc.bam scaffold1 &gt; scaffold1.sam</span><br><span class="line"></span><br><span class="line">提取scaffold1上能比对到30k到100k区域的比对结果</span><br><span class="line">$ samtools view abc.bam scaffold1:30000-100000 $gt; scaffold1_30k-100k.sam</span><br><span class="line"></span><br><span class="line">根据fasta文件，将 header 加入到 sam 或 bam 文件中</span><br><span class="line">$ samtools view -T genome.fasta -h scaffold1.sam &gt; scaffold1.h.sam</span><br></pre></td></tr></table></figure></p>
<h2 id="sort"><a href="#sort" class="headerlink" title="sort"></a>sort</h2><p>sort对bam文件进行排序。<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">Usage: samtools sort [-n] [-m &lt;maxMem&gt;] &lt;in.bam&gt; &lt;out.prefix&gt;  </span><br><span class="line">-m 参数默认下是 500,000,000 即500M（不支持K，M，G等缩写）。对于处理大数据时，如果内存够用，则设置大点的值，以节约时间。</span><br><span class="line">-n 设定排序方式按short reads的ID排序。默认下是按序列在fasta文件中的顺序（即header）和序列从左往右的位点排序。</span><br></pre></td></tr></table></figure></p>
<p>例子：<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$ samtools sort abc.bam abc.sort</span><br><span class="line">$ samtools view abc.sort.bam | less -S</span><br></pre></td></tr></table></figure></p>
<h2 id="merge"><a href="#merge" class="headerlink" title="merge"></a>merge</h2><p>将2个或2个以上的已经sort了的bam文件融合成一个bam文件。融合后的文件不需要则是已经sort过了的。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">Usage:   samtools merge [-nr] [-h inh.sam] &lt;out.bam&gt; &lt;in1.bam&gt; &lt;in2.bam&gt;[...]</span><br><span class="line"></span><br><span class="line">Options: -n       sort by read names</span><br><span class="line">         -r       attach RG tag (inferred from file names)</span><br><span class="line">         -u       uncompressed BAM output</span><br><span class="line">         -f       overwrite the output BAM if exist</span><br><span class="line">         -1       compress level 1</span><br><span class="line">         -R STR   merge file in the specified region STR [all]</span><br><span class="line">         -h FILE  copy the header in FILE to &lt;out.bam&gt; [in1.bam]</span><br><span class="line"></span><br><span class="line">Note: Samtools&apos; merge does not reconstruct the @RG dictionary in the header. Users</span><br><span class="line">      must provide the correct header with -h, or uses Picard which properly maintains</span><br><span class="line">      the header dictionary in merging.</span><br></pre></td></tr></table></figure>
<h2 id="index"><a href="#index" class="headerlink" title="index"></a>index</h2><p>必须对bam文件进行默认情况下的排序后，才能进行index。否则会报错。</p>
<p>建立索引后将产生后缀为.bai的文件，用于快速的随机处理。很多情况下需要有bai文件的存在，特别是显示序列比对情况下。比如samtool的tview命令就需要；gbrowse2显示reads的比对图形的时候也需要。<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">Usage: samtools index &lt;in.bam&gt; [out.index]</span><br></pre></td></tr></table></figure></p>
<p>例子：<br>以下两种命令结果一样<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$ samtools index abc.sort.bam</span><br><span class="line">$ samtools index abc.sort.bam abc.sort.bam.bai</span><br></pre></td></tr></table></figure></p>
<h2 id="faidx"><a href="#faidx" class="headerlink" title="faidx"></a>faidx</h2><p>对fasta文件建立索引,生成的索引文件以.fai后缀结尾。该命令也能依据索引文件快速提取fasta文件中的某一条（子）序列</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">Usage: samtools faidx &lt;in.bam&gt; [ [...]]</span><br><span class="line"></span><br><span class="line">对基因组文件建立索引</span><br><span class="line">$ samtools faidx genome.fasta</span><br><span class="line">生成了索引文件genome.fasta.fai,是一个文本文件，分成了5列。第一列是子序列的名称；</span><br><span class="line">第二列是子序列的长度；个人认为“第三列是序列所在的位置”，因为该数字从上往下逐渐变大，</span><br><span class="line">最后的数字是genome.fasta文件的大小；第4和5列不知是啥意思。于是通过此文件，可以定位子序列在fasta文件在磁盘上的存放位置，直接快速调出子序列。</span><br><span class="line"></span><br><span class="line">由于有索引文件，可以使用以下命令很快从基因组中提取到fasta格式的子序列</span><br><span class="line">$ samtools faidx genome.fasta scffold_10 &gt; scaffold_10.fasta</span><br></pre></td></tr></table></figure>
<h2 id="tview"><a href="#tview" class="headerlink" title="tview"></a>tview</h2><p>tview能直观的显示出reads比对基因组的情况，和基因组浏览器有点类似。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">Usage: samtools tview &lt;aln.bam&gt; [ref.fasta]</span><br><span class="line"></span><br><span class="line">当给出参考基因组的时候，会在第一排显示参考基因组的序列，否则，第一排全用N表示。</span><br><span class="line">按下 g ，则提示输入要到达基因组的某一个位点。例子“scaffold_10:1000&quot;表示到达第10号scaffold的第1000个碱基位点处。</span><br><span class="line">使用H(左）J（上）K（下）L（右）移动显示界面。大写字母移动快，小写字母移动慢。</span><br><span class="line">使用空格建向左快速移动（和 L 类似），使用Backspace键向左快速移动（和 H 类似）。</span><br><span class="line">Ctrl+H 向左移动1kb碱基距离； Ctrl+L 向右移动1kb碱基距离</span><br><span class="line">可以用颜色标注比对质量，碱基质量，核苷酸等。30～40的碱基质量或比对质量使用白色表示；</span><br><span class="line">20～30黄色；10～20绿色；0～10蓝色。</span><br><span class="line">使用点号&apos;.&apos;切换显示碱基和点号；使用r切换显示read name等</span><br><span class="line">还有很多其它的使用说明，具体按 ？ 键来查看。</span><br></pre></td></tr></table></figure>
<h2 id="flagstat"><a href="#flagstat" class="headerlink" title="flagstat"></a>flagstat</h2><p>给出BAM文件的比对结果<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line">Usage: samtools flagstat &lt;in.bam&gt;</span><br><span class="line"></span><br><span class="line">$ samtools flagstat example.bam</span><br><span class="line">11945742 + 0 in total (QC-passed reads + QC-failed reads)</span><br><span class="line">总共的reads数</span><br><span class="line">0 + 0 duplicates</span><br><span class="line">7536364 + 0 mapped (63.09%:-nan%)</span><br><span class="line">总体上reads的匹配率</span><br><span class="line">11945742 + 0 paired in sequencing</span><br><span class="line">有多少reads是属于paired reads</span><br><span class="line">5972871 + 0 read1</span><br><span class="line">reads1中的reads数</span><br><span class="line">5972871 + 0 read2</span><br><span class="line">reads2中的reads数</span><br><span class="line">6412042 + 0 properly paired (53.68%:-nan%)</span><br><span class="line">完美匹配的reads数：比对到同一条参考序列，并且两条reads之间的距离符合设置的阈值</span><br><span class="line">6899708 + 0 with itself and mate mapped</span><br><span class="line">paired reads中两条都比对到参考序列上的reads数</span><br><span class="line">636656 + 0 singletons (5.33%:-nan%)</span><br><span class="line">单独一条匹配到参考序列上的reads数，和上一个相加，则是总的匹配上的reads数。</span><br><span class="line">469868 + 0 with mate mapped to a different chr</span><br><span class="line">paired reads中两条分别比对到两条不同的参考序列的reads数</span><br><span class="line">243047 + 0 with mate mapped to a different chr (mapQ&gt;=5)</span><br><span class="line">同上一个，只是其中比对质量&gt;=5的reads的数量</span><br></pre></td></tr></table></figure></p>
<h2 id="depth"><a href="#depth" class="headerlink" title="depth"></a>depth</h2><p>得到每个碱基位点的测序深度,并输出到标准输出。<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">Usage: bam2depth [-r reg] [-q baseQthres] [-Q mapQthres] [-b in.bed] &lt;in1.bam&gt; [...]</span><br></pre></td></tr></table></figure></p>
<h2 id="其它有用的命令"><a href="#其它有用的命令" class="headerlink" title="其它有用的命令"></a>其它有用的命令</h2><p>reheader 替换bam文件的头<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ samtools reheader &lt;in.header.sam&gt; &lt;in.bam&gt;</span><br></pre></td></tr></table></figure></p>
<p>cat 连接多个bam文件，适用于非sorted的bam文件<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ samtools cat [-h header.sam] [-o out.bam] &lt;in1.bam&gt; &lt;in2.bam&gt; [ ... ]</span><br></pre></td></tr></table></figure></p>
<p>idxstats 统计一个表格，4列，分别为”序列名，序列长度，比对上的reads数，unmapped reads number”。第4列应该是paired reads中有一端能匹配到该scaffold上，而另外一端不匹配到任何scaffolds上的reads数。<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ samtools idxstats &lt;aln.bam&gt;</span><br></pre></td></tr></table></figure></p>
<h2 id="将bam文件转换为fastq文件"><a href="#将bam文件转换为fastq文件" class="headerlink" title="将bam文件转换为fastq文件"></a>将bam文件转换为fastq文件</h2><p>有时候，我们需要提取出比对到一段参考序列的reads，进行小范围的分析，以利于debug等。这时需要将bam或sam文件转换为fastq格式。<br>该网站提供了一个bam转换为fastq的程序：<a href="http://www.hudsonalpha.org/gsl/information/software/bam2fastq" target="_blank" rel="noopener">http://www.hudsonalpha.org/gsl/information/software/bam2fastq</a></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">$ wget http://www.hudsonalpha.org/gsl/static/software/bam2fastq-1.1.0.tgz</span><br><span class="line">$ tar zxf bam2fastq-1.1.0.tgz</span><br><span class="line">$ cd bam2fastq-1.1.0</span><br><span class="line">$ make</span><br><span class="line">$ ./bam2fastq &lt;in.bam&gt;</span><br></pre></td></tr></table></figure>
<h2 id="mpileup"><a href="#mpileup" class="headerlink" title="mpileup"></a>mpileup</h2><p>samtools还有个非常重要的命令mpileup，以前为pileup。该命令用于生成bcf文件，再使用bcftools进行SNP和Indel的分析。bcftools是samtool中附带的软件，在samtools的安装文件夹中可以找到。</p>
<p>最常用的参数有2： <code>-f</code> 来输入有索引文件的fasta参考序列； <code>-g</code> 输出到bcf格式。用法和最简单的例子如下</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">Usage: samtools mpileup [-EBug] [-C capQcoef] [-r reg] [-f in.fa] [-l list] [-M capMapQ] [-Q minBaseQ] [-q minMapQ] in.bam [in2.bam [...]]</span><br><span class="line"></span><br><span class="line">$ samtools mpileup -f genome.fasta abc.bam &gt; abc.txt</span><br><span class="line">$ samtools mpileup -gSDf genome.fasta abc.bam &gt; abc.bcf</span><br><span class="line">$ samtools mpileup -guSDf genome.fasta abc.bam | \</span><br><span class="line">           bcftools view -cvNg - &gt; abc.vcf</span><br></pre></td></tr></table></figure>
<p>mpileup不使用<code>-u</code>或<code>-g</code>参数时，则不生成二进制的bcf文件，而生成一个文本文件(输出到标准输出)。该文本文件统计了参考序列中每个碱基位点的比对情况；该文件每一行代表了参考序列中某一个碱基位点的比对结果。比如：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">scaffold_1      2841    A       11      ,,,...,....     BHIGDGIJ?FF</span><br><span class="line">scaffold_1      2842    C       12      ,$,,...,....^I. CFGEGEGGCFF+</span><br><span class="line">scaffold_1      2843    G       11      ,,...,.....     FDDDDCD?DD+</span><br><span class="line">scaffold_1      2844    G       11      ,,...,.....     FA?AAAA&lt;AA+</span><br><span class="line">scaffold_1      2845    G       11      ,,...,.....     F656666166*</span><br><span class="line">scaffold_1      2846    A       11      ,,...,.....     (1.1111)11*</span><br><span class="line">scaffold_1      2847    A       11      ,,+9acggtgaag.+9ACGGTGAAT.+9ACGGTGAAG.+9ACGGTGAAG,+9acggtgaag.+9ACGGTGAAG.+9ACGGTGAAG.+9ACGGTGAAG.+9ACGGTGAAG.+9ACGGTGAAG       %.+....-..)</span><br><span class="line">scaffold_1      2848    N       11      agGGGgGGGGG     !!$!!!!!!!!</span><br><span class="line">scaffold_1      2849    A       11      c$,...,.....    !0000000000</span><br><span class="line">scaffold_1      2850    A       10      ,...,.....      353333333</span><br></pre></td></tr></table></figure>
<p>mpileup生成的结果包含6行：参考序列名；位置；参考碱基；比对上的reads数；比对情况；比对上的碱基的质量。其中第5列比较复杂,解释如下：<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">1  ‘.’代表与参考序列正链匹配。</span><br><span class="line">2  ‘,’代表与参考序列负链匹配。</span><br><span class="line">3  ‘ATCGN’代表在正链上的不匹配。</span><br><span class="line">4  ‘atcgn’代表在负链上的不匹配。</span><br><span class="line">5  ‘*’代表模糊碱基</span><br><span class="line">6  ‘^’代表匹配的碱基是一个read的开始；’^’后面紧跟的ascii码减去33代表比对质量；这两个符号修饰的是后面的碱基，其后紧跟的碱基(.,ATCGatcgNn)代表该read的第一个碱基。</span><br><span class="line">7  ‘$’代表一个read的结束，该符号修饰的是其前面的碱基。</span><br><span class="line">8  正则式’\+[0-9]+[ACGTNacgtn]+’代表在该位点后插入的碱基；比如上例中在scaffold_1的2847后插入了9个长度的碱基acggtgaag。表明此处极可能是indel。</span><br><span class="line">9  正则式’-[0-9]+[ACGTNacgtn]+’代表在该位点后缺失的碱基；</span><br></pre></td></tr></table></figure></p>
<p><strong>pileup具体的参数如下：</strong></p>
<p><strong>输入参数</strong><br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">-6       Assume the quality is in the Illumina 1.3+ encoding. </span><br><span class="line"> -A Do not skip anomalous read pairs in variant calling. </span><br><span class="line">-B       Disable probabilistic realignment for the computation of base alignment quality (BAQ). BAQ is the Phred-scaled probability of a read base being misaligned. Applying this option greatly helps to reduce false SNPs caused by misalignments. </span><br><span class="line">-b FILE  List of input BAM files, one file per line [null]</span><br><span class="line">-C INT   Coefficient for downgrading mapping quality for reads containing excessive mismatches. Given a read with a phred-scaled probability q of being generated from the mapped position, the new mapping quality is about sqrt((INT-q)/INT)*INT. A zero value disables this functionality; if enabled, the recommended value for BWA is 50. [0] </span><br><span class="line">-d INT   At a position, read maximally INT reads per input BAM. [250] </span><br><span class="line">-E       Extended BAQ computation. This option helps sensitivity especially for MNPs, but may hurt specificity a little bit. </span><br><span class="line">-f FILE  The faidx-indexed reference file in the FASTA format. The file can be optionally compressed by razip. [null] </span><br><span class="line">-l FILE  BED or position list file containing a list of regions or sites where pileup or BCF should be generated [null] </span><br><span class="line">-M INT       cap mapping quality at INT [60]</span><br><span class="line">-q INT 	Minimum mapping quality for an alignment to be used [0] </span><br><span class="line">-Q INT 	Minimum base quality for a base to be considered [13]</span><br><span class="line">-r STR 	Only generate pileup in region STR [all sites]</span><br></pre></td></tr></table></figure></p>
<p><strong>输出参数</strong><br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"> -D 	Output per-sample read depth (require -g/-u)</span><br><span class="line"> -g 	Compute genotype likelihoods and output them in the binary call format (BCF). </span><br><span class="line"> -S 	Output per-sample Phred-scaled strand bias P-value (require -g/-u) </span><br><span class="line"> -u 	Similar to -g except that the output is uncompressed BCF, which is preferred for piping. </span><br><span class="line"></span><br><span class="line">Options for Genotype Likelihood Computation (for -g or -u):</span><br><span class="line"> -e INT 	Phred-scaled gap extension sequencing error probability. Reducing INT leads to longer indels. [20] </span><br><span class="line"> -h INT 	Coefficient for modeling homopolymer errors. Given an l-long homopolymer run, the sequencing error of an indel of size s is modeled as INT*s/l. [100] </span><br><span class="line"> -I 	Do not perform INDEL calling </span><br><span class="line"> -L INT 	Skip INDEL calling if the average per-sample depth is above INT. [250] </span><br><span class="line"> -o INT 	Phred-scaled gap open sequencing error probability. Reducing INT leads to more indel calls. [40] </span><br><span class="line"> -P STR 	Comma dilimited list of platforms (determined by @RG-PL) from which indel candidates are obtained. It is recommended to collect indel candidates from sequencing technologies that have low indel error rate such as ILLUMINA. [all]</span><br></pre></td></tr></table></figure></p>
<h2 id="使用bcftools"><a href="#使用bcftools" class="headerlink" title="使用bcftools"></a>使用bcftools</h2><p>bcftools和samtools类似，用于处理vcf(variant call format)文件和bcf(binary call format)文件。前者为文本文件，后者为其二进制文件。</p>
<p>bcftools使用简单，最主要的命令是view命令，其次还有index和cat等命令。index和cat命令和samtools中类似。此处主讲使用view命令来进行SNP和Indel calling。该命令的使用方法和例子为：<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">$ bcftools view [-AbFGNQSucgv] [-D seqDict] [-l listLoci] [-s listSample] </span><br><span class="line">          [-i gapSNPratio] [-t mutRate] [-p varThres] [-P prior] </span><br><span class="line">          [-1 nGroup1] [-d minFrac] [-U nPerm] [-X permThres] </span><br><span class="line">          [-T trioType] in.bcf [region]</span><br><span class="line"></span><br><span class="line">$ bcftools view -cvNg abc.bcf &gt; snp_indel.vcf</span><br></pre></td></tr></table></figure></p>
<p>生成的结果文件为vcf格式，有10列，分别是：</p>
<blockquote>
<p>1 参考序列名；<br>2 varianti所在的left-most位置；<br>3 variant的ID（默认未设置，用’.’表示)；<br>4 参考序列的allele；<br>5 variant的allele(有多个alleles，则用’,’分隔);<br>6 variant/reference QUALity;<br>7 FILTers applied;<br>8 variant的信息，使用分号隔开；<br>9 FORMAT of the genotype fields, separated by colon (optional)；<br>10 SAMPLE genotypes and per-sample information (optional)。</p>
</blockquote>
<p>例如：<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">scaffold_1      2847    .       A       AACGGTGAAG      194     .       INDEL;DP=11;VDB=0.0401;AF1=1;AC1=2;DP4=0,0,8,3;MQ=35;FQ=-67.5   GT:PL:GQ        1/1:235,33,0:63</span><br><span class="line">scaffold_1      3908    .       G       A       111     .       DP=13;VDB=0.0085;AF1=1;AC1=2;DP4=0,0,5,7;MQ=42;FQ=-63   GT:PL:GQ        1/1:144,36,0:69</span><br><span class="line">scaffold_1      4500    .       A       G       31.5    .       DP=8;VDB=0.0034;AF1=1;AC1=2;DP4=0,0,1,3;MQ=42;FQ=-39    GT:PL:GQ        1/1:64,12,0:21</span><br><span class="line">scaffold_1      4581    .       TGGNGG  TGG     145     .       INDEL;DP=8;VDB=0.0308;AF1=1;AC1=2;DP4=0,0,0,8;MQ=42;FQ=-58.5    GT:PL:GQ        1/1:186,24,0:45</span><br><span class="line">scaffold_1      4644    .       G       A       195     .       DP=21;VDB=0.0198;AF1=1;AC1=2;DP4=0,0,10,10;MQ=42;FQ=-87 GT:PL:GQ        1/1:228,60,0:99</span><br><span class="line">scaffold_1      4827    .       NACAAAGA        NA      4.42    .       INDEL;DP=1;AF1=1;AC1=2;DP4=0,0,1,0;MQ=40;FQ=-37.5       GT:PL:GQ        0/1:40,3,0:3</span><br><span class="line">scaffold_1      4854    .       A       G       48      .       DP=6;VDB=0.0085;AF1=1;AC1=2;DP4=0,0,2,1;MQ=41;FQ=-36    GT:PL:GQ        1/1:80,9,0:16</span><br><span class="line">scaffold_1      5120    .       A       G       85      .       DP=8;VDB=0.0355;AF1=1;AC1=2;DP4=0,0,5,3;MQ=42;FQ=-51    GT:PL:GQ        1/1:118,24,0:45</span><br></pre></td></tr></table></figure></p>
<p>第8列中显示了对variants的信息描述，比较重要，其中的 Tag 的描述如下：<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">Tag	Format	Description</span><br><span class="line">AF1	double	Max-likelihood estimate of the site allele frequency (AF) of the first ALT allele</span><br><span class="line">DP	int	Raw read depth (without quality filtering)</span><br><span class="line">DP4	int[4]	  high-quality reference forward bases, ref reverse, alternate for and alt rev bases</span><br><span class="line">FQ	int	Consensus quality. Positive: sample genotypes different; negative: otherwise</span><br><span class="line">MQ	int	Root-Mean-Square mapping quality of covering reads</span><br><span class="line">PC2	int[2]	Phred probability of AF in group1 samples being larger (,smaller) than in group2</span><br><span class="line">PCHI2	double	Posterior weighted chi^2 P-value between group1 and group2 samples</span><br><span class="line">PV4	double[4]	P-value for strand bias, baseQ bias, mapQ bias and tail distance bias</span><br><span class="line">QCHI2	int	Phred-scaled PCHI2</span><br><span class="line">RP	int	  permutations yielding a smaller PCHI2</span><br><span class="line">CLR	int	Phred log ratio of genotype likelihoods with and without the trio/pair constraint</span><br><span class="line">UGT	string	Most probable genotype configuration without the trio constraint</span><br><span class="line">CGT	string	Most probable configuration with the trio constraint</span><br></pre></td></tr></table></figure></p>
<p>bcftools view 的具体参数如下：<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><span class="line">Input/Output Options:</span><br><span class="line">-A 	Retain all possible alternate alleles at variant sites. By default, the view command discards unlikely alleles.</span><br><span class="line">-b 	Output in the BCF format. The default is VCF.</span><br><span class="line">-D FILE Sequence dictionary (list of chromosome names) for VCF-&gt;BCF conversion [null]</span><br><span class="line">-F 	Indicate PL is generated by r921 or before (ordering is different).</span><br><span class="line">-G 	Suppress all individual genotype information.</span><br><span class="line">-l FILE List of sites at which information are outputted [all sites]</span><br><span class="line">-N 	Skip sites where the REF field is not A/C/G/T</span><br><span class="line">-Q 	Output the QCALL likelihood format</span><br><span class="line">-s FILE List of samples to use. The first column in the input gives the sample names and the second gives the ploidy, which can only be 1 or 2. When the 2nd column is absent, the sample ploidy is assumed to be 2. In the output, the ordering of samples will be identical to the one in FILE. [null]</span><br><span class="line">-S 	The input is VCF instead of BCF.</span><br><span class="line">-u 	Uncompressed BCF output (force -b). </span><br><span class="line"></span><br><span class="line">Consensus/Variant Calling Options:</span><br><span class="line">-c 	Call variants using Bayesian inference. This option automatically invokes option -e.</span><br><span class="line">-d FLOAT When -v is in use, skip loci where the fraction of samples covered by reads is below FLOAT. [0]</span><br><span class="line">        当有多个sample用于variants calling时，比如多个转录组数据或多个重测序</span><br><span class="line">        数据需要比对到参考基因组上，设置该值，表明至少有该&lt;float 0~1&gt;比例的</span><br><span class="line">        samples在该位点都有覆盖才计算入variant.所以对于只有一个sample的情况</span><br><span class="line">        下，该值设置在0～1之间没有意义，大于1则得不到任何结果。</span><br><span class="line">-e 	Perform max-likelihood inference only, including estimating the site allele frequency, testing Hardy-Weinberg equlibrium and testing associations with LRT.</span><br><span class="line">-g 	Call per-sample genotypes at variant sites (force -c)</span><br><span class="line">-i FLOAT Ratio of INDEL-to-SNP mutation rate [0.15]</span><br><span class="line">-p FLOAT A site is considered to be a variant if P(ref|D)</span><br><span class="line">-t FLOAT Scaled muttion rate for variant calling [0.001]</span><br><span class="line">-T STR 	Enable pair/trio calling. For trio calling, option -s is usually needed to be applied to configure the trio members and their ordering. In the file supplied to the option -s, the first sample must be the child, the second the father and the third the mother. The valid values of STR are ‘pair’, ‘trioauto’, ‘trioxd’ and ‘trioxs’, where ‘pair’ calls differences between two input samples, and ‘trioxd’ (‘trioxs’) specifies that the input is from the X chromosome non-PAR regions and the child is a female (male). [null]</span><br><span class="line">-v 	Output variant sites only (force -c) </span><br><span class="line"></span><br><span class="line">Contrast Calling and Association Test Options:</span><br><span class="line">-1 INT 	Number of group-1 samples. This option is used for dividing the samples into two groups for contrast SNP calling or association test. When this option is in use, the following VCF INFO will be outputted: PC2, PCHI2 and QCHI2. [0]</span><br><span class="line">-U INT 	Number of permutations for association test (effective only with -1) [0]</span><br><span class="line">-X FLOAT Only perform permutations for P(chi^2)</span><br></pre></td></tr></table></figure></p>
<p>使用bcftools得到variant calling结果后。需要对结果再次进行过滤。主要依据比对结果中第8列信息。其中的 DP4 一行尤为重要，提供了4个数据：1 比对结果和正链一致的reads数、2 比对结果和负链一致的reads数、3 比对结果在正链的variant上的reads数、4 比对结果在负链的variant上的reads数。可以设定 （value3 + value4）大于某一阈值，才算是variant。比如：<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ perl -ne &apos;print $_ if /DP4=(\d+),(\d+),(\d+),(\d+)/ &amp;&amp; ($3+$4)&gt;=10 &amp;&amp; ($3+$4)/($1+$2+$3+$4)&gt;=0.8&apos; snp_indel.vcf &gt; snp_indel.final.vcf</span><br></pre></td></tr></table></figure></p>
<h2 id="rmdup"><a href="#rmdup" class="headerlink" title="rmdup"></a>rmdup</h2><p>NGS上机测序前需要进行PCR一步，使一个模板扩增出一簇，从而在上机测序的时候表现出为1个点，即一个reads。若一个模板扩增出了多簇，结果得到了多个reads，这些reads的坐标(coordinates)是相近的。在进行了reads比对后需要将这些由PCR duplicates获得的reads去掉，并只保留最高比对质量的read。使用rmdup命令即可完成.<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">Usage:  samtools rmdup [-sS]  </span><br><span class="line"> -s 对single-end reads。默认情况下，只对paired-end reads</span><br><span class="line"> -S 将Paired-end reads作为single-end reads处理。</span><br><span class="line"></span><br><span class="line">$ samtools input.sorted.bam output.bam</span><br></pre></td></tr></table></figure></p>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2018/04/01/C语言读写文件/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Naruto">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/137356299.jpg">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="木叶村">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/04/01/C语言读写文件/" itemprop="url">C语言读写文件</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-04-01T00:00:00+08:00">
                2018-04-01
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/C-C/" itemprop="url" rel="index">
                    <span itemprop="name">C/C++</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <p><strong>使用标准I/O处理文件</strong><br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br></pre></td><td class="code"><pre><span class="line">#include &lt;stdio.h&gt;</span><br><span class="line">#include &lt;stdlib.h&gt;	// exit()原型</span><br><span class="line"></span><br><span class="line">int main(int argc, char *argv[])</span><br><span class="line">&#123;</span><br><span class="line">	int ch;</span><br><span class="line">	FILE *fp;  // 文件指针</span><br><span class="line">	unsigned long count = 0;</span><br><span class="line">	if (argc != 2)</span><br><span class="line">	&#123;</span><br><span class="line">		printf(&quot;Usage: %s filename\n&quot;,argv[0]);</span><br><span class="line">		exit(EXIT_FAILURE);</span><br><span class="line">	&#125;</span><br><span class="line">	if ((fp = fopen(argv[1],&quot;r&quot;)) == NULL)</span><br><span class="line">	&#123;</span><br><span class="line">		printf(&quot;Can&apos;t open %s\n&quot;,argv[1]);</span><br><span class="line">		exit(EXIT_FAILURE);</span><br><span class="line">	&#125;</span><br><span class="line">	while ((ch = getc(fp)) != EOF)</span><br><span class="line">	&#123;</span><br><span class="line">		putc(ch, stdout); // putchar(ch)相同</span><br><span class="line">		count++;</span><br><span class="line">	&#125;</span><br><span class="line">	fclose(fp);</span><br><span class="line">	printf(&quot;File %s hash %lu characters\n&quot;, argv[1],count);</span><br><span class="line"></span><br><span class="line">	return 0;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br></pre></td><td class="code"><pre><span class="line">//输入/输出升级版，将一个文件附加到另一个文件的末尾</span><br><span class="line">//未测试</span><br><span class="line"></span><br><span class="line">#include &lt;stdio.h&gt;</span><br><span class="line">#include &lt;stdlib&gt;</span><br><span class="line">#include &lt;string.h&gt;</span><br><span class="line">#define BUFSIZE 4096</span><br><span class="line">#define SLEN 81</span><br><span class="line"></span><br><span class="line">void append(FILE *source,FILE *dest);  //函数原型,没有返回值，参数为两个文件指针</span><br><span class="line">char *s_gets(char *st,int n);</span><br><span class="line"></span><br><span class="line">int main(void)</span><br><span class="line">&#123;</span><br><span class="line">	FILE *fa, *fs;</span><br><span class="line">	int file = 0;</span><br><span class="line">	char file_app[SLEN];</span><br><span class="line">	char file_src[SLEN];</span><br><span class="line">	int ch;</span><br><span class="line">	</span><br><span class="line">	puts(&quot;Enter name of destination file:&quot;);</span><br><span class="line">	s_gets(file_app,SLEN);</span><br><span class="line">	if ((fa = fopen(file_app,&quot;a+&quot;) = NULL)</span><br><span class="line">	&#123;</span><br><span class="line">	fprintf(stderr, &quot;Can&apos;t opne %s\n&quot;, file_app);</span><br><span class="line">		exit(EXIT_FAILURE);		//程序结束失败</span><br><span class="line">	&#125;</span><br><span class="line">	if(setvbuf(fa,NULL,_IOFBF,BUFSIZE) !=0) // 完全缓冲。若无法缓冲，就返回0</span><br><span class="line">	&#123;</span><br><span class="line">		fputs(&quot;Can&apos;t creat output buffer\n&quot;,stderr);</span><br><span class="line">		exit(EXIT_FAILURE);</span><br><span class="line">	&#125;</span><br><span class="line">	puts(&quot;Enter name of first source file (empty line to quit):&quot;);</span><br><span class="line">	while(s_gets(file_src,SLEN) &amp;&amp; file_src[0] != &apos;\0&apos;)</span><br><span class="line">	&#123;</span><br><span class="line">		if (strcmp(file_src,file_app)==0)</span><br><span class="line">			fputs(&quot;Can&apos;t append file to itself\n&quot;,stderr);</span><br><span class="line">		else if ((fs = fopen(file_src,&quot;r&quot;)) == NULL)</span><br><span class="line">			fprintf(stderr, &quot;Can&apos;t open %s\n&quot;,file_src);</span><br><span class="line">		else</span><br><span class="line">		&#123;</span><br><span class="line">			 if (setvbuf(fs, NULL,_IOFBF,BUFSIZE) != 0)</span><br><span class="line">			 &#123;</span><br><span class="line">				fputs(&quot;Can&apos;t creat input buffer\n&quot;,stderr);</span><br><span class="line">				continue;</span><br><span class="line">			 &#125;</span><br><span class="line">			 append(fs,fa);</span><br><span class="line">			 if (ferror(fs) != 0)</span><br><span class="line">				 fprintf(stderr,&quot;Error in reading file %s.\n&quot;,file_src);</span><br><span class="line">			 if (ferror(fa) != 0)</span><br><span class="line">				 fprintf(stderr,&quot;Error in writing file %s.\n&quot;,file_app);</span><br><span class="line">			 fclose(fs);</span><br><span class="line">			 files++;</span><br><span class="line">			 puts(&quot;Next file (empty line to quit):&quot;);</span><br><span class="line">		&#125;</span><br><span class="line">	&#125;</span><br><span class="line">	printf(&quot;Done appending. %d files appending.\n&quot;, files );</span><br><span class="line">	rewind(fa);</span><br><span class="line">	printf(&quot;%s continue:\n&quot;, file_app);</span><br><span class="line">	while  ((ch = gets(fa) != EOF)</span><br><span class="line">		putchar(ch);</span><br><span class="line">	puts(&quot;Done displaying.&quot;);</span><br><span class="line">	fclose(fa);</span><br><span class="line">	</span><br><span class="line">	return 0;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">void append(FILE *source, FILE *dest)</span><br><span class="line">&#123;</span><br><span class="line">	size_t bytes;</span><br><span class="line">	static char temp[BUFSIZE];</span><br><span class="line">	</span><br><span class="line">	while((bytes = fread(temp,sizeof(char),BUFSIZE,source)) &gt; 0)	//二进制形式处理数据</span><br><span class="line">		fwrite(temp,sizeof(char),bytes, dest);</span><br><span class="line">	return 0;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">char *s_gets(char *st, int n)</span><br><span class="line">&#123;</span><br><span class="line">	char *ret_val;</span><br><span class="line">	char *find;</span><br><span class="line">	</span><br><span class="line">	ret_val = fgets(st,n, stdin);	// fgets(类型,长度,文件指针)</span><br><span class="line">	if(ret_val)</span><br><span class="line">	&#123;</span><br><span class="line">		find = strchr(st,&apos;\n&apos;);	//查找换行符</span><br><span class="line">		if(find)</span><br><span class="line">		</span><br><span class="line">			*find = &apos;\0&apos;;</span><br><span class="line">		else</span><br><span class="line">			while(getchar() != &apos;\n&apos;)</span><br><span class="line">				continue;</span><br><span class="line">	&#125;</span><br><span class="line">	return ret_val;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2018/03/03/You Are What You Write/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Naruto">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/137356299.jpg">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="木叶村">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/03/03/You Are What You Write/" itemprop="url">沈向洋：You Are What You Write</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-03-03T00:00:00+08:00">
                2018-03-03
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/杂谈/" itemprop="url" rel="index">
                    <span itemprop="name">杂谈</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="You-Are-What-You-Write"><a href="#You-Are-What-You-Write" class="headerlink" title="You Are What You Write"></a>You Are What You Write</h2><p><em>原文：<a href="https://www.linkedin.com/pulse/you-what-write-harry-shum?from=timeline&amp;isappinstalled=0" target="_blank" rel="noopener">https://www.linkedin.com/pulse/you-what-write-harry-shum?from=timeline&amp;isappinstalled=0</a></em></p>
<p>Are Twitter, PowerPoint, Facebook, Instagram and texting eroding our ability to think?</p>
<p>There is a Chinese proverb that says “见文如见人,” which literally means “reading the document is the same as seeing the author.” If we are what we write, then who have we, as a society, become?</p>
<p>I was sitting in a technical review recently, listening to one of our reviewers grill the engineer who was presenting: Why did you choose that design? Why is the service showing bad results? How many users will switch to the solution?</p>
<p>The presenter’s answers lacked depth. It seemed like he hadn’t done enough rigorous thinking, the kind where you sit quietly, sift through research, contemplate options, determine what you know, don’t know and where more work is required. The kind of thinking I did as a young researcher when peers took me and my work apart when I took short cuts. Back then, I practiced a disciplined approach, spending hours just thinking, and even more hours on the hardest part—writing it down.</p>
<p>Today, long-form writing is being replaced. Tweets pass for dialogue. PowerPoint condenses thoughts to bullets. Words have been traded for emojis and GIFs. And we’ve become addicted to the noise. What happens in an Internet minute? 16 million text messages. 1.8 million snaps. 452,000 Tweets. 156 million emails. Who has time to think, let alone write?</p>
<p>And maybe we, in the technology industry, have shaped this reality. We created the phones, apps and 24/7-connected world. We’ve enabled society to put down the pen. The only writing I do today is email or quick WeChat posts.</p>
<p>So now I worry that we’re losing a valuable tool that helps us to think deeply, express who we are at our greatest and expand the intellect of those around us. And for us in the technical community, this is especially troubling. The stakes are higher than ever before with AI. We’re under enormous pressure to ship quickly, to achieve more, faster, but we can’t do this at the expense of the highest engineering quality. We have to think carefully about consequences and alternatives. Who gets blamed when a self-driving car hits someone? The engineer who wrote the code is the driver. Who is accountable for the AI algorithm with bias? The engineer who created the AI.</p>
<p>I see fewer engineers writing and sharing deep thinking, but this is what will lead to far more true innovation across the industry. How will we achieve the big transformative breakthroughs versus the incremental milestones?</p>
<p>By writing. Because the way to think is actually to write.</p>
<p>Putting pen to paper forces you to develop and refine your thinking by iterating, revising and exploring alternatives. Anyone who can think deeply can write beautiful code, inspiring papers or develop the plan to bring the next big thing to life. I encourage you to read Reid Hoffman’s Series B pitch for LinkedIn in which he shares the thinking that helped him succeed. At the time, he shares that a partner in a venture firm was exposed to around 5,000 pitches, looked more closely at 600 to 800, and did between 0 and 2 deals.</p>
<p>Writing offers the possibility to create lasting artifacts. I think of papers I published that endure, albeit perhaps as reference materials. Plenoptic Sampling. Lazy Snapping. Poisson Matting. These are my work’s contribution to the field of computer vision and graphics. They will survive me and, if I’m lucky, even help shape a mind or two.</p>
<p>One of my favorite professors at Carnegie Mellon, Takeo Kanade, said that you have to write research papers like detective novels. You need story, suspense, surprise and ‘aha’ to explain your ideas to peers, to inspire others to contribute and advance your work and the whole field.</p>
<p>Writing is an equalizer to get the best from the whole team. At Amazon, presentations are done with the six-page paper. Meetings kick off with everyone reading followed by comments and questions to the author. Everyone operates from the same context, and introverts, extroverts and non-native speakers have an equal chance to get their thinking across. It’s not about the presenter’s personality, but the words.</p>
<p>Ultimately, writing helps make you successful. You might be the smartest person with the best idea, but if you can’t communicate your thinking in a compelling way, you won’t get far. Two engineers in our AI+R team who inspire me with their regular writing habits are Bill Ramsey and Ronny Kohavi. Bill has written over 250 blog posts at Microsoft, benefitting our entire technical community. With Ronny, you don’t even need to meet him—his highly cited A/B test experimentation papers say it all, and he’s publishing for the benefit of the industry on LinkedIn.</p>
<p>As you’re reading this, you may be logging your objections: I need to drive results, so I need to go straight to code. I’m known for my code, so I don’t need to write papers. I’m not a native speaker, and I speak better with my code. I don’t know what to write about. I don’t have time… But please set them aside—for your own success, for your company’s, for the industry’s advancement—and start writing.</p>
<p>I see so many occasions for building long-form writing back into the engineering culture—planning documents, project proposals, technology LRP’s, review articles—to inspire us to work together, collectively creating and cultivating big ideas and big thinking.</p>
<p>I took a first step recently, writing a research paper with my colleagues Xiaodong He and Di Li, From Eliza to XiaoIce: Challenges and Opportunities with Social Chatbots, for the first time in years, so please no judgment, only constructive feedback!</p>
<p>I challenge everyone reading this piece to write 500 words per week. If you’ve got an idea or you see a problem, write your proposal and share it!</p>
<p>Let’s rewrite our standards for thought leadership and engineering quality by writing more!</p>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2018/02/16/AWK简明教程/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Naruto">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/137356299.jpg">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="木叶村">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/02/16/AWK简明教程/" itemprop="url">AWK简明教程</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-02-16T00:00:00+08:00">
                2018-02-16
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Linux/" itemprop="url" rel="index">
                    <span itemprop="name">Linux</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h5 id="awk命令格式和选项"><a href="#awk命令格式和选项" class="headerlink" title="awk命令格式和选项"></a>awk命令格式和选项</h5><p>语法形式<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">awk [options] &apos;script&apos; var=value file(s)</span><br><span class="line">awk [options] -f scriptfile var=value file(s)</span><br></pre></td></tr></table></figure></p>
<p>常用命令选项</p>
<ul>
<li>-F fs   fs指定输入分隔符，fs可以是字符串或正则表达式，如-F:</li>
<li>-v var=value   赋值一个用户定义变量，将外部变量传递给awk</li>
<li>-f scripfile  从脚本文件中读取awk命令</li>
<li>-m[fr] val   对val值设置内在限制，-mf选项限制分配给val的最大块数目；-mr选项限制记录的最大数目。这两个功能是Bell实验室版awk的扩展功能，在标准awk中不适用。</li>
</ul>
<p><strong>工作原理</strong><br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">hawk &apos;BEGIN&#123; commands &#125; pattern&#123; commands &#125; END&#123; commands &#125;&apos;</span><br></pre></td></tr></table></figure></p>
<h5 id="内建变量"><a href="#内建变量" class="headerlink" title="内建变量"></a>内建变量</h5><div class="table-container">
<table>
<thead>
<tr>
<th>变量</th>
<th>详细说明</th>
</tr>
</thead>
<tbody>
<tr>
<td>$0</td>
<td>当前记录（这个变量中存放着整个行的内容）</td>
</tr>
<tr>
<td>$1~$n</td>
<td>当前记录的第n个字段，字段间由FS分隔</td>
</tr>
<tr>
<td>FS</td>
<td>输入字段分隔符 默认是空格或Tab (<strong>也可以是-F</strong>)</td>
</tr>
<tr>
<td>NF</td>
<td>当前记录中的字段个数，就是有多少列</td>
</tr>
<tr>
<td>NR</td>
<td>已经读出的记录数，就是行号，从1开始，如果有多个文件话，这个值也是不断累加中。</td>
</tr>
<tr>
<td>FNR</td>
<td>当前记录数，与NR不同的是，这个值会是各个文件自己的行号</td>
</tr>
<tr>
<td>RS</td>
<td>输入的记录分隔符， 默认为换行符</td>
</tr>
<tr>
<td>OFS</td>
<td>输出字段分隔符， 默认也是空格</td>
</tr>
<tr>
<td>ORS</td>
<td>输出的记录分隔符，默认为换行符</td>
</tr>
<tr>
<td>FILENAME</td>
<td>当前输入文件的名字</td>
</tr>
</tbody>
</table>
</div>
<h5 id="模式匹配"><a href="#模式匹配" class="headerlink" title="模式匹配"></a>模式匹配</h5><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ awk &apos;$6 ~ /FIN|TIME/ || NR==1 &#123;print NR,$4,$5,$6&#125;&apos; OFS=&quot;\t&quot; netstat.txt</span><br></pre></td></tr></table></figure>
<h5 id="统计"><a href="#统计" class="headerlink" title="统计"></a>统计</h5><p>下面的命令计算所有的C文件，CPP文件和H文件的文件大小总和。<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$ ls -l  *.cpp *.c *.h | awk &apos;&#123;sum+=$5&#125; END &#123;print sum&#125;&apos;</span><br><span class="line">2511401</span><br></pre></td></tr></table></figure></p>
<p>统计每个用户的进程的占了多少内存<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">$ ps aux | awk &apos;NR!=1&#123;a[$1]+=$6;&#125; END &#123; for(i in a) print i &quot;, &quot; a[i]&quot;KB&quot;;&#125;&apos;</span><br><span class="line">dbus, 540KB</span><br><span class="line">mysql, 99928KB</span><br><span class="line">www, 3264924KB</span><br><span class="line">root, 63644KB</span><br><span class="line">hchen, 6020KB</span><br></pre></td></tr></table></figure></p>
<p><strong>++控制结构和脚本语言待补充==++</strong></p>
<p><strong>参考链接：</strong></p>
<p><a href="http://man.linuxde.net/awk" target="_blank" rel="noopener">http://man.linuxde.net/awk</a><br><a href="https://coolshell.cn/articles/9070.html" target="_blank" rel="noopener">https://coolshell.cn/articles/9070.html</a></p>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2018/02/15/注释格式说明/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Naruto">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/137356299.jpg">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="木叶村">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/02/15/注释格式说明/" itemprop="url">突变注释格式说明</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-02-15T00:00:00+08:00">
                2018-02-15
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生信/" itemprop="url" rel="index">
                    <span itemprop="name">生信</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <p><img src="/myphoto/bdvj80uudkbj.png" alt="image"></p>
<div class="table-container">
<table>
<thead>
<tr>
<th>核酸变化</th>
<th style="text-align:center">氨基酸改变</th>
<th style="text-align:center">详细说明</th>
</tr>
</thead>
<tbody>
<tr>
<td>c.1915A&gt;G(exon16)</td>
<td style="text-align:center">p.N639D</td>
<td style="text-align:center">非同义突变，第639氨基酸蛋白位置上，氨基酸N突变为氨基酸D</td>
</tr>
<tr>
<td>c.112C&gt;T(exon1)</td>
<td style="text-align:center">p.R38X,70</td>
<td style="text-align:center">非同义突变，提前终止，第20氨基酸蛋白位置上，氨基酸C突变为终止密码子（用X表示），且相对原来缩短170个氨基酸</td>
</tr>
<tr>
<td>c.3052T&gt;C(exon20)</td>
<td style="text-align:center">p.X1018Q,11</td>
<td style="text-align:center">非同义突变，延长终止，第1018氨基酸蛋白位置为终止密码子（用X表示），突变为Q，且相对原来预测延长11个氨基酸，如果用“？”则表示无法计算延长长度</td>
</tr>
<tr>
<td>c.67(exon1)_c.69(exon1)delCTG</td>
<td style="text-align:center">p.L23del</td>
<td style="text-align:center">非框移突变，第23氨基酸蛋白位置缺失</td>
</tr>
<tr>
<td>c.3364(exon33)_c.3372(exon33)delGGCCCCCCA</td>
<td style="text-align:center">p.G1122_P1124del</td>
<td style="text-align:center">非框移突变，第1122位(G)至1124位(P)氨基酸缺失</td>
</tr>
<tr>
<td>c.3675(exon2)_c.3676(exon2)insAGC</td>
<td style="text-align:center">p.S1225_E1226insS</td>
<td style="text-align:center">非框移突变，第1225（S）至1226（E）位氨基酸蛋白发生之间插入氨基酸S</td>
</tr>
<tr>
<td>c.311(exon2)_c.313(exon2)delCCC</td>
<td style="text-align:center">p.T104_L105delinsM</td>
<td style="text-align:center">非框移突变，第104（T）至105（L）位氨基酸突变为氨基酸M，也可理解为：第104位至第105位置的氨基酸替换为M</td>
</tr>
<tr>
<td>c.1072(exon10)_c.1073(exon10)insAGGGACTGGGCGGGGCCATGGTCT</td>
<td style="text-align:center">p.W358delinsX,76</td>
<td style="text-align:center">非框移突变，导致第358位氨基酸蛋白W突变为终止密码子(用X表示)，且相对原来长度缩短76个氨基酸</td>
</tr>
<tr>
<td>c.417(exon3)delA</td>
<td style="text-align:center">p.X139Wfs*8</td>
<td style="text-align:center">框移突变，延长终止，导致终止密码子（第139位，用X表示）突变为氨基酸W，且相对原长度预测延长8个氨基酸（或从第139位氨基酸开始住数后第8个终止），如果用“？”则表示无法计算延长长度，即在3’UTR上找不到终止密码子</td>
</tr>
<tr>
<td>c.191(exon4)delA</td>
<td style="text-align:center">p.Y64Lfs*30</td>
<td style="text-align:center">框移突变，导致其编码的蛋白从第64位氨基酸开始算起，再翻译29个氨基酸（包含第64位）后终止，且这29个氨基酸是移码的，序列和原来的序列不同</td>
</tr>
</tbody>
</table>
</div>
<ul>
<li>c.-20T&gt;G(exon1)：基因5’utr上距离起始密码子的第-20位碱基T突变为G，位于第1外显子 </li>
<li>c.1A&gt;G(exon1)：基因编码区第1位碱基A突变为G，位于第1外显子 </li>
<li>c.21T&gt;A(exon1)：基因编码区第21位碱基T突变为A，位于第1外显子 </li>
<li>c.21+2T&gt;A(IVS1)：基因的第1个内含子第2位碱基G突变为A，距离第1外显子边界（c.21）2bp的位置 </li>
<li>c.61-2G&gt;A(IVS2)：基因的第2个内含子上距离第3外显子边界（c.61）2bp的位置 </li>
<li>c.*22G&gt;A(exon4)：基因3’utr上距离终止密码子的第22位碱基G突变为A，位于第4外显子</li>
</ul>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2018/02/14/二代数据质控及Trimmomatic的使用/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Naruto">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/137356299.jpg">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="木叶村">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/02/14/二代数据质控及Trimmomatic的使用/" itemprop="url">(转载)二代数据质控及Trimmomatic的使用</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-02-14T00:00:00+08:00">
                2018-02-14
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生信/" itemprop="url" rel="index">
                    <span itemprop="name">生信</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h3 id="数据质控的意义"><a href="#数据质控的意义" class="headerlink" title="数据质控的意义"></a>数据质控的意义</h3><p>现在的NGS测序，以illumina为首基本都是运用边合成边测序的技术。碱基的合成依靠的是化学反应，这使得碱基链可以不断地从5’端一直往3’端合成并延伸下去。但在这个合成的过程中随着合成链的增长，DNA聚合酶的效率会不断下降，特异性也开始变差，这就会带来一个问题,—越到后面碱基合成的错误率就会越高，这也是为何当前NGS测序读长普遍偏短的一个原因。</p>
<p><strong>注</strong>: 有时候测序仪在刚开始进行合成反应的时候也会由于反应还不够稳定，同样会带来质量值的波动，不过这个波动一般都在高质量值区域（如下图）。<br><img src="/myphoto/1.png" alt="image"></p>
<p>一般我们可以从如下几个方面来分析：</p>
<ul>
<li>read各个位置的碱基质量值分布</li>
<li>碱基的总体质量值分布</li>
<li>read各个位置上碱基分布比例，目的是为了分析碱基的分离程度</li>
<li>GC含量分布</li>
<li>read各位置的N含量</li>
<li>read是否还包含测序的接头序列</li>
<li>read重复率，这个是实验的扩增过程所引入的</li>
</ul>
<p>下面以软件FastQC质控结果为例，进行详细说明</p>
<p>在做read质量值分析的时候，FastQC并不单独查看具体某一条read中碱基的质量值，而是将Fastq文件中所有的read数据都综合起来一起分析。下图是一个测序质量非常好的read各位置碱基质量分布图（如下图）。</p>
<p><img src="/myphoto/2.png" alt="image"></p>
<p>这个图的横轴是read上碱基的位置，纵轴是碱基质量值。在这个例子中，read的长度是126bp（来自HiSeq X10的测序结果），这应该算是比较长的二代测序序列了。我们可以看到read上的每一个位置都有一个黄色的箱型图表示在该位置上所有碱基的质量分布情况。除了最后一个碱基之外，其他的碱基质量值都基本都在大于30，而且波动很小，说明质量很稳定，这其实是一个非常高质量的结果。而且我们可以看到图中质量值的分布都在绿色背景（代表高质量）的区域。</p>
<p>那如果是质量很差的结果看起来会是怎么样的呢？我手边一时找不到这样的数据，就在网上找到了一个代替品，样子如下</p>
<p><img src="/myphoto/3.png" alt="image"></p>
<p>在这个图中我们可以明显看到，read各个位置上的碱基质量分布波动都比较大，特别从第18个碱基往后全部出现了大幅度的波动，而且很多read的碱基质量值都掉到非常低（红色）的区域中了，说明这个数据的测序结果真的非常差，有着大量不及格的read。最好的情况是重新测序，但如果不得不使用这个数据，就要把这些低质量的数据全都去除掉才行，同时还需留意是否还存在其他的问题，但不管如何都一定会丢掉很大一部分的数据。</p>
<p>除了上面read各位置的碱基质量值分布之外，FastQC还会为我们计算其他几个非常有价值的统计结果，包括：</p>
<p>1）碱基总体质量值分布，只要大部分都高于20，那么就比较正常。</p>
<p><img src="/myphoto/4.png" alt="image"></p>
<p>一般来说，对于二代测序，<strong>最好是达到Q20的碱基要在95%以上（最差不低于90%），Q30要求大于85%（最差也不要低于80%）。</strong></p>
<p>2）read各个位置上碱基比例分布</p>
<p><img src="/myphoto/5.png" alt="image"></p>
<p>这个是为了分析碱基的分离程度。何为碱基分离？我们知道AT配对，CG配对，假如测序过程是比较随机的话（随机意味着好），那么在每个位置上A和T比例应该差不多，C和G的比例也应该差不多，如上图所示，两者之间即使有偏差也不应该太大，最好平均在1%以内，如果过高，除非有合理的原因，比如某些特定的捕获测序所致，否则都需要注意是不是测序过程有什么偏差。</p>
<p>3）GC含量分布图</p>
<p><img src="/myphoto/6.png" alt="image"></p>
<p>GC含量指的是G和C这两种碱基占总碱基的比例。二代测序平台或多或少都存在一定的测序偏向性，我们可以通过查看这个值来协助判断测序过程是否足够随机。对于人类来说，我们基因组的GC含量一般在40%左右。因此，如果发现GC含量的图谱明显偏离这个值那么说明测序过程存在较高的序列偏向性，结果就是基因组中某些特定区域被反复测序的几率高于平均水平，除了覆盖度会有偏离之后，将会影响下游的变异检测和CNV分析。</p>
<p>4）N含量分布图</p>
<p><img src="/myphoto/7.png" alt="image"></p>
<p>N在测序数据中一般是不应该出现的，如果出现则意味着，测序的光学信号无法被清晰分辨，如果这种情况多的话，往往意味着测序系统或者测序试剂的错误。</p>
<p>5）接头序列</p>
<p><img src="/myphoto/8.png" alt="image"></p>
<p>在测序之前需要构建测序文库，测序接头就是在这个时候加上的，其目的一方面是为了能够结合到flowcell上，另一方面是当有多个样本同时测序的时候能够利用接头信息进行区分。当测序read的长度大于被测序的DNA片段时，就会在read的末尾测到这些接头序列（如下图）。一般的WGS测序是不会测到这些接头序列的，因为构建WGS测序的文库序列（插入片段）都比较长，约几百bp，而read的测序长度都在100bp-150bp这个范围。不过在进行一些RNA测序的时候，由于它们的序列本来就比较短，很多只有几十bp长（特别是miRNA），那么就很容易会出现read测通的现象，这个时候就会在read的末尾测到这些接头序列。</p>
<p><img src="/myphoto/9.png" alt="image"></p>
<p>最后，这些被测到的接头序列和低质量碱基一样都是需要在正式分析之前进行切除的read片段。</p>
<h3 id="Trimmomatic的使用"><a href="#Trimmomatic的使用" class="headerlink" title="Trimmomatic的使用"></a>Trimmomatic的使用</h3><h4 id="Trimmomatic-介绍"><a href="#Trimmomatic-介绍" class="headerlink" title="Trimmomatic 介绍"></a>Trimmomatic 介绍</h4><p>Trimmomatic 发表的文章至今已被引用了 2810 次，是一个广受欢迎的 Illumina 平台数据过滤工具。其他平台的数据例如 Iron torrent ，PGM 测序数据可以用 fastx_toolkit 、NGSQC toolkit 来过滤。</p>
<p>Trimmomatic 支持多线程，处理数据速度快，主要用来去除 Illumina 平台的 Fastq 序列中的接头，并根据碱基质量值对 Fastq 进行修剪。软件有两种过滤模式，分别对应 SE 和 PE 测序数据，同时支持 gzip 和 bzip2 压缩文件。</p>
<p>另外也支持 phred-33 和 phred-64 格式互相转化，现在之所以会出现 phred-33 和 phred-64 格式的困惑，都是 Illumina 公司的锅（damn you, Illumina!），不过现在绝大部分 Illumina 平台的产出数据也都转为使用 phred-33 格式了。</p>
<h4 id="Trimmomatic-过滤步骤"><a href="#Trimmomatic-过滤步骤" class="headerlink" title="Trimmomatic 过滤步骤"></a>Trimmomatic 过滤步骤</h4><p><strong>Trimmomatic 过滤数据的步骤与命令行中过滤参数的顺序有关</strong>，通常的过滤步骤如下：</p>
<ol>
<li>ILLUMINACLIP: 过滤 reads 中的 Illumina 测序接头和引物序列，并决定是否去除反向互补的 R1/R2 中的 R2。</li>
<li>SLIDINGWINDOW: 从 reads 的 5’ 端开始，进行滑窗质量过滤，切掉碱基质量平均值低于阈值的滑窗。<br>MAXINFO: 一个自动调整的过滤选项，在保证 reads 长度的情况下尽量降低测序错误率，最大化 reads 的使用价值。</li>
<li>LEADING: 从 reads 的开头切除质量值低于阈值的碱基。</li>
<li>TRAILING: 从 reads 的末尾开始切除质量值低于阈值的碱基。</li>
<li>CROP: 从 reads 的末尾切掉部分碱基使得 reads 达到指定长度。</li>
<li>HEADCROP: 从 reads 的开头切掉指定数量的碱基。</li>
<li>MINLEN: 如果经过剪切后 reads 的长度低于阈值则丢弃这条 reads。</li>
<li>AVGQUAL: 如果 reads 的平均碱基质量值低于阈值则丢弃这条 reads。</li>
<li>TOPHRED33: 将 reads 的碱基质量值体系转为 phred-33。</li>
<li>TOPHRED64: 将 reads 的碱基质量值体系转为 phred-64。</li>
</ol>
<h4 id="Trimmomatic-简单用法"><a href="#Trimmomatic-简单用法" class="headerlink" title="Trimmomatic 简单用法"></a>Trimmomatic 简单用法</h4><p>由于 Trimmomatic 过滤数据的步骤与命令行中过滤参数的顺序有关，因此，如果需要去接头，建议第一步就去接头，否则接头序列被其他的过滤参数剪切掉部分之后就更难匹配更难去除干净了。</p>
<h5 id="单末端测序模式"><a href="#单末端测序模式" class="headerlink" title="单末端测序模式"></a>单末端测序模式</h5><p>在 SE 模式下，只有一个输入文件和一个过滤之后的输出文件：<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">java -jar &lt;path to trimmomatic jar&gt; SE [-threads &lt;threads&gt;] [-phred33 | -phred64] [-trimlog</span><br><span class="line">&lt;logFile&gt;] &lt;input&gt; &lt;output&gt; &lt;step 1&gt; &lt;step 2&gt; ...</span><br></pre></td></tr></table></figure></p>
<p><code>-trimlog</code> 参数指定了过滤日志文件名，日志中包含以下四列内容:</p>
<ul>
<li>read ID</li>
<li>过滤之后剩余序列长度</li>
<li>过滤之后的序列起始碱基位置（序列开头处被切掉了多少个碱基）</li>
<li>过滤之后的序列末端碱基位置</li>
<li>序列末端处被剪切掉的碱基数</li>
</ul>
<p>由于生成的 trimlog 文件中包含了每一条 reads 的处理记录，因此文件体积巨大（GB 级别），如果后面不会用到 trim 日志，建议不要使用这个参数。 </p>
<h5 id="双末端测序模式"><a href="#双末端测序模式" class="headerlink" title="双末端测序模式"></a>双末端测序模式</h5><p>在 PE 模式下，有两个输入文件，正向测序序列和反向测序序列，但是过滤之后输出文件有四个，过滤之后双端序列都保留的就是 paired，反之如果其中一端序列过滤之后被丢弃了另一端序列保留下来了就是 unpaired 。</p>
<p><img src="/myphoto/10.png" alt="image"></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">java -jar &lt;path to trimmomatic.jar&gt; PE [-threads &lt;threads] [-phred33 | -phred64] [-trimlog</span><br><span class="line">&lt;logFile&gt;] &gt;] [-basein &lt;inputBase&gt; | &lt;input 1&gt; &lt;input 2&gt;] [-baseout &lt;outputBase&gt; |</span><br><span class="line">&lt;paired output 1&gt; &lt;unpaired output 1&gt; &lt;paired output 2&gt; &lt;unpaired output 2&gt; &lt;step 1&gt; &lt;step 2&gt; ...</span><br></pre></td></tr></table></figure>
<p>其中 -phred33 和 -phred64 参数指定 fastq 的质量值编码格式，如果不设置这个参数，软件会自动判断输入文件是哪种格式（v0.32 之后的版本都支持），虽然软件默认的参数是 phred64，如果不确定序列是哪种质量编码格式，可以不设置这个参数。</p>
<h5 id="输入输出文件"><a href="#输入输出文件" class="headerlink" title="输入输出文件"></a>输入输出文件</h5><p>PE 模式的两个输入文件：<code>sample_R1.fastq</code> <code>sample_R2.fastq</code>以及四个输出文件：<code>sample_paired_R1.clean.fastq</code> <code>sample_unpaired_R1.clean.fastq</code> <code>sample_paired_R1.clean.fastq</code> <code>sample_unpaired_R1.clean.fastq</code></p>
<p>通常 PE 测序的两个文件，R1 和 R2 的文件名是类似的，因此可以使用 <code>-basein</code> 参数指定其中 R1 文件名即可，软件会推测出 R2 的文件名，但是这个功能实测并不好用，因为软件只能自动识别推测三种种格式的 <code>-basein</code>:</p>
<ul>
<li>Sample_Name_R1_001.fq.gz -&gt; Sample_Name_R2_001.fq.gz</li>
<li>Sample_Name.f.fastq -&gt; Sample_Name.r.fastq</li>
<li>Sample_Name.1.sequence.txt -&gt; Sample_Name.2.sequence.txt</li>
</ul>
<p>建议不用 <code>-basein</code> 参数，直接指定两个文件名（R1 和 R2）作为输入。</p>
<p>输出文件有四个，当然也可以像上文一样指定四个文件名，但是参数太长有点麻烦，有个省心的方法，使用 <code>-baseout</code> 参数指定输出文件的 <code>basename</code>，软件会自动为四个输出文件命名。例如 <code>-baseout mySampleFiltered.fq.gz</code> ，文件名中添加 <code>.gz</code> 后缀，软件会自动将输出结果进行 <code>gzip</code> 压缩。输出的四个文件分别会自动命名为：</p>
<ul>
<li>mySampleFiltered_1P.fq.gz - for paired forward reads</li>
<li>mySampleFiltered_1U.fq.gz - for unpaired forward reads</li>
<li>mySampleFiltered_2P.fq.gz - for paired reverse reads</li>
<li>mySampleFiltered_2U.fq.gz - for unpaired reverse reads</li>
</ul>
<p>此外，如果直接指定输入输出文件名，文件名后添加 <code>.gz</code> 后缀就是告诉软件输入文件是 <code>.gz</code> 压缩文件，输出文件需要用 gzip 压缩。</p>
<h5 id="常用参数说明"><a href="#常用参数说明" class="headerlink" title="常用参数说明"></a>常用参数说明</h5><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line">PE/SE</span><br><span class="line">    设定对Paired-End或Single-End的reads进行处理，其输入和输出参数稍有不一样。</span><br><span class="line">-threads</span><br><span class="line">    设置多线程运行数</span><br><span class="line">-phred33</span><br><span class="line">    设置碱基的质量格式，可选pred64</span><br><span class="line">ILLUMINACLIP:TruSeq3-PE.fa:2:30:10</span><br><span class="line">    切除adapter序列。参数后面分别接adapter序列的fasta文件：允许的最大mismatch</span><br><span class="line">数：palindrome模式下匹配碱基数阈值：simple模式下的匹配碱基数阈值。</span><br><span class="line">LEADING:3</span><br><span class="line">    切除首端碱基质量小于3的碱基</span><br><span class="line">TRAILING:3</span><br><span class="line">    切除尾端碱基质量小于3的碱基</span><br><span class="line">SLIDINGWINDOW:4:15</span><br><span class="line">    从5&apos;端开始进行滑动，当滑动位点周围一段序列(window)的平均碱基低于阈值，则从该处进行切除。Windows的size是4个碱基，其平均碱基</span><br><span class="line">质量小于15，则切除。</span><br><span class="line">MINLEN:50</span><br><span class="line">    最小的reads长度</span><br><span class="line">CROP:&lt;length&gt;</span><br><span class="line">    保留reads到指定的长度</span><br><span class="line">HEADCROP:&lt;length&gt;</span><br><span class="line">    在reads的首端切除指定的长度</span><br><span class="line">TOPHRED33</span><br><span class="line">    将碱基质量转换为pred33格式</span><br><span class="line">TOPHRED64</span><br><span class="line">    将碱基质量转换为pred64格式</span><br></pre></td></tr></table></figure>
<h5 id="过滤原理"><a href="#过滤原理" class="headerlink" title="过滤原理"></a>过滤原理</h5><p>每一步的过滤如果需要多个参数，通常用冒号:将各个参数隔开，当然参数的先后顺序是有要求的。</p>
<h6 id="ILUMINACLIP"><a href="#ILUMINACLIP" class="headerlink" title="ILUMINACLIP"></a>ILUMINACLIP</h6><p>从名字可以看出，这一步是为了去除 illumina 接头的，这个软件其实就是专为 illumina 平台数据而设计的。</p>
<p>为了更好理解测序 reads 中为什么会有引物和接头序列，我画了一个文库加上接头之后的结构示意图，也把引物结合部位大概标了出来：</p>
<p><img src="/myphoto/11.png" alt="image"></p>
<p>这个文库结构示意图理解之后就容易理解测序过程了。</p>
<p>去除接头以及引物序列看似简单，但需要权衡灵敏度（保证接头和引物去除干净）和特异性（保证不是接头和引物的序列不被误切除），由于测序中可能存在的随机错误让去接头这样一个简单的操作变的复杂。</p>
<p>虽然理论上接头序列和引物序列可能出现在 reads 中的任何位置，但实际上序列中出现接头和引物大部分情况下都是由于文库插入片段比测序读长短导致的，这种情况在 reads 的开头部分是有一段可用序列的，末端包含了接头的全长或部分序列，如果末端只有接头的一部分序列，那么去除这残缺的接头序列也不是容易的事。</p>
<p>然而，在 PE 测序模式下如果文库的插入片段比测序读长短，那么 read1 和 read2 中非接头序列的那部分会完全反向互补，Trimmomatic 有一个 ‘palindrome’ 模式会利用这个特点进行接头序列的去除。</p>
<p>下图中 A、B、C、D 四种情况就是 Trimmomatic 去除接头和引物的四种模式：</p>
<ul>
<li>红色条形：被切除的序列</li>
<li>绿色条形：保留下来的有效读长 </li>
<li>深蓝色条形：接头序列</li>
<li>浅蓝色条形：引物序列</li>
</ul>
<p><img src="/myphoto/12.png" alt="image"></p>
<p><strong>A 模式</strong>：测序 reads 从起始位置开始就包含了完整的接头序列，那么根据 Illumina 测序原理，这整条 reads 都不可能包含有用序列了，整条 reads 被丢弃。</p>
<p><strong>B 模式</strong>：这种相对常见，由于文库插入片段比测序读长短，会在 reads 末端包含部分接头序列，若是这部分接头序列足够长是可以识别并去除的，但如果接头序列太短，比接头匹配参数设置的最短长度还短，那么就无法去除。但是，如果是 PE 测序，可以按照 D 模式去除 reads 末端的很短的接头序列。</p>
<p><strong>C 模式</strong>：PE 测序可能出现这种情况，正向测序和反向测序有部分完全反向互补，但是空载的文库，两个接头直接互连，这样的 reads 不包含任何有用序列，正反向测序 reads 都被丢弃。</p>
<p><strong>D 模式</strong>：是 Trimmomatic 利用 PE 测序进行短接头序列去除的典范，如果文库插入片段比测序读长短，利用正反向测序 reads 中一段碱基可以完全反向互补的特点，将两个接头序列与 reads 进行比对，同时两条 reads 之间也互相比对，可以将 3’ 末端哪怕只有 1bp 的接头序列都可以被准确去除，相对 B 模式去除接头污染更彻底。</p>
<p>Trimmomatic 使用了一种类似序列比对软件（例如 Isaac aligner，一个超快速的 alignment 软件）的两步策略来搜索潜在的接头序列。首先，使用接头序列中的一段种子序列（seed 长度不超过 16bp）与测序 reads 进行比对，如果种子序列在测序 reads 中有足够好的比对结果（具体由 seedMismatch 参数决定），就启动第二步的接头全长与 reads 比对。第一步的 seed 搜索速度很快，可以过滤掉没有接头污染的 reads ，这种两步搜索的方法使得接头序列的查找效率很高。</p>
<p>在第二步的接头序列和测序 reads 全长比对统计比对分值时，罚分策略考虑了测序碱基的质量值Q，每一个比对上的碱基加分 0.6，每一个错配的碱基减分 Q/10，考虑碱基质量值可以降低低质量碱基（高测序错误率）错配对整个比对得分的影响。在这个规则下，一段 12bp 的接头序列完全比对到 reads 上得分为 7.2, 25bp 的接头序列完全比对到 reads 上得分为 15。因此在 ILLUMINACLIP 参数中 simple clip threshold 的值建议为 7-15 之间（即上图中 A/B 比对模式比对得分阈值）。</p>
<p>对于 palindromic 模式的比对（上图中 D 模式），可以比对上的序列长度会更长，为了保证识别接头序列的准确率，比对得分的阈值也更高，例如 reads的 R1 和 R2 中有 50bp 序列可以反向互补匹配，得分为 30。这种模式下，Trimmomatic 可以识别并去除 reads 中非常短的接头序列。</p>
<p><strong>ILLUMINACLIP 参数说明</strong>：按照规定顺序，ILLUMINACLIP 的参数列表如下（各个参数之间以冒号分开），PE 测序需要注意最后一个参数。对于 SE 测序最后两个参数可以不设置。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">ILLUMINACLIP:&lt;fastaWithAdaptersEtc&gt;:&lt;seed mismatches&gt;:&lt;palindrome clip threshold&gt;:&lt;simple clip threshold&gt;:&lt;minAdapterLength&gt;:&lt;keepBothReads&gt;</span><br><span class="line"></span><br><span class="line">ILLUMINACLIP:TruSeq3-SE:2:30:10 #接头和引物序列在 TruSeq3-SE 中，第一步 seed 搜索允许2个碱基错配，palindrome 比对分值阈值 30，simple clip 比对分值阈值 10，palindrome 模式允许切除的最短接头序列为 8bp（默认值），palindrome 模式去除与 R1 完全反向互补的 R2（默认去除）</span><br></pre></td></tr></table></figure>
<p><code>fastaWithAdaptersEtc</code>：指定包含接头和引物序列（所有被视为污染的序列）的 fasta 文件路径，Trimmomatic 自带了一个包含 Illumina 平台接头和引物序列的 fasta 文件，可以直接用这个。</p>
<p><code>seedMismatches</code>：指定第一步 seed 搜索时允许的错配碱基个数，例如 2。</p>
<p><code>palindrome clip threshold</code>：指定针对 PE 的 palindrome clip 模式下，需要 R1 和 R2 之间至少多少比对分值（上图中 D 模式），才会进行接头切除，例如 30。</p>
<p><code>simple clip threshold</code>：指定切除接头序列的最低比对分值（上图 A/B 模式），通常 7-15 之间。</p>
<p><code>minAdapterLength</code>：只对 PE 测序的 palindrome clip 模式有效，指定 palindrome 模式下可以切除的接头序列最短长度，由于历史的原因，默认值是 8，但实际上 palindrome 模式可以切除短至 1bp 的接头污染，所以可以设置为 1 。</p>
<p><code>keepBothReads</code>：只对 PE 测序的 palindrome clip 模式有效，这个参数很重要，在上图中 D 模式下， R1 和 R2 在去除了接头序列之后剩余的部分是完全反向互补的，默认参数 false，意味着整条去除与 R1 完全反向互补的 R2，当做重复去除掉，但在有些情况下，例如需要用到 paired reads 的 bowtie2 流程，就要将这个参数改为 true，否则会损失一部分 paired reads。</p>
<p>看一个 PE150 数据的测试，就知道 <code>keepBothReads</code> 参数的重要性了：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">$ java -jar trimmomatic-0.36.jar PE -phred33 F-2-test_R1.fastq.gz F-2-test_R2.fastq.gz -baseout F-2.fastq.gz ILLUMINACLIP:TruSeq3-PE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:51</span><br><span class="line"></span><br><span class="line">ILLUMINACLIP: Using 1 prefix pairs, 0 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences</span><br><span class="line">Input Read Pairs: 2500 Both Surviving: 1633 (65.32%) Forward Only Surviving: 828 (33.12%) Reverse Only Surviving: 12 (0.48%) Dropped: 27 (1.08%)</span><br><span class="line">TrimmomaticPE: Completed successfully</span><br><span class="line"># 使用 ILLUMINACLIP 默认的第六个参数 false，只有 65.32% paired reads 保留下来</span><br><span class="line"></span><br><span class="line">$ java -jar trimmomatic-0.36.jar PE -phred33 F-2-test_R1.fastq.gz F-2-test_R2.fastq.gz -baseout F-2.fastq.gz ILLUMINACLIP:TruSeq3-PE.fa:2:30:10:8:true LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:51</span><br><span class="line"></span><br><span class="line">ILLUMINACLIP: Using 1 prefix pairs, 0 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences</span><br><span class="line">Input Read Pairs: 2500 Both Surviving: 2439 (97.56%) Forward Only Surviving: 22 (0.88%) Reverse Only Surviving: 16 (0.64%) Dropped: 23 (0.92%)</span><br><span class="line">TrimmomaticPE: Completed successfully</span><br><span class="line"># 将 ILLUMINACLIP 第六个参数改为 true，其余所有参数均相同，结果有 97.56% paired reads 保留下来</span><br></pre></td></tr></table></figure>
<h6 id="SLIDINGWINDOW"><a href="#SLIDINGWINDOW" class="headerlink" title="SLIDINGWINDOW"></a>SLIDINGWINDOW</h6><p>滑窗剪切，统计滑窗口中所有碱基的平均质量值，如果低于设定的阈值，则切掉窗口。<br>SLIDINGWINDOW 参数如下：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">SLIDINGWINDOW:&lt;windowSize&gt;:&lt;requiredQuality&gt;</span><br><span class="line"></span><br><span class="line">SLIDINGWINDOW:4:15  #设置 4bp 窗口，碱基平均质量值阈值 15</span><br></pre></td></tr></table></figure>
<p><code>widowSize</code>:设置窗口大小<br><code>requiredQuality</code>：设置窗口碱基平均质量阈值</p>
<h6 id="MAXINFO"><a href="#MAXINFO" class="headerlink" title="MAXINFO"></a>MAXINFO</h6><p>包含一个可以自动调整的过滤条件，在保留尽可能长的序列和保持序列中碱基测序错误率尽可能低之间进行平衡，以达到 trim 之后保留序列的价值最大化。</p>
<p>对于不同的应用场景，一条 reads 序列的价值由以下三个因素决定：</p>
<ol>
<li>允许的最短 read 长度：只有在测序 reads 足够长的情况下，才能将 reads map 到参考序列的唯一位置上。过短的序列往往能 map 到参考基因组上的多个位置（特异性差），这种非特异性的 map 结果可以给出的有价值信息很少（对后续分析没什么用）。保证 reads 可以 map 到参考序列的唯一位置上需要的 reads 最短长度也与参考序列本身的长度和序列复杂度有关，但是通常最短 reads 要求不低于 40bp。</li>
<li>需求的额外 reads 长度：在保证 reads 可以 map 到参考序列上的唯一位置的情况下， reads 序列越长，可能对后续分析越有利。当然，这也需要看应用场景，对于 RNA-seq 来说只需要统计某个位置 map 上的 reads 数目，这种情况下 reads 长度只要满足唯一比对要求即可。而对于 de-novo 序列组装或者 Resequencing 检测变异，reads 序列长度越长越好，reads 越长可以为后续分析提供越多的有效证据。</li>
<li>对测序错误的敏感性：即测序错误率对后续分析工作是否造成严重影响，有的应用中对 reads 测序错误零容忍，就需要非常严格的过滤条件，而有的应用对测序错误不敏感，就可以使用宽松的过滤条件。</li>
</ol>
<p>MAXINFO 有两个参数，第一个 <code>target read length</code> 控制上面的因素一，即允许的最短 read 长度。第二个参数 <code>strictness</code> 是控制因素二和因素三之间的平衡，即满足最短 read 长度的情况下，是保留尽可能多的碱基，还是保证尽可能低的测序错误率。</p>
<p>MAXINFO 过滤从 reads 3’ 端开始进行剪切，在考虑上述三个因素的情况下统计所有可能的 trim 方式的到的 clean reads 的 INFO 分值（即所谓的 reads 价值），这三个因素分别以不同的方式影响最终的 reads INFO 分值：</p>
<ol>
<li>最短 read 长度：最终保留的 reads 长度与 INFO 分值之间是 logistic 函数关系，即 INFO 分值与 reads 长度之间是 S 形增长曲线（类似大肠杆菌生长曲线），大概相当于，当保留下来的 reads 长度小于最短长度时，INFO 分值随着 reads 长度增加呈指数级增长，当保留下来的 reads 长度达到最短 read 长度要求之后，INFO 分值的增长会变缓。</li>
<li>超出最短 read 长度的部分：INFO 分值会随着额外的 reads 长度增加了线性增长，线性系数与 MAXINFO 的第二个参数有关，为 1 - strictness。</li>
<li>测序错误敏感性：最终保留下来的所有碱基的质量值都被用来计算 reads 包含测序错误的概率，出现测序错误的概率提升会降低 INFO 分值，错误概率对 INFO 分值的影响权重与 strictness 相关。</li>
</ol>
<p>针对一条 read 的任何可能的剪切方式都计算出 INFO 分值，最终的 reads 长度和切除的碱基由 INFO 最大值决定。实际上这三个影响因子作用的方式不同：</p>
<ol>
<li>当 reads 长度比最短 read 长度还短时 INFO 分值被序列长度主导，因为太短的 reads 根本无法提供足够的有用信息。</li>
<li>当 reads 长度满足最短 read 长度要求时，reads 的长度因素就不再是影响 INFO 分值的主导因素了，而且一旦 reads 长度足够长时，由于 logistic 函数的关系，reads 长度就不怎么贡献 INFO 分值了。</li>
<li>超出最短 read 长度要求的碱基对 INFO 贡献有限，而另一方面，由于碱基的增加导致 reads 中出现测序错误的概率增大，这会导致对 INFO 的罚分。控制这超出最短 read 长度的序列对 INFO 分值的影响到底是罚分还是得分，就看 strictness 参数了。</li>
<li>对于大部分足够长的 reads，保留超出最低要求的碱基序列是利还是弊，会依照碱基质量值分布和 strictness 参数来决定如何 trim。</li>
</ol>
<p>参数说明：<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">MAXINFO:&lt;targetLength&gt;:&lt;strictness&gt;</span><br></pre></td></tr></table></figure></p>
<p><code>targetLength</code>:使得 reads 可以 map 到参考序列上唯一位置的最短长度（likely）。<br><code>strictness</code>：一个介于 0 - 1 之间的小数，决定如何平衡 最大化 reads 长度 或者 最小化 reads 出现错误的概率，当参数设置小于 0.2 时倾向于最大化 reads 长度，当参数设置大于 0.8 时倾向于最小化 reads 中出现测序错误的概率。</p>
<h6 id="LEADING"><a href="#LEADING" class="headerlink" title="LEADING"></a>LEADING</h6><p>从 reads 的起始端开始切除质量值低于设定的阈值的碱基，直到有一个碱基其质量值达到阈值。<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">LEADING:&lt;quality&gt;</span><br></pre></td></tr></table></figure></p>
<p><code>quality</code>:设定碱基质量值阈值，低于这个阈值将被切除。</p>
<h6 id="TRAILING"><a href="#TRAILING" class="headerlink" title="TRAILING"></a>TRAILING</h6><p>从 reads 的末端开始切除质量值低于设定阈值的碱基，直到有一个碱基质量值达到阈值。Illumina 平台有些低质量的碱基质量值被标记为 2 ，所以设置为 3 可以过滤掉这部分低质量碱基。官方推荐使用 Sliding Window 或 MaxInfo 来代替 LEADING 和 TAILING。<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">TRAILING:&lt;quality&gt;</span><br></pre></td></tr></table></figure></p>
<p><code>quality</code>:设定碱基质量值阈值，低于这个阈值将被切除。</p>
<h6 id="CROP"><a href="#CROP" class="headerlink" title="CROP"></a>CROP</h6><p>不管碱基质量，从 reads 的起始开始保留设定长度的碱基，其余全部切除。一刀切，把所有 reads 切成相同的长度。<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">CROP:&lt;length&gt;</span><br></pre></td></tr></table></figure></p>
<p><code>length</code>：reads 从末端除之后保留下来的序列长度</p>
<h6 id="HEADCROP"><a href="#HEADCROP" class="headerlink" title="HEADCROP"></a>HEADCROP</h6><p>不管碱基质量，从 reads 的起始开始直接切除部分碱基。<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">HEADCROP:&lt;length&gt;</span><br></pre></td></tr></table></figure></p>
<p><code>length</code>：从 reads 的起始开始切除的碱基数</p>
<h6 id="MINLEN"><a href="#MINLEN" class="headerlink" title="MINLEN"></a>MINLEN</h6><p>设定一个最短 read 长度，当 reads 经过前面的过滤之后，如果保留下来的长度低于这个阈值，整条 reads 被丢弃。被丢弃的 reads 数会被统计在 Trimmomatic 日志的 dropped reads 中。<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">MINLEN:&lt;length&gt;</span><br></pre></td></tr></table></figure></p>
<p><code>length</code>：可被保留的最短 read 长度</p>
<h6 id="TOPHRED33"><a href="#TOPHRED33" class="headerlink" title="TOPHRED33"></a>TOPHRED33</h6><p>此选项可以将过滤之后的 Fastq 文件中质量值这一行转为 phred-33 格式。</p>
<h6 id="TOPHRED64"><a href="#TOPHRED64" class="headerlink" title="TOPHRED64"></a>TOPHRED64</h6><p>此选项可以将过滤之后的 Fastq 文件中质量值这一行转为 phred-64 格式。</p>
<h6 id="接头序列和引物序列"><a href="#接头序列和引物序列" class="headerlink" title="接头序列和引物序列"></a>接头序列和引物序列</h6><p>Trimmomatic 也可以自己制作包含接头和引物序列的 fasta 文件，格式可以参考软件自带的 adapters 文件夹中的格式。</p>
<p>adapters 文件夹中包含 illumina 测序 TruSeq2，TruSeq3 针对 SE 和 PE 的通用接头和引物序列。</p>
<p><strong>参考链接</strong>：<br><a href="https://www.jianshu.com/p/a8935adebaae" target="_blank" rel="noopener">https://www.jianshu.com/p/a8935adebaae</a><br><a href="https://www.jianshu.com/p/7b5591673255" target="_blank" rel="noopener">https://www.jianshu.com/p/7b5591673255</a><br><a href="http://www.huangshujia.me/2017/08/25/2017-08-25-Begining-WGS-Data-Analysis-Fastq-Data-Quality-Control.html" target="_blank" rel="noopener">http://www.huangshujia.me/2017/08/25/2017-08-25-Begining-WGS-Data-Analysis-Fastq-Data-Quality-Control.html</a></p>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2018/02/13/IchorCNA/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Naruto">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/137356299.jpg">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="木叶村">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/02/13/IchorCNA/" itemprop="url">ichorCNA</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-02-13T00:00:00+08:00">
                2018-02-13
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生信/" itemprop="url" rel="index">
                    <span itemprop="name">生信</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h4 id="1-简介"><a href="#1-简介" class="headerlink" title="1. 简介"></a>1. 简介</h4><p>ichorCNA是从超低通全基因组测序（ULP-WGS，0.1x覆盖率）估计无细胞DNA中肿瘤分数的工具。</p>
<p><strong>描述</strong></p>
<p>ichorCNA使用概率模型，作为隐马尔可夫模型（HMM）实现，以同时分割基因组，预测大规模拷贝数改变，并估计超低通全基因组测序样本（ULP-WGS ）。ichorCNA针对样本的低覆盖率（〜0.1x）测序进行了优化，并使用患者和健康供体cfDNA样本进行了基准测试。</p>
<p><strong>用途</strong></p>
<p>ichorCNA可用于通知肿瘤来源DNA的存在或缺失，并指导决定进行整个外显子或更深的全基因组测序。此外，我们可以使用肿瘤分数的定量估计来校准期望的测序深度，以达到用于鉴定无细胞DNA中的突变的统计功效。最后，ichorCNA可以利用超低通量测序的成本效益方法来检测来自大群体的大规模拷贝数变化。</p>
<p><strong>文献</strong></p>
<p>Adalsteinsson, Ha, Freeman, et al. Scalable whole-exome sequencing of cell-free DNA reveals high concordance with metastatic tumors</p>
<h4 id="2-安装"><a href="#2-安装" class="headerlink" title="2. 安装"></a>2. 安装</h4><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">#R控制台中</span><br><span class="line">install.packages(&quot;plyr&quot;)</span><br><span class="line">source(&quot;https://bioconductor.org/biocLite.R&quot;)</span><br><span class="line">biocLite(&quot;HMMcopy&quot;)</span><br><span class="line">biocLite(&quot;GenomeInfoDb&quot;)</span><br><span class="line">install.packages(&quot;devtools&quot;)</span><br><span class="line">library(devtools)</span><br><span class="line">install_github(&quot;broadinstitute/ichorCNA&quot;, &quot;--no-docs&quot;)</span><br><span class="line">#依赖的包有很多，安装过程报错信息可以看出需要依赖的其他包</span><br></pre></td></tr></table></figure>
<h4 id="3-使用"><a href="#3-使用" class="headerlink" title="3. 使用"></a>3. 使用</h4><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line">#以样本DYDFA-842-31为例</span><br><span class="line">$bwa mem -t 4 -B 3 -O 11 -E 4 -k 32 -Y hg19_reference_with_NC.fasta </span><br><span class="line">DYDFA-842-31/DYDFA-842-31_R1.fq.gz DYDFA-842-31_R2.fq.gz &gt; ichorCNA/test31/31.sam</span><br><span class="line"></span><br><span class="line">$samtools view -bS ichorCNA/test31/31.sam &gt; ichorCNA/test31/31.bam</span><br><span class="line"></span><br><span class="line">$samtools sort 31.bam &gt; 31.sort.bam &amp;</span><br><span class="line"></span><br><span class="line">$samtools  index 31.sort.bam &amp;</span><br><span class="line"></span><br><span class="line">$java -jar picard-2.8.2.jar MarkDuplicates I=./31.sort.bam O=./31.rmdup.bam M=rmdup_stat.txt REMOVE_DUPLICATES=true &amp;</span><br><span class="line"></span><br><span class="line">$samtools index 31.rmdup.bam  &amp;</span><br><span class="line"></span><br><span class="line">hmmcopy_utils-master/bin/readCounter --window 1000000 --quality 20 --chromosome chr1,chr2,chr3,chr4,chr5,chr6,chr7,chr8,chr9,chr10,chr11,chr12,chr13,chr14,chr15,chr16,chr17,chr18,chr19,chr20,chr21,chr22,chrX,chrY 31.rmdup.bam &gt; 31.wig &amp;</span><br><span class="line"></span><br><span class="line">$Rscript ichorCNA-master/scripts/runIchorCNA.R --id test31 --WIG test31.wig \</span><br><span class="line">    --gcWig ichorCNA-master/inst/extdata/gc_hg19_1000kb.wig \</span><br><span class="line">    --mapWig ichorCNA-master/inst/extdata/map_hg19_1000kb.wig \</span><br><span class="line">    --centromere ichorCNA-master/inst/extdata/GRCh37.p13_centromere_UCSC-gapTable.txt \</span><br><span class="line">    --normalPanel ichorCNA-master/inst/extdata/HD_ULP_PoN_1Mb_median_normAutosome_mapScoreFiltered_median.rds \</span><br><span class="line">    --normal &quot;c(0.5,0.6,0.7,0.8,0.9)&quot;  --scStates &quot;c(1,3)&quot; \</span><br><span class="line">    --txnE 0.9999 --txnStrength 10000 --outDir ./</span><br></pre></td></tr></table></figure>
<h4 id="4-输出"><a href="#4-输出" class="headerlink" title="4. 输出"></a>4. 输出</h4><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">test31</span><br><span class="line">Gender: male</span><br><span class="line">Tumor Fraction: 0</span><br><span class="line">Ploidy: 2.001</span><br><span class="line">Subclone Fraction:      NA</span><br></pre></td></tr></table></figure>
<p><img src="/myphoto/clipboard31.png" alt="31"><br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">test32</span><br><span class="line">Gender: male</span><br><span class="line">Tumor Fraction: 0</span><br><span class="line">Ploidy: 2</span><br><span class="line">Subclone Fraction:      NA</span><br></pre></td></tr></table></figure></p>
<p><img src="/myphoto/clipboard32.png" alt="32"><br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">test33</span><br><span class="line">Gender: female</span><br><span class="line">Tumor Fraction: 0</span><br><span class="line">Ploidy: 2</span><br><span class="line">Subclone Fraction:      NA</span><br></pre></td></tr></table></figure></p>
<p><img src="/myphoto/clipboard33.png" alt="33"><br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">test34</span><br><span class="line">Gender: female</span><br><span class="line">Tumor Fraction: 0</span><br><span class="line">Ploidy: 1.999</span><br><span class="line">Subclone Fraction:      NA</span><br></pre></td></tr></table></figure></p>
<p><img src="/myphoto/clipboard34.png" alt="34"><br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">test35</span><br><span class="line">Gender: female</span><br><span class="line">Tumor Fraction: 0.404</span><br><span class="line">Ploidy: 2.98</span><br><span class="line">Subclone Fraction:      0.5017</span><br></pre></td></tr></table></figure></p>
<p><img src="/myphoto/clipboard35.png" alt="35"></p>
<h4 id="5-解释结果"><a href="#5-解释结果" class="headerlink" title="5. 解释结果"></a>5. 解释结果</h4><h4 id="6-帮助文档"><a href="#6-帮助文档" class="headerlink" title="6. 帮助文档"></a>6. 帮助文档</h4><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br></pre></td><td class="code"><pre><span class="line">Rscript runIchorCNA.R --help</span><br><span class="line">Usage: runIchorCNA.R [options]</span><br><span class="line"></span><br><span class="line">Options:</span><br><span class="line">--WIG=WIG</span><br><span class="line">	WIG 文件(必须)</span><br><span class="line"></span><br><span class="line">--NORMWIG=NORMWIG</span><br><span class="line">	标准 WIG 文件 Default: [NULL]</span><br><span class="line"></span><br><span class="line">--gcWig=GCWIG</span><br><span class="line">	GC-content WIG 文件; (必须);软件给出的是hg19的GC WIG文件</span><br><span class="line"></span><br><span class="line">--mapWig=MAPWIG</span><br><span class="line">	 WIG 文件  Default: [NULL]</span><br><span class="line"></span><br><span class="line">--normalPanel=NORMALPANEL</span><br><span class="line">	深度矫正文件 Default: [NULL]</span><br><span class="line"></span><br><span class="line">--exons.bed=EXONS.BED</span><br><span class="line">	外显子bed文件 Default: [NULL]</span><br><span class="line"></span><br><span class="line">--id=ID</span><br><span class="line">	样本名 Default: [test]</span><br><span class="line"></span><br><span class="line">--centromere=CENTROMERE</span><br><span class="line">	着丝粒位置文件;ichorCNA提供的是hg19版本的. Default: [NULL]</span><br><span class="line"></span><br><span class="line">--rmCentromereFlankLength=RMCENTROMEREFLANKLENGTH</span><br><span class="line">    去除着丝粒侧翼的长度. Default: [1e+05]</span><br><span class="line"></span><br><span class="line">--normal=NORMAL</span><br><span class="line">	Initial normal contamination; can be more than one value if additional normal initializations are desired. Default: [0.5]</span><br><span class="line">    #初始正常污染; 如果需要额外的正常初始化，可以是多于一个值</span><br><span class="line">--scStates=SCSTATES</span><br><span class="line">	亚克隆状态. Default: [NULL]</span><br><span class="line"></span><br><span class="line">--coverage=COVERAGE</span><br><span class="line">	PICARD 软件计算出的覆盖度文件. Default: [NULL]</span><br><span class="line"></span><br><span class="line">--lambda=LAMBDA</span><br><span class="line">	Initial Student&apos;s t precision; must contain 4 values (e.g. c(1500,1500,1500,1500)); if not provided then will automatically use based on variance of data. Default: [NULL]</span><br><span class="line">    # 初始T-test精确度; 必须包含4个值（例如c（1500,1500,1500,1500））;如果没有提供，则会根据数据的方差自动使用</span><br><span class="line">--lambdaScaleHyperParam=LAMBDASCALEHYPERPARAM</span><br><span class="line">	Hyperparameter (scale) for Gamma prior on Student&apos;s-t precision. Default: [3]</span><br><span class="line">    #Gamma之前的超参数（规模）的t精度</span><br><span class="line">--ploidy=PLOIDY</span><br><span class="line">	Initial tumour ploidy; can be more than one value if additional ploidy initializations are desired. Default: [2]</span><br><span class="line">    # 最初的肿瘤倍性; 默认二倍体；</span><br><span class="line">--maxCN=MAXCN</span><br><span class="line">	Total clonal CN states. Default: [7]</span><br><span class="line"></span><br><span class="line">--estimateNormal=ESTIMATENORMAL</span><br><span class="line">	Estimate normal. Default: [TRUE]</span><br><span class="line"></span><br><span class="line">--estimateScPrevalence=ESTIMATESCPREVALENCE</span><br><span class="line">	Estimate subclonal prevalence. Default: [TRUE]</span><br><span class="line"></span><br><span class="line">--estimatePloidy=ESTIMATEPLOIDY</span><br><span class="line">	Estimate tumour ploidy. Default: [TRUE]</span><br><span class="line"></span><br><span class="line">--maxFracCNASubclone=MAXFRACCNASUBCLONE</span><br><span class="line">	Exclude solutions with fraction of subclonal events greater than this value. Default: [0.7]</span><br><span class="line">    # 排除大于此值的部分亚克隆事件的解决方案</span><br><span class="line">--maxFracGenomeSubclone=MAXFRACGENOMESUBCLONE</span><br><span class="line">	Exclude solutions with subclonal genome fraction greater than this value. Default: [0.5]</span><br><span class="line">    # 排除亚克隆基因组分数大于此值的解决方案</span><br><span class="line">--minSegmentBins=MINSEGMENTBINS</span><br><span class="line">	Minimum number of bins for largest segment threshold required to estimate tumor fraction; if below this threshold, then will be assigned zero tumor fraction.</span><br><span class="line"></span><br><span class="line">--altFracThreshold=ALTFRACTHRESHOLD</span><br><span class="line">	Minimum proportion of bins altered required to estimate tumor fraction; if below this threshold, then will be assigned zero tumor fraction. Default: [0.05]</span><br><span class="line">    # 评估肿瘤分数所需的最大分割阈值的最小分箱数量; 如果低于这个阈值，则将被分配零肿瘤分数</span><br><span class="line">--chrNormalize=CHRNORMALIZE</span><br><span class="line">	Specify chromosomes to normalize GC/mappability biases. Default: [c(1:22)]</span><br><span class="line">    #指定染色体</span><br><span class="line">--chrTrain=CHRTRAIN</span><br><span class="line">	Specify chromosomes to estimate params. Default: [c(1:22)]</span><br><span class="line">    #指定染色体</span><br><span class="line">--chrs=CHRS</span><br><span class="line">	Specify chromosomes to analyze. Default: [c(1:22,&quot;X&quot;)]</span><br><span class="line">    #指定染色体</span><br><span class="line">--normalizeMaleX=NORMALIZEMALEX</span><br><span class="line">	If male, then normalize chrX by median. Default: [TRUE]</span><br><span class="line">    # 男性按照X染色体分析</span><br><span class="line">--fracReadsInChrYForMale=FRACREADSINCHRYFORMALE</span><br><span class="line">	Threshold for fraction of reads in chrY to assign as male. Default: [0.001]</span><br><span class="line">    #将chrY中的读数部分的阈值分配为男性</span><br><span class="line">--includeHOMD=INCLUDEHOMD</span><br><span class="line">	If FALSE, then exclude HOMD state. Useful when using large bins (e.g. 1Mb). Default: [FALSE]</span><br><span class="line">    # 如果FALSE，则排除HOMD状态。 使用大容量箱（如1Mb）时很有用</span><br><span class="line">--txnE=TXNE</span><br><span class="line">	Self-transition probability. Increase to decrease number of segments. Default: [0.9999999]</span><br><span class="line">    #　自我转换概率。 增加以减少段数</span><br><span class="line">--txnStrength=TXNSTRENGTH</span><br><span class="line">	Transition pseudo-counts. Exponent should be the same as the number of decimal places of --txnE. Default: [1e+07]</span><br><span class="line">    # 转换伪计数。 指数应与--txnE的小数位数相同</span><br><span class="line">--plotFileType=PLOTFILETYPE</span><br><span class="line">	File format for output plots. Default: [pdf]</span><br><span class="line">    #图片输出格式</span><br><span class="line">--plotYLim=PLOTYLIM</span><br><span class="line">	ylim to use for chromosome plots. Default: [c(-2,2)]</span><br><span class="line">    y轴范围</span><br><span class="line">--outDir=OUTDIR</span><br><span class="line">	Output Directory. Default: [./]</span><br><span class="line"></span><br><span class="line">--libdir=LIBDIR</span><br><span class="line">	Script library path. Usually exclude this argument unless custom modifications have been made to the ichorCNA R package code and the user would like to source those R files. Default: [NULL]</span><br></pre></td></tr></table></figure>
          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2018/02/13/安装python3及模块/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Naruto">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/137356299.jpg">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="木叶村">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/02/13/安装python3及模块/" itemprop="url">安装python3及模块</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-02-13T00:00:00+08:00">
                2018-02-13
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/python3/" itemprop="url" rel="index">
                    <span itemprop="name">python3</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h3 id="python"><a href="#python" class="headerlink" title="python"></a>python</h3><p><strong>linux下安装python</strong></p>
<ol>
<li>官网下载源码压缩包，解压 </li>
<li>setup.py 里面可以设置sqlite3的搜索路径， 以防止模块安装不上</li>
<li><p>./configure</p>
<blockquote>
<p>安装文件有很多选项</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br><span class="line">125</span><br><span class="line">126</span><br><span class="line">127</span><br><span class="line">128</span><br><span class="line">129</span><br><span class="line">130</span><br><span class="line">131</span><br><span class="line">132</span><br><span class="line">133</span><br><span class="line">134</span><br><span class="line">135</span><br><span class="line">136</span><br><span class="line">137</span><br><span class="line">138</span><br><span class="line">139</span><br><span class="line">140</span><br><span class="line">141</span><br><span class="line">142</span><br><span class="line">143</span><br><span class="line">144</span><br><span class="line">145</span><br><span class="line">146</span><br><span class="line">147</span><br><span class="line">148</span><br><span class="line">149</span><br><span class="line">150</span><br><span class="line">151</span><br><span class="line">152</span><br><span class="line">153</span><br></pre></td><td class="code"><pre><span class="line">./configure -h</span><br><span class="line">`configure&apos; configures python 3.6 to adapt to many kinds of systems.</span><br><span class="line"></span><br><span class="line">Usage: ./configure [OPTION]... [VAR=VALUE]...</span><br><span class="line"></span><br><span class="line">To assign environment variables (e.g., CC, CFLAGS...), specify them as</span><br><span class="line">VAR=VALUE.  See below for descriptions of some of the useful variables.</span><br><span class="line"></span><br><span class="line">Defaults for the options are specified in brackets.</span><br><span class="line"></span><br><span class="line">Configuration:</span><br><span class="line">  -h, --help              display this help and exit</span><br><span class="line">      --help=short        display options specific to this package</span><br><span class="line">      --help=recursive    display the short help of all the included packages</span><br><span class="line">  -V, --version           display version information and exit</span><br><span class="line">  -q, --quiet, --silent   do not print `checking ...&apos; messages</span><br><span class="line">      --cache-file=FILE   cache test results in FILE [disabled]</span><br><span class="line">  -C, --config-cache      alias for `--cache-file=config.cache&apos;</span><br><span class="line">  -n, --no-create         do not create output files</span><br><span class="line">      --srcdir=DIR        find the sources in DIR [configure dir or `..&apos;]</span><br><span class="line"></span><br><span class="line">Installation directories:</span><br><span class="line">  --prefix=PREFIX         install architecture-independent files in PREFIX      ###  安装路径</span><br><span class="line">                          [/usr/local]</span><br><span class="line">  --exec-prefix=EPREFIX   install architecture-dependent files in EPREFIX</span><br><span class="line">                          [PREFIX]</span><br><span class="line"></span><br><span class="line">By default, `make install&apos; will install all the files in</span><br><span class="line">`/usr/local/bin&apos;, `/usr/local/lib&apos; etc.  You can specify</span><br><span class="line">an installation prefix other than `/usr/local&apos; using `--prefix&apos;,</span><br><span class="line">for instance `--prefix=$HOME&apos;.</span><br><span class="line"></span><br><span class="line">For better control, use the options below.</span><br><span class="line"></span><br><span class="line">Fine tuning of the installation directories:</span><br><span class="line">  --bindir=DIR            user executables [EPREFIX/bin]</span><br><span class="line">  --sbindir=DIR           system admin executables [EPREFIX/sbin]</span><br><span class="line">  --libexecdir=DIR        program executables [EPREFIX/libexec]</span><br><span class="line">  --sysconfdir=DIR        read-only single-machine data [PREFIX/etc]</span><br><span class="line">  --sharedstatedir=DIR    modifiable architecture-independent data [PREFIX/com]</span><br><span class="line">  --localstatedir=DIR     modifiable single-machine data [PREFIX/var]</span><br><span class="line">  --libdir=DIR            object code libraries [EPREFIX/lib]</span><br><span class="line">  --includedir=DIR        C header files [PREFIX/include]</span><br><span class="line">  --oldincludedir=DIR     C header files for non-gcc [/usr/include]</span><br><span class="line">  --datarootdir=DIR       read-only arch.-independent data root [PREFIX/share]</span><br><span class="line">  --datadir=DIR           read-only architecture-independent data [DATAROOTDIR]</span><br><span class="line">  --infodir=DIR           info documentation [DATAROOTDIR/info]</span><br><span class="line">  --localedir=DIR         locale-dependent data [DATAROOTDIR/locale]</span><br><span class="line">  --mandir=DIR            man documentation [DATAROOTDIR/man]</span><br><span class="line">  --docdir=DIR            documentation root [DATAROOTDIR/doc/python]</span><br><span class="line">  --htmldir=DIR           html documentation [DOCDIR]</span><br><span class="line">  --dvidir=DIR            dvi documentation [DOCDIR]</span><br><span class="line">  --pdfdir=DIR            pdf documentation [DOCDIR]</span><br><span class="line">  --psdir=DIR             ps documentation [DOCDIR]</span><br><span class="line"></span><br><span class="line">System types:</span><br><span class="line">  --build=BUILD     configure for building on BUILD [guessed]</span><br><span class="line">  --host=HOST       cross-compile to build programs to run on HOST [BUILD]</span><br><span class="line">  --target=TARGET   configure for building compilers for TARGET [HOST]</span><br><span class="line"></span><br><span class="line">Optional Features:</span><br><span class="line">  --disable-option-checking  ignore unrecognized --enable/--with options</span><br><span class="line">  --disable-FEATURE       do not include FEATURE (same as --enable-FEATURE=no)</span><br><span class="line">  --enable-FEATURE[=ARG]  include FEATURE [ARG=yes]</span><br><span class="line">  --enable-universalsdk[=SDKDIR]</span><br><span class="line">                          Build fat binary against Mac OS X SDK</span><br><span class="line">  --enable-framework[=INSTALLDIR]</span><br><span class="line">                          Build (MacOSX|Darwin) framework</span><br><span class="line">  --enable-shared         disable/enable building shared python library</span><br><span class="line">  --enable-profiling      enable C-level code profiling</span><br><span class="line">  --enable-optimizations  Enable expensive, stable optimizations (PGO, etc).</span><br><span class="line">                          Disabled by default.</span><br><span class="line">  --enable-loadable-sqlite-extensions</span><br><span class="line">                          support loadable extensions in _sqlite module</span><br><span class="line">  --enable-ipv6           Enable ipv6 (with ipv4) support</span><br><span class="line">  --disable-ipv6          Disable ipv6 support</span><br><span class="line">  --enable-big-digits[=BITS]</span><br><span class="line">                          use big digits for Python longs [[BITS=30]]</span><br><span class="line"></span><br><span class="line">Optional Packages:</span><br><span class="line">  --with-PACKAGE[=ARG]    use PACKAGE [ARG=yes]</span><br><span class="line">  --without-PACKAGE       do not use PACKAGE (same as --with-PACKAGE=no)</span><br><span class="line">  --with-universal-archs=ARCH</span><br><span class="line">                          select architectures for universal build (&quot;32-bit&quot;,</span><br><span class="line">                          &quot;64-bit&quot;, &quot;3-way&quot;, &quot;intel&quot;, &quot;intel-32&quot;, or &quot;all&quot;)</span><br><span class="line">  --with-framework-name=FRAMEWORK</span><br><span class="line">                          specify an alternate name of the framework built</span><br><span class="line">                          with --enable-framework</span><br><span class="line">  --without-gcc           never use gcc</span><br><span class="line">  --with-icc              build with icc</span><br><span class="line">  --with-cxx-main=&lt;compiler&gt;</span><br><span class="line">                          compile main() and link python executable with C++</span><br><span class="line">                          compiler</span><br><span class="line">  --with-suffix=.exe      set executable suffix</span><br><span class="line">  --with-pydebug          build with Py_DEBUG defined</span><br><span class="line">  --with-assertions       build with C assertions enabled</span><br><span class="line">  --with-lto              Enable Link Time Optimization in PGO builds.</span><br><span class="line">                          Disabled by default.</span><br><span class="line">  --with-hash-algorithm=[fnv|siphash24]</span><br><span class="line">                          select hash algorithm</span><br><span class="line">  --with-address-sanitizer</span><br><span class="line">                          enable AddressSanitizer</span><br><span class="line">  --with-libs=&apos;lib1 ...&apos;  link against additional libs</span><br><span class="line">  --with-system-expat     build pyexpat module using an installed expat</span><br><span class="line">                          library</span><br><span class="line">  --with-system-ffi       build _ctypes module using an installed ffi library</span><br><span class="line">  --with-system-libmpdec  build _decimal module using an installed libmpdec</span><br><span class="line">                          library</span><br><span class="line">  --with-tcltk-includes=&apos;-I...&apos;</span><br><span class="line">                          override search for Tcl and Tk include files</span><br><span class="line">  --with-tcltk-libs=&apos;-L...&apos;</span><br><span class="line">                          override search for Tcl and Tk libs</span><br><span class="line">  --with-dbmliborder=db1:db2:...</span><br><span class="line">                          order to check db backends for dbm. Valid value is a</span><br><span class="line">                          colon separated string with the backend names</span><br><span class="line">                          `ndbm&apos;, `gdbm&apos; and `bdb&apos;.</span><br><span class="line">  --with(out)-threads[=DIRECTORY]</span><br><span class="line">                          disable/enable thread support</span><br><span class="line">  --with(out)-thread[=DIRECTORY]</span><br><span class="line">                          deprecated; use --with(out)-threads</span><br><span class="line">  --with(out)-doc-strings disable/enable documentation strings</span><br><span class="line">  --with(out)-pymalloc    disable/enable specialized mallocs</span><br><span class="line">  --with-valgrind         Enable Valgrind support</span><br><span class="line">  --with(out)-dtrace      disable/enable DTrace support</span><br><span class="line">  --with-fpectl           enable SIGFPE catching</span><br><span class="line">  --with-libm=STRING      math library</span><br><span class="line">  --with-libc=STRING      C library</span><br><span class="line">  --with(out)-computed-gotos</span><br><span class="line">                          Use computed gotos in evaluation loop (enabled by</span><br><span class="line">                          default on supported compilers)</span><br><span class="line">  --with(out)-ensurepip=[=upgrade]</span><br><span class="line">                          &quot;install&quot; or &quot;upgrade&quot; using bundled pip</span><br><span class="line"></span><br><span class="line">Some influential environment variables:</span><br><span class="line">  MACHDEP     name for machine-dependent library files</span><br><span class="line">  CC          C compiler command</span><br><span class="line">  CFLAGS      C compiler flags</span><br><span class="line">  LDFLAGS     linker flags, e.g. -L&lt;lib dir&gt; if you have libraries in a</span><br><span class="line">              nonstandard directory &lt;lib dir&gt;</span><br><span class="line">  LIBS        libraries to pass to the linker, e.g. -l&lt;library&gt;</span><br><span class="line">  CPPFLAGS    (Objective) C/C++ preprocessor flags, e.g. -I&lt;include dir&gt; if</span><br><span class="line">              you have headers in a nonstandard directory &lt;include dir&gt;</span><br><span class="line">  CPP         C preprocessor</span><br><span class="line">  PKG_CONFIG  path to pkg-config utility</span><br><span class="line">  PKG_CONFIG_PATH</span><br><span class="line">              directories to add to pkg-config&apos;s search path</span><br><span class="line">  PKG_CONFIG_LIBDIR</span><br><span class="line">              path overriding pkg-config&apos;s built-in search path</span><br><span class="line"></span><br><span class="line">Use these variables to override the choices made by `configure&apos; or to help</span><br><span class="line">it to find libraries and programs with nonstandard names/locations.</span><br><span class="line"></span><br><span class="line">Report bugs to &lt;https://bugs.python.org/&gt;.</span><br></pre></td></tr></table></figure>
</blockquote>
</li>
<li><p>make &amp;&amp; make install </p>
</li>
<li>安装模块 pip3 install modulename</li>
</ol>
<p><strong>注意-h 下有响应时间等其他参数 按需设置,每个参数具体意义可自行搜索</strong></p>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2018/02/13/(转载)linux下R安装(无root权限)/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Naruto">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/137356299.jpg">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="木叶村">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/02/13/(转载)linux下R安装(无root权限)/" itemprop="url">(转载)linux下R安装(无root权限)</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-02-13T00:00:00+08:00">
                2018-02-13
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/R/" itemprop="url" rel="index">
                    <span itemprop="name">R</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h5 id="R依赖库的安装"><a href="#R依赖库的安装" class="headerlink" title="R依赖库的安装"></a>R依赖库的安装</h5><p>在安装R之前首先要安装R依赖的包，如何知道R安装需要什么包呢？可以去看R的官方文档<a href="https://cran.r-project.org/doc/manuals/r-release/R-admin.html#Essential-and-useful-other-programs-under-a-Unix_002dalike" target="_blank" rel="noopener">R Installation and Administration</a> ,我们主要要安装的依赖包是zlib，bzip，liblzma(这个包如果直接搜索liblzma出来的会是XZ，没错，需要安装的就是XZ)以及PCRE。不过首先需要注意的是，我安装的R版本是R-3.4.2 Short Summer，2017-09-28发行，而依照的官方文档也是Version 3.4.2 2017-09-28，因此其他版本的不一定完全适用，但是一般来说R所依赖的包大致都是这些，可能有版本的区别，安装R的大致顺序也差不多。其次，在安装这些库时最好遵照我下面给出的顺序来依次进行。否则就会报错。</p>
<ol>
<li><p>安装zlib1.2.5版本及以上</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">cd zlib-1.2.11</span><br><span class="line">./configure --prefix =/bulider/software/zlib</span><br><span class="line">make  </span><br><span class="line">make install</span><br></pre></td></tr></table></figure>
</li>
<li><p>安装bzip</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">cd bzip2-1.0.6</span><br><span class="line">make -f Makefile-libbz2_so</span><br><span class="line">#修改Makefile 中的PREFIX=/bulider/software/bzip2</span><br><span class="line">make &amp;&amp; make install</span><br></pre></td></tr></table></figure>
</li>
<li><p>安装liblzma5.0.3版本及以上</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">cd xz-5.2.3</span><br><span class="line">./configure -prefix=/bulider/software/xz</span><br><span class="line">make</span><br><span class="line">make install</span><br></pre></td></tr></table></figure>
</li>
<li><p>安装pcre</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">tar -zxvf pcre-8.41.tar.gz</span><br><span class="line">cd pcre-8.41</span><br><span class="line">./configure --enable-utf --enable-unicode-properties --enable-jit --disable-cpp --prefix /bulider/software/pcre</span><br><span class="line">##这个包的安装参数是有要求的，因此用的是官方文档的推荐命令</span><br><span class="line">make</span><br><span class="line">make install</span><br></pre></td></tr></table></figure>
</li>
<li><p>安装libcurl7.22.0及以上，但是不要超过版本8</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">tar -zxf curl-7.28.0.tar.gz</span><br><span class="line">cd curl-7.28.0</span><br><span class="line">./configure --prefix=/bulider/software/curl</span><br><span class="line">make</span><br><span class="line">make install</span><br></pre></td></tr></table></figure>
</li>
</ol>
<h5 id="安装R"><a href="#安装R" class="headerlink" title="安装R"></a>安装R</h5><ol>
<li>进行configure<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">/source/R-3.4.2/configure --prefix=/builder/software/R --en</span><br><span class="line">able-R-shlib --with-cairo  --with-jpeglib --with-readline --with-tcltk  --with-blas --with-lapack --enable-R-profiling LDFLAGS=&quot;-L/builder/software/zlib/lib  -L/builder/software/bzip2/lib -L/builder/software/xz/lib -L/builder/software/pcre/lib -L/builder/so</span><br><span class="line">ftware/curl/lib&quot; CPPFLAGS=&quot;-I/builder/zlib/include -I/builder/softw</span><br><span class="line">are/bzip2/include -I/builder/software/xz/include -I/builder/software/pcre/in</span><br><span class="line">clude -I/builder/software/curl/include&quot;##第一步检查环境配置，生成Makefile</span><br></pre></td></tr></table></figure>
</li>
</ol>
<p>首先注意一点，如果你要把R安装到另一个文件夹里而不是源码所在的文件夹里，那么你首先要创建一个文件夹，然后在这个文件夹里对R进行编译，在<code>configure</code>时要加上全路径<code>/source/R-3.4.2/configure</code>，同时在后面加上<code>--prefix=</code>参数，比方我要安装到<code>/builder/software/R</code>,那么就需要加入<code>--prefix=/builder/software/R</code>这个参数，而我的源码在<code>/source/R-3.4.2</code>这个文件夹里。</p>
<p>其次由于R是用C语言写的，因此要从源码编译R需要把R的源码需要通过gcc将源码编译为可执行文件。在前面我们安装了一些R的依赖库，这些依赖库是在编译R的时候要用到的，如果在一般情况下这些依赖库都会放到系统指定的位置来进行使用，但是由于我没有root权限，因此无法将这些库文件放到系统指定的库路径或者修改相应的配置文件，只能把这些库的路径通过<code>LDFLAGS</code>和<code>CPPFLAGS</code>这两个参数来传递给gcc，关于这两个参数可以参考。告诉gcc这些路径里有相应的库。其中<code>LDFLAGS</code>传递的是给链接器的参数，<code>CPPFLAGS</code>传递的是相应的头文件。这两个参数都是<code>configure</code>的参数。</p>
<ol>
<li>通过上一步生成了<code>Makefile</code>后，觉得胜利已经在前方向你招手，于是你进行了<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">make</span><br></pre></td></tr></table></figure>
</li>
</ol>
<p>但是出现了报错！<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br></pre></td><td class="code"><pre><span class="line">sr/bin/ld: warning: libpcre.so.1, needed by ../../lib/libR.so, not found (try using -rpath or -rpath-link)</span><br><span class="line">/usr/bin/ld: warning: liblzma.so.5, needed by ../../lib/libR.so, not found (try using -rpath or -rpath-link)</span><br><span class="line">../../lib/libR.so: undefined reference to `pcre_fullinfo&apos;</span><br><span class="line">../../lib/libR.so: undefined reference to `lzma_lzma_preset@XZ_5.0&apos;</span><br><span class="line">../../lib/libR.so: undefined reference to `lzma_alone_decoder@XZ_5.0&apos;</span><br><span class="line">../../lib/libR.so: undefined reference to `lzma_crc64@XZ_5.0&apos;</span><br><span class="line">../../lib/libR.so: undefined reference to `lzma_raw_encoder@XZ_5.0&apos;</span><br><span class="line">../../lib/libR.so: undefined reference to `pcre_config&apos;</span><br><span class="line">../../lib/libR.so: undefined reference to `lzma_code@XZ_5.0&apos;</span><br><span class="line">../../lib/libR.so: undefined reference to `lzma_stream_decoder@XZ_5.0&apos;</span><br><span class="line">../../lib/libR.so: undefined reference to `pcre_free&apos;</span><br><span class="line">../../lib/libR.so: undefined reference to `lzma_raw_decoder@XZ_5.0&apos;</span><br><span class="line">../../lib/libR.so: undefined reference to `pcre_free_study&apos;</span><br><span class="line">../../lib/libR.so: undefined reference to `pcre_assign_jit_stack&apos;</span><br><span class="line">../../lib/libR.so: undefined reference to `pcre_exec&apos;</span><br><span class="line">../../lib/libR.so: undefined reference to `lzma_version_string@XZ_5.0&apos;</span><br><span class="line">../../lib/libR.so: undefined reference to `pcre_maketables&apos;</span><br><span class="line">../../lib/libR.so: undefined reference to `lzma_stream_encoder@XZ_5.0&apos;</span><br><span class="line">../../lib/libR.so: undefined reference to `pcre_compile&apos;</span><br><span class="line">../../lib/libR.so: undefined reference to `pcre_study&apos;</span><br><span class="line">../../lib/libR.so: undefined reference to `pcre_version&apos;</span><br><span class="line">../../lib/libR.so: undefined reference to `lzma_end@XZ_5.0&apos;</span><br><span class="line">../../lib/libR.so: undefined reference to `pcre_jit_stack_alloc&apos;</span><br><span class="line">collect2: error: ld returned 1 exit status</span><br><span class="line">make[3]: *** [R.bin] Error 1</span><br><span class="line">make[3]: Leaving directory `/builder/software/R/src/main&apos;</span><br><span class="line">make[2]: *** [R] Error 2</span><br><span class="line">make[2]: Leaving directory `/builder/software/R/src/main&apos;</span><br><span class="line">make[1]: *** [R] Error 1</span><br><span class="line">make[1]: Leaving directory `/builder/software/R/src&apos;</span><br><span class="line">make: *** [R] Error 1</span><br></pre></td></tr></table></figure></p>
<p>这个报错是因为什么原因导致的？我们可以看到里面说<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">usr/bin/ld: warning: libpcre.so.1, needed by ../../lib/libR.so, not found (try using -rpath or -rpath-link)</span><br><span class="line">/usr/bin/ld: warning: liblzma.so.5, needed by ../../lib/libR.so, not found (try using -rpath or -rpath-link)</span><br></pre></td></tr></table></figure></p>
<p>说缺少<code>libpcre.so.1</code>和<code>liblzma.so.5</code>这两个库文件，我心里浮现了一副黑人问号脸我在前面<code>./configure</code>的时候还传入了<code>PCRE</code>和<code>XZ</code>这两个库的路径，为什么还是会报错？<br>于是我先是进行了谷歌，从谷歌上找到了一篇R安装攻略,在这篇攻略里也提到了这个问题，</p>
<blockquote>
<p>此处报错是由于没有找到动态库，看第一二行</p>
<p>解决方法：添加动态库</p>
<p>cat /etc/ld.so.conf</p>
<p>include ld.so.conf.d/*.conf</p>
<p>/opt/pcre-8.39/lib</p>
<p>/opt/xz-5.2.2/lib</p>
</blockquote>
<p>但是在这个攻略里最后的解决方法是修改<code>/etc/ld.so.conf</code>文件，将前面安装的库的路径添加到这个文件中，但是我没有root权限，并不能修改这个文件，因此这个方法对我并不适用。后来在鹏帆和许骄的提醒下，我们在后面看到建议说try using <code>-rpath or -rpath-link</code>,后来我搜索了这两个命令，这两个命令是用于通过gcc向ld传递参数的时候用的，</p>
<blockquote>
<p>gcc编译链接动态库时，很有可能编译通过，但是执行时，找不到动态链接库，那是因为-L选项指定的路径只在编译时有效，解决方法是通过-Wl(注意，是字母表K之后的小写L),<code>rpath=&lt;your_lib_dir&gt;</code>，使得execute记住链接库的位置</p>
</blockquote>
<p>于是我试着在configure的时候加入这两个参数<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">/builder/software/R-3.4.2/configure --prefix=/builder/software/R --en</span><br><span class="line">able-R-shlib --with-cairo  --with-jpeglib --with-readline --with-tcltk  --with-blas --with-lapack --enable-R-profiling LDFLAGS=&quot;-L/builder/software/zlib/lib  -L/builder/software/bzip2/lib -L/builder/software/xz/lib  -Wl,--rpath/builder/software/pcre/lib -Wl,--rpath/builder/so</span><br><span class="line">ftware/curl/lib&quot; CPPFLAGS=&quot;-I/builder/software/zlib/include -I/builder/softw</span><br><span class="line">are/bzip2/include -I/builder/software/xz/include -I/builder/software/pcre/in</span><br><span class="line">clude -I/builder/software/curl/include</span><br></pre></td></tr></table></figure></p>
<p>但是这样总会报错，后来我知道，configure的时候我们传入的参数都会写到Makeconf文件里面，那么我可不可以先按照原来的我方法先进行configure，然后再对Makeconf文件进行修改呢？于是我还是先按照之前的方法进行configure<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">/builder/software/R-3.4.2/configure --prefix=/builder/software/R --en</span><br><span class="line">able-R-shlib --with-cairo  --with-jpeglib --with-readline --with-tcltk  --with-blas --with-lapack --enable-R-profiling LDFLAGS=&quot;-L/builder/software/zlib/lib  -L/builder/software/bzip2/lib -L/builder/software/xz/lib -L/builder/software/pcre/lib -L/builder/so</span><br><span class="line">ftware/curl/lib&quot; CPPFLAGS=&quot;-I/builder/software/zlib/include -I/builder/softw</span><br><span class="line">are/bzip2/include -I/builder/software/xz/include -I/builder/software/pcre/in</span><br><span class="line">clude -I/builder/software/curl/include</span><br></pre></td></tr></table></figure></p>
<p>然后在make之前修改Makeconf中的参数<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">LDFLAGS = -L/builder/software/zlib/lib  -L/builder/software/bzip2/lib   -L/builder/software/curl/lib -Wl,-rpath=/builder/software/xz/lib -Wl,-rpath=/builder/software/pcre/lib</span><br></pre></td></tr></table></figure></p>
<p>再进行<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">make</span><br></pre></td></tr></table></figure></p>
<p>之前的报错终于不见了!但是出现了新的报错心里有句mmp:<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">/usr/bin/ld: cannot find -lpcre</span><br><span class="line">collect2: error: ld returned 1 exit status</span><br><span class="line">make[3]: *** [libR.so] Error 1</span><br><span class="line">make[3]: Leaving directory `/builder/software/R-3.4.2/src/main&apos;</span><br><span class="line">make[2]: *** [R] Error 2</span><br><span class="line">make[2]: Leaving directory `/builder/software/R-3.4.2/src/main&apos;</span><br><span class="line">make[1]: *** [R] Error 1</span><br><span class="line">make[1]: Leaving directory `/builder/software/R-3.4.2/src&apos;</span><br><span class="line">make: *** [R] Error 1</span><br></pre></td></tr></table></figure></p>
<p>于是我在想，可能是前面我把LDFLAGS里面参数由-L/builder/software/pcre/lib修改为了-Wl,-rpath=/builder/software/pcre/lib之后不再出现原来的报错，但是现在又出现了新的报错，而我在使用原来的参数时不会出现这个新的报错，结合之前我搜索到的结果,是不是因为我之前的写法只能让路径gcc在编译时有效,而后来的写法只能让其在进行运行时有效？那么如果我让两种参数同时出现，虽然我把库的路径传递了两次，但是会不会就不再报错了？时间是检验真理的唯一标准！先试验一把！于是我把两种参数的路径都加到了Makeconf中<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">LDFLAGS = -L/builder/software/zlib/lib  -L/builder/software/bzip2/lib -L/builder/software/xz/lib -L/builder/software/pcre/lib -L/builder/software/curl/lib -Wl,-rpath=/builder/software/xz/lib -Wl,-rpath=/builder/software/pcre/lib</span><br></pre></td></tr></table></figure></p>
<p>再次进行<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">make</span><br></pre></td></tr></table></figure></p>
<p>终于没有了报错。说明make成功了，于是我再进行<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">make install</span><br></pre></td></tr></table></figure></p>
<p>终于安装完成！你挑着担我牵着马</p>
<h5 id="总结"><a href="#总结" class="headerlink" title="总结"></a>总结</h5><p>1.在Linux下用源码安装R是个大坑尤其是在没有root权限的情况下</p>
<p>2.安装R之前一定要将R要求的依赖库预先安装好，版本和安装顺序都要对。具体可以参考R官方文档R Installation and Administration 。</p>
<p>3.在安装好R的依赖库之后，在进行R的安装的时候要注意将依赖库的路径传入到参数中，而且一定要注意传入的方式。</p>
<p>4.注意分析报错信息，从中寻找原因，不要盲目瞎想，多谷歌，多问老司机。</p>
<p>5.关于-Wl -rpath更详细的介绍<a href="https://my.oschina.net/shelllife/blog/115958" target="_blank" rel="noopener">参考</a></p>
<p>3.在安装好R的依赖库之后，在进行R的安装的时候要注意将依赖库的路径传入到参数中，而且一定要注意传入的方式。<br>4.注意分析报错信息，从中寻找原因，不要盲目瞎想，多谷歌，多问老司机。<br>5.关于-Wl -rpath更详细的介绍参考</p>
<p>参考链接：<a href="http://www.jianshu.com/p/5bfe154f1aa4" target="_blank" rel="noopener">http://www.jianshu.com/p/5bfe154f1aa4</a></p>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2018/02/13/perl 单行命令/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Naruto">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/137356299.jpg">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="木叶村">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/02/13/perl 单行命令/" itemprop="url">perl 单行命令</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-02-13T00:00:00+08:00">
                2018-02-13
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/perl/" itemprop="url" rel="index">
                    <span itemprop="name">perl</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <p><strong>与One-Liner相关的perl参数</strong></p>
<ul>
<li>-a 自动分隔模式，用空格分隔$并保存在@F中，也就是@F=split //, $</li>
<li>-F 指定-a的分隔符</li>
<li>-l 对输入的内容进行自动chomp，对输出的内容自动加换行符</li>
<li>-n 相当于while(&lt;&gt;)</li>
<li>-e 执行命令，也就是脚本</li>
<li>-p 自动循环+输出，也就是while(&lt;&gt;){命令（脚本）; print;}</li>
</ul>
<p><img src="http://files.jb51.net/file_images/article/201709/2017910200851051.png?201781020923" alt="完整版本"></p>
<ol>
<li><p>基本格式：perl -参数 ‘命令’ 输入文件</p>
</li>
<li><p><code>perl -e</code>为必须要写的参数</p>
</li>
<li><p><code>perl -pe</code> 可用于读取文件每行，并按照给定的命令进行处理，最后输出；如将文件1.txt中的aaa替换为AAA</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">perl -pe &apos;s/aaa/AAA/g&apos; 1.txt</span><br></pre></td></tr></table></figure>
</li>
<li><p><code>perl -l</code>参数几乎可以跟n搭配代替perl经常用的<code>while(&lt;&gt;){chomp;}</code>语法</p>
</li>
<li><p>如果需要处理tab分割的文件的每一行内容，那么<code>perl -alne</code>参数几乎可以说是必备的，例如:<code>while(&lt;&gt;){chomp;@F=split /\s+/,$_;print &quot;$F[0]\n&quot;}</code>相当于<code>perl -alne &#39;print $F[0]&#39;</code></p>
</li>
<li><p>perl单行命令脚本里的变量都不需要预先声明，如想打印出每空行，并且每行以行数开头</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">perl -ne &apos;print ++$a.&quot; $_&quot; if /./&apos;</span><br></pre></td></tr></table></figure>
</li>
<li><p>perl单行命令有时优于sed/grep等shell命令是由于其优秀的正则匹配，通常简单的匹配可以如：匹配上的行号，模仿grep -c的功能：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">perl -lne &apos;$a++ if /regex/; END &#123;print $a+0&#125;&apos;</span><br></pre></td></tr></table></figure>
</li>
<li><p>perl单行命令可以使用perl的模块，如使用sum函数的模块：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">perl -MList::Util=sum -alne &apos;print sum @F&apos;</span><br></pre></td></tr></table></figure>
</li>
<li><p>perl也可以像awk一样使用END命令，如打印出文件中总单词个数</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">perl -alne &apos;$t += @F; END &#123; print $t&#125;&apos;</span><br></pre></td></tr></table></figure>
</li>
<li><p>perl也可以使用<code>map{}</code>等函数，如打印出匹配上的单词的总个数</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">perl -alne &apos;map &#123; /regex/ &amp;&amp; $t++ &#125; @F; END &#123; print $t &#125;&apos;</span><br></pre></td></tr></table></figure>
</li>
<li><p>perl单行命令可以说是将perl的简洁用到了极致，如打印出匹配上的行：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">perl -ne &apos;/regex/ &amp;&amp; print&apos;</span><br></pre></td></tr></table></figure>
</li>
<li><p>perl单行命令能像perl一样灵活的使用则正表达式</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">perl -ne &apos;print if /^\d+$/&apos;</span><br></pre></td></tr></table></figure>
</li>
</ol>
<p>以上例子均出自于 <a href="http://www.catonmat.net/blog/perl-one-liners-explained-part-one/" target="_blank" rel="noopener">http://www.catonmat.net/blog/perl-one-liners-explained-part-one/</a></p>
<p>使用perl来处理数据的我们，会一点Perl one line可以有效的减少编写重复命令的时间，尤其是那些就用1-2次就不会用的脚本，尤其在window系统下不方便使用shell命令的时候。</p>
<p>PS.当然在windows系统下也可以借用git模拟Unix命令环境</p>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
  </section>

  
  <nav class="pagination">
    <a class="extend prev" rel="prev" href="/page/3/"><i class="fa fa-angle-left"></i></a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/3/">3</a><span class="page-number current">4</span><a class="page-number" href="/page/5/">5</a><a class="extend next" rel="next" href="/page/5/"><i class="fa fa-angle-right"></i></a>
  </nav>



          </div>
          


          

        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      

      <section class="site-overview sidebar-panel sidebar-panel-active">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
          <img class="site-author-image" itemprop="image"
               src="/images/137356299.jpg"
               alt="Naruto" />
          <p class="site-author-name" itemprop="name">Naruto</p>
           
              <p class="site-description motion-element" itemprop="description"></p>
          
        </div>
        <nav class="site-state motion-element">

          
            <div class="site-state-item site-state-posts">
              <a href="/archives">
                <span class="site-state-item-count">42</span>
                <span class="site-state-item-name">日志</span>
              </a>
            </div>
          

          
            
            
            <div class="site-state-item site-state-categories">
              <a href="/categories/index.html">
                <span class="site-state-item-count">8</span>
                <span class="site-state-item-name">分类</span>
              </a>
            </div>
          

          
            
            
            <div class="site-state-item site-state-tags">
              <a href="/tags/index.html">
                <span class="site-state-item-count">20</span>
                <span class="site-state-item-name">标签</span>
              </a>
            </div>
          

        </nav>

        

        <div class="links-of-author motion-element">
          
            
              <span class="links-of-author-item">
                <a href="zhmqsnq90@foxmail.com" target="_blank" title="E-Mail">
                  
                    
                      E-Mail
                    
                </a>
              </span>
            
          
        </div>

        
        

        
        

        


      </section>

      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright" >
  
  &copy; 
  <span itemprop="copyrightYear">2018</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Naruto</span>
</div>


<div class="powered-by">
  由 <a class="theme-link" href="https://hexo.io">Hexo</a> 强力驱动
</div>

<div class="theme-info">
  主题 -
  <a class="theme-link" href="https://github.com/iissnan/hexo-theme-next">
    NexT.Mist
  </a>
</div>


        

        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>

  
  <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.2"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.2"></script>



  
  

  
    <script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.2"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.1.2"></script>

  

  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.2"></script>



  


  




	





  





  






  





  

  

  

  
  
    <script type="text/x-mathjax-config">
      MathJax.Hub.Config({
        tex2jax: {
          inlineMath: [ ['$','$'], ["\\(","\\)"]  ],
          processEscapes: true,
          skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
        }
      });
    </script>

    <script type="text/x-mathjax-config">
      MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax(), i;
        for (i=0; i < all.length; i += 1) {
          all[i].SourceElement().parentNode.className += ' has-jax';
        }
      });
    </script>
    <script type="text/javascript" src="//cdn.bootcss.com/mathjax/2.7.1/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
  


  

  

</body>
</html>
