<!DOCTYPE html>



  


<html class="theme-next mist use-motion" lang="zh-Hans">
<head>
  <meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>
<meta name="theme-color" content="#222">









<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />















  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />




  
  
  
  

  
    
    
  

  

  

  

  

  
    
    
    <link href="//fonts.googleapis.com/css?family=Lato:300,300italic,400,400italic,700,700italic&subset=latin,latin-ext" rel="stylesheet" type="text/css">
  






<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.2" rel="stylesheet" type="text/css" />


  <meta name="keywords" content="生信理论基础,软件," />








  <link rel="shortcut icon" type="image/x-icon" href="/favicon.ico?v=5.1.2" />






<meta name="description" content="软件作者简介：陈实富，海普洛斯 联合创始人 / CTO 文章转载自：https://zhuanlan.zhihu.com/p/33601691 1引言各位做生信的小伙伴都知道，对于下机的FASTQ数据需要进行质控和预处理，以保证下游分析输入的数据都是干净可靠的。通常我们都是使用FASTQC等软件进行质控，使用cutadapt软件去除接头，使用Trimmomatic等软件进行剪裁，然后使用一些自已开">
<meta name="keywords" content="生信理论基础,软件">
<meta property="og:type" content="article">
<meta property="og:title" content="(转)fastp： 一款超快速全功能的FASTQ文件自动化质控+过滤+校正+预处理软件">
<meta property="og:url" content="http://yoursite.com/2018/11/15/fastp/index.html">
<meta property="og:site_name" content="木叶村">
<meta property="og:description" content="软件作者简介：陈实富，海普洛斯 联合创始人 / CTO 文章转载自：https://zhuanlan.zhihu.com/p/33601691 1引言各位做生信的小伙伴都知道，对于下机的FASTQ数据需要进行质控和预处理，以保证下游分析输入的数据都是干净可靠的。通常我们都是使用FASTQC等软件进行质控，使用cutadapt软件去除接头，使用Trimmomatic等软件进行剪裁，然后使用一些自已开">
<meta property="og:locale" content="zh-Hans">
<meta property="og:image" content="https://pic2.zhimg.com/80/v2-6920178b4b210ff1e37089ff8e1b289d_hd.jpg">
<meta property="og:image" content="https://pic1.zhimg.com/80/v2-202f214ad93c99a305f6b3510ddf3f78_hd.jpg">
<meta property="og:image" content="https://pic4.zhimg.com/80/v2-78e31fe48bbdc53f5c1a1427b6170a53_hd.jpg">
<meta property="og:image" content="https://pic3.zhimg.com/80/v2-ee5358c3a2a93af49171459278aa751e_hd.jpg">
<meta property="og:image" content="https://pic3.zhimg.com/80/v2-ee5358c3a2a93af49171459278aa751e_hd.jpg">
<meta property="og:image" content="https://pic2.zhimg.com/80/v2-c2ca5f1b36bf7c9010d43773dffaf66d_hd.jpg">
<meta property="og:image" content="https://pic2.zhimg.com/80/v2-4d30aaf1dfe395d89d36d9ae5c3e135d_hd.jpg">
<meta property="og:image" content="https://pic3.zhimg.com/80/v2-a69e014cd516d44ac2bb0d3cb8d6d67e_hd.jpg">
<meta property="og:updated_time" content="2018-11-15T13:32:37.518Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="(转)fastp： 一款超快速全功能的FASTQ文件自动化质控+过滤+校正+预处理软件">
<meta name="twitter:description" content="软件作者简介：陈实富，海普洛斯 联合创始人 / CTO 文章转载自：https://zhuanlan.zhihu.com/p/33601691 1引言各位做生信的小伙伴都知道，对于下机的FASTQ数据需要进行质控和预处理，以保证下游分析输入的数据都是干净可靠的。通常我们都是使用FASTQC等软件进行质控，使用cutadapt软件去除接头，使用Trimmomatic等软件进行剪裁，然后使用一些自已开">
<meta name="twitter:image" content="https://pic2.zhimg.com/80/v2-6920178b4b210ff1e37089ff8e1b289d_hd.jpg">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Mist',
    sidebar: {"position":"left","display":"post","offset":12,"offset_float":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: true,
    duoshuo: {
      userId: '0',
      author: '博主'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://yoursite.com/2018/11/15/fastp/"/>





  <title>(转)fastp： 一款超快速全功能的FASTQ文件自动化质控+过滤+校正+预处理软件 | 木叶村</title>
  














</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-Hans">

  
  
    
  

  <div class="container sidebar-position-left page-post-detail ">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">木叶村</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle"></p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            首页
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            分类
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            归档
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            标签
          </a>
        </li>
      

      
    </ul>
  

  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2018/11/15/fastp/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Naruto">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/137356299.jpg">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="木叶村">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">(转)fastp： 一款超快速全功能的FASTQ文件自动化质控+过滤+校正+预处理软件</h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-11-15T00:00:00+08:00">
                2018-11-15
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生信/" itemprop="url" rel="index">
                    <span itemprop="name">生信</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <p>软件作者简介：陈实富，海普洛斯 联合创始人 / CTO</p>
<p>文章转载自：<a href="https://zhuanlan.zhihu.com/p/33601691" target="_blank" rel="noopener">https://zhuanlan.zhihu.com/p/33601691</a></p>
<h2 id="1引言"><a href="#1引言" class="headerlink" title="1引言"></a>1引言</h2><p>各位做生信的小伙伴都知道，对于下机的FASTQ数据需要进行质控和预处理，以保证下游分析输入的数据都是干净可靠的。通常我们都是使用FASTQC等软件进行质控，使用cutadapt软件去除接头，使用Trimmomatic等软件进行剪裁，然后使用一些自已开发的脚本进行过滤。这一过程可能需要使用多个软件，相当繁琐，而且速度较慢，这些软件大多又不支持多线程，遇到较大的FASTQ文件，处理起来可真是让人等得心急如焚。</p>
<p>所以今天给大家介绍一款新的软件，名字叫fastp，它可以仅仅扫描FASTQ文件一次，就完成比FASTQC+ cutadapt + Trimmomatic 这三个软件加起来的功能还多很多的功能，而且速度上比仅仅使用Trimmomatic一个软件还要快3倍左右，因为它使用C++开发，处处使用了高效算法，而且完美支持多线程！<br>github地址请戳：<a href="https://github.com/OpenGene/fastp" target="_blank" rel="noopener">https://github.com/OpenGene/fastp</a></p>
<h2 id="2功能特点"><a href="#2功能特点" class="headerlink" title="2功能特点"></a>2功能特点</h2><p>该软件的功能列表：</p>
<ul>
<li>对数据自动进行全方位质控，生成人性化的报告。</li>
<li>过滤功能（低质量，太短，太多N……）。</li>
<li>对每一个序列的头部或尾部，计算滑动窗内的质量均值，并将均值较低的子序列进行切除（类似Trimmomatic的做法，但是快非常多）。</li>
<li>全局剪裁 （在头/尾部，不影响去重），对于Illumina下机数据往往最后一到两个cycle需要这样处理。</li>
<li>去除接头污染。厉害的是，你不用输入接头序列，因为算法会自动识别接头序列并进行剪裁。</li>
<li>对于双端测序（PE）的数据，软件会自动查找每一对read的重叠区域，并对该重叠区域中不匹配的碱基对进行校正。</li>
<li>去除尾部的polyG。对于Illumina NextSeq/NovaSeq的测序数据，因为是两色法发光，polyG是常有的事，所以该特性对该两类测序平台默认打开。</li>
<li>对于PE数据中的overlap区间中不一致的碱基对，依据质量值进行校正</li>
<li>可以对带分子标签（UMI）的数据进行预处理，不管UMI在插入片段还是在index上，都可以轻松处理。</li>
<li>可以将输出进行分拆，而且支持两种模式，分别是指定分拆的个数，或者分拆后每个文件的行数。</li>
</ul>
<h2 id="3轻松上手"><a href="#3轻松上手" class="headerlink" title="3轻松上手"></a>3轻松上手</h2><p>SE:</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">fastp -i in.fq -o out.fq</span><br></pre></td></tr></table></figure>
<p>PE:</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">fastp -i in.R1.fq -o out.R1.fq -I in.R2.fq -O out.R2.fq</span><br><span class="line">fastp-i in.R1.fq.gz -I in.R2.fq.gz -o out.R1.fq.gz -O out.R2.fq.gz</span><br></pre></td></tr></table></figure>
<h2 id="4安装fastp"><a href="#4安装fastp" class="headerlink" title="4安装fastp"></a>4安装fastp</h2><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">git clone https://github.com/OpenGene/fastp.git</span><br><span class="line">cd fastp</span><br><span class="line">make</span><br><span class="line">sudo make install</span><br></pre></td></tr></table></figure>
<h2 id="5功能介绍"><a href="#5功能介绍" class="headerlink" title="5功能介绍"></a>5功能介绍</h2><h3 id="5-1过滤"><a href="#5-1过滤" class="headerlink" title="5.1过滤"></a>5.1过滤</h3><p>fastp可以对低质量序列，较多N的序列，该功能默认是启用的，但可以使用<code>-Q</code>参数关闭。使用<code>-q</code>参数来指定合格的phred质量值，比如<code>-q 15</code>表示质量值大于等于Q15的即为合格，然后使用<code>-u</code>参数来指定最多可以有多少百分比的质量不合格碱基。比如<code>-q 15 -u 40</code>表示一个read最多只能有40%的碱基的质量值低于Q15，否则会被扔掉。使用<code>-n</code>可以限定一个read中最多能有多少个N。</p>
<p>fastp还默认启用了read长度过滤，但可以使用<code>-L</code>参数关闭。使用<code>-l</code>参数指定最低要求一个read有多长，比如<code>-l 30</code>表示低于30个碱基的read会被扔掉。这个功能可以用于实现常用的discard模式，以保证所有输出的序列都一样长。<br>在fastp的HTML报告中，最头上的Summary表格很清楚地显示了过滤的统计信息，如下图所示：<br><img src="https://pic2.zhimg.com/80/v2-6920178b4b210ff1e37089ff8e1b289d_hd.jpg" alt="5.1"></p>
<h3 id="5-2接头处理"><a href="#5-2接头处理" class="headerlink" title="5.2接头处理"></a>5.2接头处理</h3><p>接头（adapter）污染的处理是FASTQ文件预处理中很重要的一步。fastp默认启用了接头处理，但是可以使用<code>-A</code>命令来关掉。fastp可以自动化地查找接头序列并进行剪裁，也就是说你可以不输入任何的接头序列，fastp全自动搞定了！对于SE数据，你还是可以<code>-a</code>参数来输入你的接头，而对于PE数据则完全没有必要，fastp基于PE数据的overlap分析可以更准确地查找接头，去得更干净，而且对于一些接头本身就有碱基不匹配情况处理得更好。fastp对于接头去除会有一个汇总的报告，如下图所示：<br><img src="https://pic1.zhimg.com/80/v2-202f214ad93c99a305f6b3510ddf3f78_hd.jpg" alt="5.2"></p>
<h3 id="5-3滑窗质量剪裁"><a href="#5-3滑窗质量剪裁" class="headerlink" title="5.3滑窗质量剪裁"></a>5.3滑窗质量剪裁</h3><p>很多时候，一个read的低质量序列都是集中在read的末端，也有少部分是在read的开头。fastp支持像Trimmomatic那样对滑动窗口中的碱基计算平均质量值，然后将不符合的滑窗直接剪裁掉。使用<code>-5</code>参数开启在5’端，也就是read的开头的剪裁，使用<code>-3</code>参数开启在3’端，也就是read的末尾的剪裁。使用<code>-W</code>参数指定滑动窗大小，默认是4，使用<code>-M</code>参数指定要求的平均质量值，默认是20，也就是Q20。</p>
<h3 id="5-4-PE数据的碱基校正"><a href="#5-4-PE数据的碱基校正" class="headerlink" title="5.4 PE数据的碱基校正"></a>5.4 PE数据的碱基校正</h3><p>fastp支持对PE数据的每一对read进行分析，查找它们的overlap区间，然后对于overlap区间中不一致的碱基，如果发现其中一个质量非常高，而另一个非常低，则可以将非常低质量的碱基改为相应的非常高质量值的碱基值，如下图所示：<br><img src="https://pic4.zhimg.com/80/v2-78e31fe48bbdc53f5c1a1427b6170a53_hd.jpg" alt="5.4"></p>
<p>上图中所示的标红的T碱基是低质量序列，和高质量的A不匹配，它会被校正为A。该校正功能默认没有开启使用<code>-c</code>参数可以启用，对于一些对噪声容忍度低的应用，比如液体活检，建议开启。</p>
<h3 id="5-5全局剪裁"><a href="#5-5全局剪裁" class="headerlink" title="5.5全局剪裁"></a>5.5全局剪裁</h3><p>fastp可以对所有read在头部和尾部进行统一剪裁，该功能在去除一些测序质量不好的cycle比较有用，比如151*2的PE测序中，最后一个cycle通常质量是非常低的，需要剪裁掉。使用<code>-f</code>和<code>-t</code>分别指定read1的头部和尾部的剪裁，使用<code>-F</code>和<code>-T</code>分别指定read2的头部和尾部的剪裁。</p>
<h3 id="5-6-polyG剪裁"><a href="#5-6-polyG剪裁" class="headerlink" title="5.6 polyG剪裁"></a>5.6 polyG剪裁</h3><p>对于两色发光法的Illumina设备（NextSeq /NovaSeq），因为在没有光信号情况下base calling的结果会返回G，所以在序列的尾端可能会出现较多的polyG，需要被去除。fastp会自动化地识别NextSeq / NovaSeq的数据，然后进行polyG识别和剪裁。如果你想强制开启该功能，可以指定<code>-g</code>参数，如果想强制关闭该功能，则可以指定<code>-G</code>参数。</p>
<h3 id="5-7分子标签UMI处理"><a href="#5-7分子标签UMI处理" class="headerlink" title="5.7分子标签UMI处理"></a>5.7分子标签UMI处理</h3><p>UMI在处理ctDNA类似的超低频突变检测应用中是十分有用的，为了更好地对带UMI的FASTQ文件进行预处理，fastp也很好地支持了UMI预处理功能。该功能默认没有启用，需要使用<code>-U</code>参数开启，另外需要使用<code>--umi_loc</code>来指定UMI所在的位置，它可以是（index1、 index2、 read1、 read2、 per_index、 per_read ）中的一种，分别表示UMI是在index位置上，还是在插入片段中。如果指定了是在插入序列中，还需要使用 <code>--umi_len</code> 参数来指定UMI所占的碱基长度。</p>
<h3 id="5-8输出文件切分"><a href="#5-8输出文件切分" class="headerlink" title="5.8输出文件切分"></a>5.8输出文件切分</h3><p>很多时候我们需要对输出的FASTQ进行切分，分成大小均匀的多个文件，这样可以使用比对软件并行地比对，提高并行处理的速度。fastp软件也提供了相应的功能，并且支持了两种模式，分别是使用参数<code>-s</code>指定切分后文件的个数，或者使用<code>-S</code>参数指定每个切分后文件的行数。</p>
<h2 id="6质控报告解读"><a href="#6质控报告解读" class="headerlink" title="6质控报告解读"></a>6质控报告解读</h2><p>接下来，我们再看一下如何理解fastp生成的质控报告。fastp的报告在单一文件中同时包含了过滤前和过滤后的统计结果，如果是PE数据，则同时包含了read1和read2的统计结果。之前我们已经说过了，fastp会生成HTML的报告和JSON格式的报告。HTML报告的默认文件名是fastp.html，但是可以通过-h参数修改，JSON报告的默认文件名是fastp.json，但是可以通过-j参数修改。而且fastp报告还有一个标题，默认是fastp report，这个也可以通过-R参数修改为你想要的标题。JSON格式的报告是优化过的，人机皆可读，适合进阶的用户使用程序解析，而这里我们重点关注HTML格式的报告。</p>
<h3 id="6-1质量分布曲线图"><a href="#6-1质量分布曲线图" class="headerlink" title="6.1质量分布曲线图"></a>6.1质量分布曲线图</h3><p>我们第一关注的当然是质量，所以fastp提供了质量分布曲线，即每一个cycle的平均质量值，而且fastp同时提供了A/T/C/G四种不同碱基的平均质量，以及总的平均质量，如下图所示：<br><img src="https://pic3.zhimg.com/80/v2-ee5358c3a2a93af49171459278aa751e_hd.jpg" alt="6.1"><br><img src="https://pic3.zhimg.com/80/v2-ee5358c3a2a93af49171459278aa751e_hd.jpg" alt="6.1_1"></p>
<p>从上图我们可以看到，一共有5条曲线，分别是A/T/C/G和mean。而且HTML报告中的每一个项目和分项目都是可以点击进行隐藏和展开的。</p>
<h3 id="6-2碱基含量分布曲线"><a href="#6-2碱基含量分布曲线" class="headerlink" title="6.2碱基含量分布曲线"></a>6.2碱基含量分布曲线</h3><p>和质量分布曲线类似，碱基含量分布曲线也是按照每一个cycle来的，显示了每一个位置的碱基含量。如下图所示：<br><img src="https://pic2.zhimg.com/80/v2-c2ca5f1b36bf7c9010d43773dffaf66d_hd.jpg" alt="6.2"><br><img src="https://pic2.zhimg.com/80/v2-4d30aaf1dfe395d89d36d9ae5c3e135d_hd.jpg" alt="6.2_1"></p>
<p>从图中可以看到，fastp同时显示了A/T/C/G/N/GC的每一个位置的比例和总的比例。而且如果你觉得头部那里比较乱看不清的话，可以用鼠标拉一个框，它就放大了</p>
<h3 id="6-3-KMER统计表格"><a href="#6-3-KMER统计表格" class="headerlink" title="6.3 KMER统计表格"></a>6.3 KMER统计表格</h3><p>fastp对5个碱基长度的所有组合的出现次数进行了统计，然后把它放在了一张表格中，表格的每一个元素为深背景白字，背景越深，则表示重复次数越多。这样，一眼望去，就可以发现有哪一些异常的信息。<br><img src="https://pic3.zhimg.com/80/v2-a69e014cd516d44ac2bb0d3cb8d6d67e_hd.jpg" alt="6.3"></p>
<p>从上面的KMER表格中，我们可以发现，GGGGG的颜色特别深，从鼠标移上去之后显示的信息中我们可以发现它的出现次数是平均次数的12.8倍，这是不正常的，因为GGGGG的正常倍数应该在1倍左右。幸好我们有fastp，它可以过滤掉这种polyG，让数值较多地回归正常。</p>
<h3 id="6-4过表达序列-overrepresented-sequence"><a href="#6-4过表达序列-overrepresented-sequence" class="headerlink" title="6.4过表达序列(overrepresented sequence)"></a>6.4过表达序列(overrepresented sequence)</h3><p>fastp的最新版本（v0.12）提供了overrepresetned<br>sequence的分析，而且不但提供了这些overrepresented sequence的序列个数和占比，还提供了他们在测序cycles中的分布情况，这有利于分析各种问题</p>
<h2 id="7-完整参数"><a href="#7-完整参数" class="headerlink" title="7 完整参数"></a>7 完整参数</h2><p>使用示例：<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">fastp -i R1.fastq.gz -I R2.clean.fastq.gz -o R1.clean.fastq.gz -O R2.clean.fastq.gz -f 5 -F 5 -3 -W 8 -M 25 -q 20 -u 25 -n 3 -l 50 -c -g --poly_g_min_len 6 -x -y -Y 20 -j out.json -h out.html --adapter_sequence GATCGGAAGAGCACACGTCTGAACTCCAGTCAC --adapter_sequence_r2 AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT &gt; fastp.log</span><br></pre></td></tr></table></figure></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br></pre></td><td class="code"><pre><span class="line">$fastp</span><br><span class="line">fastp: an ultra-fast all-in-one FASTQ preprocessor</span><br><span class="line">version 0.19.3</span><br><span class="line">usage: fastp --in1=string [options] ...</span><br><span class="line">options:</span><br><span class="line">  -i, --in1                            read1 input file name (string)</span><br><span class="line">  -o, --out1                           read1 output file name (string [=])</span><br><span class="line">  -I, --in2                            read2 input file name (string [=])</span><br><span class="line">  -O, --out2                           read2 output file name (string [=])</span><br><span class="line">  -6, --phred64                        indicate the input is using phred64 scoring (it&apos;ll be converted to phred33, so the output will still be phred33)</span><br><span class="line">  -z, --compression                    compression level for gzip output (1 ~ 9). 1 is fastest, 9 is smallest, default is 4. (int [=4])</span><br><span class="line">      --stdout                         stream passing-filters reads to STDOUT. This option will result in interleaved FASTQ output for paired-end input. Disabled by defaut.</span><br><span class="line">      --interleaved_in                 indicate that &lt;in1&gt; is an interleaved FASTQ which contains both read1 and read2. Disabled by defaut.</span><br><span class="line">      --reads_to_process               specify how many reads/pairs to be processed. Default 0 means process all reads. (int [=0])</span><br><span class="line">      --dont_overwrite                 don&apos;t overwrite existing files. Overwritting is allowed by default.</span><br><span class="line">  -V, --verbose                        output verbose log information (i.e. when every 1M reads are processed).</span><br><span class="line">  -A, --disable_adapter_trimming       adapter trimming is enabled by default. If this option is specified, adapter trimming is disabled</span><br><span class="line">  -a, --adapter_sequence               the adapter for read1. For SE data, if not specified, the adapter will be auto-detected. For PE data, this is used if R1/R2 are found not overlapped. (string [=auto])</span><br><span class="line">      --adapter_sequence_r2            the adapter for read2 (PE data only). This is used if R1/R2 are found not overlapped. If not specified, it will be the same as &lt;adapter_sequence&gt; (string [=])</span><br><span class="line">  -f, --trim_front1                    trimming how many bases in front for read1, default is 0 (int [=0])</span><br><span class="line">  -t, --trim_tail1                     trimming how many bases in tail for read1, default is 0 (int [=0])</span><br><span class="line">  -F, --trim_front2                    trimming how many bases in front for read2. If it&apos;s not specified, it will follow read1&apos;s settings (int [=0])</span><br><span class="line">  -T, --trim_tail2                     trimming how many bases in tail for read2. If it&apos;s not specified, it will follow read1&apos;s settings (int [=0])</span><br><span class="line">  -g, --trim_poly_g                    force polyG tail trimming, by default trimming is automatically enabled for Illumina NextSeq/NovaSeq data</span><br><span class="line">      --poly_g_min_len                 the minimum length to detect polyG in the read tail. 10 by default. (int [=10])</span><br><span class="line">  -G, --disable_trim_poly_g            disable polyG tail trimming, by default trimming is automatically enabled for Illumina NextSeq/NovaSeq data</span><br><span class="line">  -x, --trim_poly_x                    enable polyX trimming in 3&apos; ends.</span><br><span class="line">      --poly_x_min_len                 the minimum length to detect polyX in the read tail. 10 by default. (int [=10])</span><br><span class="line">  -5, --cut_by_quality5                enable per read cutting by quality in front (5&apos;), default is disabled (WARNING: this will interfere deduplication for both PE/SE data)</span><br><span class="line">  -3, --cut_by_quality3                enable per read cutting by quality in tail (3&apos;), default is disabled (WARNING: this will interfere deduplication for SE data)</span><br><span class="line">  -W, --cut_window_size                the size of the sliding window for sliding window trimming, default is 4 (int [=4])</span><br><span class="line">  -M, --cut_mean_quality               the bases in the sliding window with mean quality below cutting_quality will be cut, default is Q20 (int [=20])</span><br><span class="line">  -Q, --disable_quality_filtering      quality filtering is enabled by default. If this option is specified, quality filtering is disabled</span><br><span class="line">  -q, --qualified_quality_phred        the quality value that a base is qualified. Default 15 means phred quality &gt;=Q15 is qualified. (int [=15])</span><br><span class="line">  -u, --unqualified_percent_limit      how many percents of bases are allowed to be unqualified (0~100). Default 40 means 40% (int [=40])</span><br><span class="line">  -n, --n_base_limit                   if one read&apos;s number of N base is &gt;n_base_limit, then this read/pair is discarded. Default is 5 (int [=5])</span><br><span class="line">  -L, --disable_length_filtering       length filtering is enabled by default. If this option is specified, length filtering is disabled</span><br><span class="line">  -l, --length_required                reads shorter than length_required will be discarded, default is 15. (int [=15])</span><br><span class="line">      --length_limit                   reads longer than length_limit will be discarded, default 0 means no limitation. (int [=0])</span><br><span class="line">  -y, --low_complexity_filter          enable low complexity filter. The complexity is defined as the percentage of base that is different from its next base (base[i] != base[i+1]).</span><br><span class="line">  -Y, --complexity_threshold           the threshold for low complexity filter (0~100). Default is 30, which means 30% complexity is required. (int [=30])</span><br><span class="line">      --filter_by_index1               specify a file contains a list of barcodes of index1 to be filtered out, one barcode per line (string [=])</span><br><span class="line">      --filter_by_index2               specify a file contains a list of barcodes of index2 to be filtered out, one barcode per line (string [=])</span><br><span class="line">      --filter_by_index_threshold      the allowed difference of index barcode for index filtering, default 0 means completely identical. (int [=0])</span><br><span class="line">  -c, --correction                     enable base correction in overlapped regions (only for PE data), default is disabled</span><br><span class="line">      --overlap_len_require            the minimum length of the overlapped region for overlap analysis based adapter trimming and correction. 30 by default. (int [=30])</span><br><span class="line">      --overlap_diff_limit             the maximum difference of the overlapped region for overlap analysis based adapter trimming and correction. 5 by default. (int [=5])</span><br><span class="line">  -U, --umi                            enable unique molecular identifer (UMI) preprocessing</span><br><span class="line">      --umi_loc                        specify the location of UMI, can be (index1/index2/read1/read2/per_index/per_read, default is none (string [=])</span><br><span class="line">      --umi_len                        if the UMI is in read1/read2, its length should be provided (int [=0])</span><br><span class="line">      --umi_prefix                     if specified, an underline will be used to connect prefix and UMI (i.e. prefix=UMI, UMI=AATTCG, final=UMI_AATTCG). No prefix by default (string [=])</span><br><span class="line">      --umi_skip                       if the UMI is in read1/read2, fastp can skip several bases following UMI, default is 0 (int [=0])</span><br><span class="line">  -p, --overrepresentation_analysis    enable overrepresented sequence analysis.</span><br><span class="line">  -P, --overrepresentation_sampling    one in (--overrepresentation_sampling) reads will be computed for overrepresentation analysis (1~10000), smaller is slower, default is 20. (int [=20])</span><br><span class="line">  -j, --json                           the json format report file name (string [=fastp.json])</span><br><span class="line">  -h, --html                           the html format report file name (string [=fastp.html])</span><br><span class="line">  -R, --report_title                   should be quoted with &apos; or &quot;, default is &quot;fastp report&quot; (string [=fastp report])</span><br><span class="line">  -w, --thread                         worker thread number, default is 2 (int [=2])</span><br><span class="line">  -s, --split                          split output by limiting total split file number with this option (2~999), a sequential number prefix will be added to output name ( 0001.out.fq, 0002.out.fq...), disabled by default (int [=0])</span><br><span class="line">  -S, --split_by_lines                 split output by limiting lines of each file with this option(&gt;=1000), a sequential number prefix will be added to output name ( 0001.out.fq, 0002.out.fq...), disabled by default (long [=0])</span><br><span class="line">  -d, --split_prefix_digits            the digits for the sequential number padding (1~10), default is 4, so the filename will be padded as 0001.xxx, 0 to disable padding (int[=4])</span><br><span class="line">  -?, --help                           print this message</span><br></pre></td></tr></table></figure>
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/生信理论基础/" rel="tag"># 生信理论基础</a>
          
            <a href="/tags/软件/" rel="tag"># 软件</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2018/11/14/python3标准库系列(五)-subprocess/" rel="next" title="python3 标准库系列(五)---subprocess">
                <i class="fa fa-chevron-left"></i> python3 标准库系列(五)---subprocess
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2018/11/16/Linux系列---find/" rel="prev" title="Linux命令：find">
                Linux命令：find <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>



    <div class="post-spread">
      
    </div>
  </div>


          </div>
          


          
  <div class="comments" id="comments">
    
  </div>


        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap" >
            文章目录
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview">
            站点概览
          </li>
        </ul>
      

      <section class="site-overview sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
          <img class="site-author-image" itemprop="image"
               src="/images/137356299.jpg"
               alt="Naruto" />
          <p class="site-author-name" itemprop="name">Naruto</p>
           
              <p class="site-description motion-element" itemprop="description"></p>
          
        </div>
        <nav class="site-state motion-element">

          
            <div class="site-state-item site-state-posts">
              <a href="/archives">
                <span class="site-state-item-count">46</span>
                <span class="site-state-item-name">日志</span>
              </a>
            </div>
          

          
            
            
            <div class="site-state-item site-state-categories">
              <a href="/categories/index.html">
                <span class="site-state-item-count">10</span>
                <span class="site-state-item-name">分类</span>
              </a>
            </div>
          

          
            
            
            <div class="site-state-item site-state-tags">
              <a href="/tags/index.html">
                <span class="site-state-item-count">22</span>
                <span class="site-state-item-name">标签</span>
              </a>
            </div>
          

        </nav>

        

        <div class="links-of-author motion-element">
          
            
              <span class="links-of-author-item">
                <a href="zhmqsnq90@foxmail.com" target="_blank" title="E-Mail">
                  
                    
                      E-Mail
                    
                </a>
              </span>
            
          
        </div>

        
        

        
        

        


      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#1引言"><span class="nav-number">1.</span> <span class="nav-text">1引言</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#2功能特点"><span class="nav-number">2.</span> <span class="nav-text">2功能特点</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#3轻松上手"><span class="nav-number">3.</span> <span class="nav-text">3轻松上手</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#4安装fastp"><span class="nav-number">4.</span> <span class="nav-text">4安装fastp</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#5功能介绍"><span class="nav-number">5.</span> <span class="nav-text">5功能介绍</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#5-1过滤"><span class="nav-number">5.1.</span> <span class="nav-text">5.1过滤</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#5-2接头处理"><span class="nav-number">5.2.</span> <span class="nav-text">5.2接头处理</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#5-3滑窗质量剪裁"><span class="nav-number">5.3.</span> <span class="nav-text">5.3滑窗质量剪裁</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#5-4-PE数据的碱基校正"><span class="nav-number">5.4.</span> <span class="nav-text">5.4 PE数据的碱基校正</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#5-5全局剪裁"><span class="nav-number">5.5.</span> <span class="nav-text">5.5全局剪裁</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#5-6-polyG剪裁"><span class="nav-number">5.6.</span> <span class="nav-text">5.6 polyG剪裁</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#5-7分子标签UMI处理"><span class="nav-number">5.7.</span> <span class="nav-text">5.7分子标签UMI处理</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#5-8输出文件切分"><span class="nav-number">5.8.</span> <span class="nav-text">5.8输出文件切分</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#6质控报告解读"><span class="nav-number">6.</span> <span class="nav-text">6质控报告解读</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#6-1质量分布曲线图"><span class="nav-number">6.1.</span> <span class="nav-text">6.1质量分布曲线图</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#6-2碱基含量分布曲线"><span class="nav-number">6.2.</span> <span class="nav-text">6.2碱基含量分布曲线</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#6-3-KMER统计表格"><span class="nav-number">6.3.</span> <span class="nav-text">6.3 KMER统计表格</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#6-4过表达序列-overrepresented-sequence"><span class="nav-number">6.4.</span> <span class="nav-text">6.4过表达序列(overrepresented sequence)</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#7-完整参数"><span class="nav-number">7.</span> <span class="nav-text">7 完整参数</span></a></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright" >
  
  &copy; 
  <span itemprop="copyrightYear">2018</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Naruto</span>
</div>


<div class="powered-by">
  由 <a class="theme-link" href="https://hexo.io">Hexo</a> 强力驱动
</div>

<div class="theme-info">
  主题 -
  <a class="theme-link" href="https://github.com/iissnan/hexo-theme-next">
    NexT.Mist
  </a>
</div>


        

        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>

  
  <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.2"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.2"></script>



  
  

  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.2"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.1.2"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.2"></script>



  


  




	





  





  






  





  

  

  

  
  


  

  

</body>
</html>
