<!DOCTYPE HTML>
<html lang="zh-CN">
<head>
    <!--Setting-->
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, user-scalable=no, initial-scale=1.0, maximum-scale=1.0, minimum-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1">
    <meta http-equiv="Cache-Control" content="no-siteapp">
    <meta http-equiv="Cache-Control" content="no-transform">
    <meta name="renderer" content="webkit|ie-comp|ie-stand">
    <meta name="apple-mobile-web-app-capable" content="木叶村">
    <meta name="apple-mobile-web-app-status-bar-style" content="black">
    <meta name="format-detection" content="telephone=no,email=no,adress=no">
    <meta name="browsermode" content="application">
    <meta name="screen-orientation" content="portrait">
    <link rel="dns-prefetch" href="https://jeweizhong.github.io">
    <!--SEO-->

    <meta name="keywords" content="软件" />


    <meta name="description" content="1. 简介ichorCNA是从超低通全基因组测序（ULP-WGS，0.1x覆盖率）估计无细胞DNA中肿瘤分数的工具。
描述
ichorCNA使用概率模型，作为隐马尔可夫模型（HMM）实现，以同时..." />



<meta name="robots" content="all" />
<meta name="google" content="all" />
<meta name="googlebot" content="all" />
<meta name="verify" content="all" />

    <!--Title-->


<title>ichorCNA | 木叶村</title>


    <link rel="alternate" href="/atom.xml" title="木叶村" type="application/atom+xml">


    <link rel="icon" href="/bitbug_favicon.ico">

    



<link rel="stylesheet" href="/css/bootstrap.min.css?rev=3.3.7">
<link rel="stylesheet" href="/css/font-awesome.min.css?rev=4.5.0">
<link rel="stylesheet" href="/css/style.css?rev=@@hash">




    
	<div class="hide">
		<script type="text/javascript">
			var cnzz_protocol = (("https:" == document.location.protocol) ? " https://" : " http://");document.write(unescape("%3Cspan class='cnzz_stat_icon_1263868967 hide' %3E%3Cscript%20src%3D%22https%3A%2F%2Fs95.cnzz.com%2Fz_stat.php%3Fweb_id%3D1272564536%22%3E%3C%2Fscript%3E%3C/span%3E%3Cscript src='" + cnzz_protocol + "s19.cnzz.com/z_stat.php%3Fid%3D1263868967%26show%3Dpic1' type='text/javascript'%3E%3C/script%3E"));
		</script>
	</div>






    

</head>


<!--[if lte IE 8]>
<style>
    html{ font-size: 1em }
</style>
<![endif]-->
<!--[if lte IE 9]>
<div style="ie">你使用的浏览器版本过低，为了你更好的阅读体验，请更新浏览器的版本或者使用其他现代浏览器，比如Chrome、Firefox、Safari等。</div>
<![endif]-->

<body>
    <header class="main-header"  style="background-image:url(http://snippet.shenliyang.com/img/banner.jpg)"  >
    <div class="main-header-box">
        <a class="header-avatar" href="/" title='Naruto'>
            <img src="/img/avatar.jpg" alt="logo头像" class="img-responsive center-block">
        </a>
        <div class="branding">
        	<!--<h2 class="text-hide">Snippet主题,从未如此简单有趣</h2>-->
            
                <h2> 无人为孤岛，一书一世界 </h2>
            
    	</div>
    </div>
</header>
    <nav class="main-navigation">
    <div class="container">
        <div class="row">
            <div class="col-sm-12">
                <div class="navbar-header"><span class="nav-toggle-button collapsed pull-right" data-toggle="collapse" data-target="#main-menu" id="mnav">
                    <span class="sr-only"></span>
                        <i class="fa fa-bars"></i>
                    </span>
                    <a class="navbar-brand" href="https://jeweizhong.github.io">木叶村</a>
                </div>
                <div class="collapse navbar-collapse" id="main-menu">
                    <ul class="menu">
                        
                            <li role="presentation" class="text-center">
                                <a href="/"><i class="fa "></i>首页</a>
                            </li>
                        
                            <li role="presentation" class="text-center">
                                <a href="/categories/计算机"><i class="fa "></i>计算机</a>
                            </li>
                        
                            <li role="presentation" class="text-center">
                                <a href="/categories/生信"><i class="fa "></i>生信</a>
                            </li>
                        
                            <li role="presentation" class="text-center">
                                <a href="/categories/数学与统计"><i class="fa "></i>数学与统计</a>
                            </li>
                        
                            <li role="presentation" class="text-center">
                                <a href="/archives/"><i class="fa "></i>时间轴</a>
                            </li>
                        
                    </ul>
                </div>
            </div>
        </div>
    </div>
</nav>
    <section class="content-wrap">
        <div class="container">
            <div class="row">
                <main class="col-md-8 main-content m-post">
                    <p id="process"></p>
<article class="post">
    <div class="post-head">
        <h1 id="ichorCNA">
            
	            ichorCNA
            
        </h1>
        <div class="post-meta">
    
        <span class="categories-meta fa-wrap">
            <i class="fa fa-folder-open-o"></i>
            <a class="category-link" href="/categories/生信/">生信</a>
        </span>
    

    
        <span class="fa-wrap">
            <i class="fa fa-tags"></i>
            <span class="tags-meta">
                
                    <a class="tag-link" href="/tags/软件/">软件</a>
                
            </span>
        </span>
    

    
        
        <span class="fa-wrap">
            <i class="fa fa-clock-o"></i>
            <span class="date-meta">2018/02/13</span>
        </span>
        
    
</div>
            
            
            <p class="fa fa-exclamation-triangle warning">
                本文于<strong>457</strong>天之前发表，文中内容可能已经过时。
            </p>
        
    </div>
    
    <div class="post-body post-content">
        <h4 id="1-简介"><a href="#1-简介" class="headerlink" title="1. 简介"></a>1. 简介</h4><p>ichorCNA是从超低通全基因组测序（ULP-WGS，0.1x覆盖率）估计无细胞DNA中肿瘤分数的工具。</p>
<p><strong>描述</strong></p>
<p>ichorCNA使用概率模型，作为隐马尔可夫模型（HMM）实现，以同时分割基因组，预测大规模拷贝数改变，并估计超低通全基因组测序样本（ULP-WGS ）。ichorCNA针对样本的低覆盖率（〜0.1x）测序进行了优化，并使用患者和健康供体cfDNA样本进行了基准测试。</p>
<p><strong>用途</strong></p>
<p>ichorCNA可用于通知肿瘤来源DNA的存在或缺失，并指导决定进行整个外显子或更深的全基因组测序。此外，我们可以使用肿瘤分数的定量估计来校准期望的测序深度，以达到用于鉴定无细胞DNA中的突变的统计功效。最后，ichorCNA可以利用超低通量测序的成本效益方法来检测来自大群体的大规模拷贝数变化。</p>
<p><strong>文献</strong></p>
<p>Adalsteinsson, Ha, Freeman, et al. Scalable whole-exome sequencing of cell-free DNA reveals high concordance with metastatic tumors</p>
<h4 id="2-安装"><a href="#2-安装" class="headerlink" title="2. 安装"></a>2. 安装</h4><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">#R控制台中</span><br><span class="line">install.packages(&quot;plyr&quot;)</span><br><span class="line">source(&quot;https://bioconductor.org/biocLite.R&quot;)</span><br><span class="line">biocLite(&quot;HMMcopy&quot;)</span><br><span class="line">biocLite(&quot;GenomeInfoDb&quot;)</span><br><span class="line">install.packages(&quot;devtools&quot;)</span><br><span class="line">library(devtools)</span><br><span class="line">install_github(&quot;broadinstitute/ichorCNA&quot;, &quot;--no-docs&quot;)</span><br><span class="line">#依赖的包有很多，安装过程报错信息可以看出需要依赖的其他包</span><br></pre></td></tr></table></figure>
<h4 id="3-使用"><a href="#3-使用" class="headerlink" title="3. 使用"></a>3. 使用</h4><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line">#以样本DYDFA-842-31为例</span><br><span class="line">$bwa mem -t 4 -B 3 -O 11 -E 4 -k 32 -Y hg19_reference_with_NC.fasta </span><br><span class="line">DYDFA-842-31/DYDFA-842-31_R1.fq.gz DYDFA-842-31_R2.fq.gz &gt; ichorCNA/test31/31.sam</span><br><span class="line"></span><br><span class="line">$samtools view -bS ichorCNA/test31/31.sam &gt; ichorCNA/test31/31.bam</span><br><span class="line"></span><br><span class="line">$samtools sort 31.bam &gt; 31.sort.bam &amp;</span><br><span class="line"></span><br><span class="line">$samtools  index 31.sort.bam &amp;</span><br><span class="line"></span><br><span class="line">$java -jar picard-2.8.2.jar MarkDuplicates I=./31.sort.bam O=./31.rmdup.bam M=rmdup_stat.txt REMOVE_DUPLICATES=true &amp;</span><br><span class="line"></span><br><span class="line">$samtools index 31.rmdup.bam  &amp;</span><br><span class="line"></span><br><span class="line">hmmcopy_utils-master/bin/readCounter --window 1000000 --quality 20 --chromosome chr1,chr2,chr3,chr4,chr5,chr6,chr7,chr8,chr9,chr10,chr11,chr12,chr13,chr14,chr15,chr16,chr17,chr18,chr19,chr20,chr21,chr22,chrX,chrY 31.rmdup.bam &gt; 31.wig &amp;</span><br><span class="line"></span><br><span class="line">$Rscript ichorCNA-master/scripts/runIchorCNA.R --id test31 --WIG test31.wig \</span><br><span class="line">    --gcWig ichorCNA-master/inst/extdata/gc_hg19_1000kb.wig \</span><br><span class="line">    --mapWig ichorCNA-master/inst/extdata/map_hg19_1000kb.wig \</span><br><span class="line">    --centromere ichorCNA-master/inst/extdata/GRCh37.p13_centromere_UCSC-gapTable.txt \</span><br><span class="line">    --normalPanel ichorCNA-master/inst/extdata/HD_ULP_PoN_1Mb_median_normAutosome_mapScoreFiltered_median.rds \</span><br><span class="line">    --normal &quot;c(0.5,0.6,0.7,0.8,0.9)&quot;  --scStates &quot;c(1,3)&quot; \</span><br><span class="line">    --txnE 0.9999 --txnStrength 10000 --outDir ./</span><br></pre></td></tr></table></figure>
<h4 id="4-输出"><a href="#4-输出" class="headerlink" title="4. 输出"></a>4. 输出</h4><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">test31</span><br><span class="line">Gender: male</span><br><span class="line">Tumor Fraction: 0</span><br><span class="line">Ploidy: 2.001</span><br><span class="line">Subclone Fraction:      NA</span><br></pre></td></tr></table></figure>
<p><img src="/myphoto/clipboard31.png" alt="31"><br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">test32</span><br><span class="line">Gender: male</span><br><span class="line">Tumor Fraction: 0</span><br><span class="line">Ploidy: 2</span><br><span class="line">Subclone Fraction:      NA</span><br></pre></td></tr></table></figure></p>
<p><img src="/myphoto/clipboard32.png" alt="32"><br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">test33</span><br><span class="line">Gender: female</span><br><span class="line">Tumor Fraction: 0</span><br><span class="line">Ploidy: 2</span><br><span class="line">Subclone Fraction:      NA</span><br></pre></td></tr></table></figure></p>
<p><img src="/myphoto/clipboard33.png" alt="33"><br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">test34</span><br><span class="line">Gender: female</span><br><span class="line">Tumor Fraction: 0</span><br><span class="line">Ploidy: 1.999</span><br><span class="line">Subclone Fraction:      NA</span><br></pre></td></tr></table></figure></p>
<p><img src="/myphoto/clipboard34.png" alt="34"><br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">test35</span><br><span class="line">Gender: female</span><br><span class="line">Tumor Fraction: 0.404</span><br><span class="line">Ploidy: 2.98</span><br><span class="line">Subclone Fraction:      0.5017</span><br></pre></td></tr></table></figure></p>
<p><img src="/myphoto/clipboard35.png" alt="35"></p>
<h4 id="5-解释结果"><a href="#5-解释结果" class="headerlink" title="5. 解释结果"></a>5. 解释结果</h4><h4 id="6-帮助文档"><a href="#6-帮助文档" class="headerlink" title="6. 帮助文档"></a>6. 帮助文档</h4><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br></pre></td><td class="code"><pre><span class="line">Rscript runIchorCNA.R --help</span><br><span class="line">Usage: runIchorCNA.R [options]</span><br><span class="line"></span><br><span class="line">Options:</span><br><span class="line">--WIG=WIG</span><br><span class="line">	WIG 文件(必须)</span><br><span class="line"></span><br><span class="line">--NORMWIG=NORMWIG</span><br><span class="line">	标准 WIG 文件 Default: [NULL]</span><br><span class="line"></span><br><span class="line">--gcWig=GCWIG</span><br><span class="line">	GC-content WIG 文件; (必须);软件给出的是hg19的GC WIG文件</span><br><span class="line"></span><br><span class="line">--mapWig=MAPWIG</span><br><span class="line">	 WIG 文件  Default: [NULL]</span><br><span class="line"></span><br><span class="line">--normalPanel=NORMALPANEL</span><br><span class="line">	深度矫正文件 Default: [NULL]</span><br><span class="line"></span><br><span class="line">--exons.bed=EXONS.BED</span><br><span class="line">	外显子bed文件 Default: [NULL]</span><br><span class="line"></span><br><span class="line">--id=ID</span><br><span class="line">	样本名 Default: [test]</span><br><span class="line"></span><br><span class="line">--centromere=CENTROMERE</span><br><span class="line">	着丝粒位置文件;ichorCNA提供的是hg19版本的. Default: [NULL]</span><br><span class="line"></span><br><span class="line">--rmCentromereFlankLength=RMCENTROMEREFLANKLENGTH</span><br><span class="line">    去除着丝粒侧翼的长度. Default: [1e+05]</span><br><span class="line"></span><br><span class="line">--normal=NORMAL</span><br><span class="line">	Initial normal contamination; can be more than one value if additional normal initializations are desired. Default: [0.5]</span><br><span class="line">    #初始正常污染; 如果需要额外的正常初始化，可以是多于一个值</span><br><span class="line">--scStates=SCSTATES</span><br><span class="line">	亚克隆状态. Default: [NULL]</span><br><span class="line"></span><br><span class="line">--coverage=COVERAGE</span><br><span class="line">	PICARD 软件计算出的覆盖度文件. Default: [NULL]</span><br><span class="line"></span><br><span class="line">--lambda=LAMBDA</span><br><span class="line">	Initial Student&apos;s t precision; must contain 4 values (e.g. c(1500,1500,1500,1500)); if not provided then will automatically use based on variance of data. Default: [NULL]</span><br><span class="line">    # 初始T-test精确度; 必须包含4个值（例如c（1500,1500,1500,1500））;如果没有提供，则会根据数据的方差自动使用</span><br><span class="line">--lambdaScaleHyperParam=LAMBDASCALEHYPERPARAM</span><br><span class="line">	Hyperparameter (scale) for Gamma prior on Student&apos;s-t precision. Default: [3]</span><br><span class="line">    #Gamma之前的超参数（规模）的t精度</span><br><span class="line">--ploidy=PLOIDY</span><br><span class="line">	Initial tumour ploidy; can be more than one value if additional ploidy initializations are desired. Default: [2]</span><br><span class="line">    # 最初的肿瘤倍性; 默认二倍体；</span><br><span class="line">--maxCN=MAXCN</span><br><span class="line">	Total clonal CN states. Default: [7]</span><br><span class="line"></span><br><span class="line">--estimateNormal=ESTIMATENORMAL</span><br><span class="line">	Estimate normal. Default: [TRUE]</span><br><span class="line"></span><br><span class="line">--estimateScPrevalence=ESTIMATESCPREVALENCE</span><br><span class="line">	Estimate subclonal prevalence. Default: [TRUE]</span><br><span class="line"></span><br><span class="line">--estimatePloidy=ESTIMATEPLOIDY</span><br><span class="line">	Estimate tumour ploidy. Default: [TRUE]</span><br><span class="line"></span><br><span class="line">--maxFracCNASubclone=MAXFRACCNASUBCLONE</span><br><span class="line">	Exclude solutions with fraction of subclonal events greater than this value. Default: [0.7]</span><br><span class="line">    # 排除大于此值的部分亚克隆事件的解决方案</span><br><span class="line">--maxFracGenomeSubclone=MAXFRACGENOMESUBCLONE</span><br><span class="line">	Exclude solutions with subclonal genome fraction greater than this value. Default: [0.5]</span><br><span class="line">    # 排除亚克隆基因组分数大于此值的解决方案</span><br><span class="line">--minSegmentBins=MINSEGMENTBINS</span><br><span class="line">	Minimum number of bins for largest segment threshold required to estimate tumor fraction; if below this threshold, then will be assigned zero tumor fraction.</span><br><span class="line"></span><br><span class="line">--altFracThreshold=ALTFRACTHRESHOLD</span><br><span class="line">	Minimum proportion of bins altered required to estimate tumor fraction; if below this threshold, then will be assigned zero tumor fraction. Default: [0.05]</span><br><span class="line">    # 评估肿瘤分数所需的最大分割阈值的最小分箱数量; 如果低于这个阈值，则将被分配零肿瘤分数</span><br><span class="line">--chrNormalize=CHRNORMALIZE</span><br><span class="line">	Specify chromosomes to normalize GC/mappability biases. Default: [c(1:22)]</span><br><span class="line">    #指定染色体</span><br><span class="line">--chrTrain=CHRTRAIN</span><br><span class="line">	Specify chromosomes to estimate params. Default: [c(1:22)]</span><br><span class="line">    #指定染色体</span><br><span class="line">--chrs=CHRS</span><br><span class="line">	Specify chromosomes to analyze. Default: [c(1:22,&quot;X&quot;)]</span><br><span class="line">    #指定染色体</span><br><span class="line">--normalizeMaleX=NORMALIZEMALEX</span><br><span class="line">	If male, then normalize chrX by median. Default: [TRUE]</span><br><span class="line">    # 男性按照X染色体分析</span><br><span class="line">--fracReadsInChrYForMale=FRACREADSINCHRYFORMALE</span><br><span class="line">	Threshold for fraction of reads in chrY to assign as male. Default: [0.001]</span><br><span class="line">    #将chrY中的读数部分的阈值分配为男性</span><br><span class="line">--includeHOMD=INCLUDEHOMD</span><br><span class="line">	If FALSE, then exclude HOMD state. Useful when using large bins (e.g. 1Mb). Default: [FALSE]</span><br><span class="line">    # 如果FALSE，则排除HOMD状态。 使用大容量箱（如1Mb）时很有用</span><br><span class="line">--txnE=TXNE</span><br><span class="line">	Self-transition probability. Increase to decrease number of segments. Default: [0.9999999]</span><br><span class="line">    #　自我转换概率。 增加以减少段数</span><br><span class="line">--txnStrength=TXNSTRENGTH</span><br><span class="line">	Transition pseudo-counts. Exponent should be the same as the number of decimal places of --txnE. Default: [1e+07]</span><br><span class="line">    # 转换伪计数。 指数应与--txnE的小数位数相同</span><br><span class="line">--plotFileType=PLOTFILETYPE</span><br><span class="line">	File format for output plots. Default: [pdf]</span><br><span class="line">    #图片输出格式</span><br><span class="line">--plotYLim=PLOTYLIM</span><br><span class="line">	ylim to use for chromosome plots. Default: [c(-2,2)]</span><br><span class="line">    y轴范围</span><br><span class="line">--outDir=OUTDIR</span><br><span class="line">	Output Directory. Default: [./]</span><br><span class="line"></span><br><span class="line">--libdir=LIBDIR</span><br><span class="line">	Script library path. Usually exclude this argument unless custom modifications have been made to the ichorCNA R package code and the user would like to source those R files. Default: [NULL]</span><br></pre></td></tr></table></figure>
    </div>
    
    <div class="post-footer">
        <div>
            
                转载声明：商业转载请联系作者获得授权,非商业转载请注明出处 © <a href="" target="_blank">Snippet</a>
            
        </div>
        <div>
            
        </div>
    </div>
</article>

<div class="article-nav prev-next-wrap clearfix">
    
        <a href="/2018/02/13/perl-dan-xing-ming-ling/" class="pre-post btn btn-default" title='perl 单行命令'>
            <i class="fa fa-angle-left fa-fw"></i><span class="hidden-lg">上一篇</span>
            <span class="hidden-xs">perl 单行命令</span>
        </a>
    
    
        <a href="/2018/02/13/zhuan-zai-linux-xia-r-an-zhuang-wu-root-quan-xian/" class="next-post btn btn-default" title='(转载)linux下R安装(无root权限)'>
            <span class="hidden-lg">下一篇</span>
            <span class="hidden-xs">(转载)linux下R安装(无root权限)</span><i class="fa fa-angle-right fa-fw"></i>
        </a>
    
</div>


    <div id="comments">
        
   <p>评论系统未开启，无法评论！</p>

    </div>





                </main>
                
                    <aside id="article-toc" role="navigation" class="col-md-4">
    <div class="widget">
        <h3 class="title">文章目录</h3>
        
            <ol class="toc"><li class="toc-item toc-level-4"><a class="toc-link" href="#1-简介"><span class="toc-text">1. 简介</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#2-安装"><span class="toc-text">2. 安装</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#3-使用"><span class="toc-text">3. 使用</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#4-输出"><span class="toc-text">4. 输出</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#5-解释结果"><span class="toc-text">5. 解释结果</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#6-帮助文档"><span class="toc-text">6. 帮助文档</span></a></li></ol>
        
    </div>
</aside>

                
            </div>
        </div>
    </section>
    <footer class="main-footer">
    <div class="container">
        <div class="row">
        </div>
    </div>
</footer>

<a id="back-to-top" class="icon-btn hide">
	<i class="fa fa-chevron-up"></i>
</a>




    <div class="copyright">
    <div class="container">
        <div class="row">
            <div class="col-sm-12">
                <div class="busuanzi">
    
</div>

            </div>
            <div class="col-sm-12">
                <span>Copyright &copy; 2017
                </span> |
                <span>
                    Powered by <a href="//hexo.io" class="copyright-links" target="_blank" rel="nofollow">Hexo</a>
                </span> |
                <span>
                    Theme by <a href="//github.com/shenliyang/hexo-theme-snippet.git" class="copyright-links" target="_blank" rel="nofollow">Snippet</a>
                </span>
            </div>
        </div>
    </div>
</div>







<script src="/js/app.js?rev=@@hash"></script>

<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        tex2jax: {
            inlineMath: [ ["$","$"], ["\\(","\\)"] ],
            skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code'],
            processEscapes: true
        }
    });
    MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax();
        for (var i = 0; i < all.length; ++i)
            all[i].SourceElement().parentNode.className += ' has-jax';
    });
</script>
<script src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
</body>
</html>